--- a/minecraft_server/net/minecraft/src/BehaviorDispenseArmor.java
+++ b/minecraft_server/net/minecraft/src/BehaviorDispenseArmor.java
@@ -7,6 +7,8 @@
     /**
      * Dispense the specified stack, play the dispense sound and spawn particles.
      */
+	// FCMOD: Code removed to prevent usage
+	/*
     protected ItemStack dispenseStack(IBlockSource par1IBlockSource, ItemStack par2ItemStack)
     {
         EnumFacing var3 = BlockDispenser.getFacing(par1IBlockSource.getBlockMetadata());
@@ -33,4 +35,6 @@
             return super.dispenseStack(par1IBlockSource, par2ItemStack);
         }
     }
+    */
+    // END FCMOD
 }
