--- a/minecraft_server/net/minecraft/src/BlockLog.java
+++ b/minecraft_server/net/minecraft/src/BlockLog.java
@@ -123,4 +123,12 @@
     {
         return new ItemStack(this.blockID, 1, limitToValidMetadata(par1));
     }
+    // FCMOD: Code added
+    protected BlockLog( int iBlockID, Material material )
+    {
+        super(iBlockID, material );
+        
+        setCreativeTab( CreativeTabs.tabBlock );
 }
+    // END FCMOD
+}
