--- a/minecraft_server/net/minecraft/src/BlockTripWire.java
+++ b/minecraft_server/net/minecraft/src/BlockTripWire.java
@@ -128,7 +128,7 @@
     {
         if (!par1World.isRemote)
         {
-            if (par6EntityPlayer.getCurrentEquippedItem() != null && par6EntityPlayer.getCurrentEquippedItem().itemID == Item.shears.itemID)
+        	if (par6EntityPlayer.getCurrentEquippedItem() != null && par6EntityPlayer.getCurrentEquippedItem().getItem() instanceof FCItemShears)
             {
                 par1World.setBlockMetadata(par2, par3, par4, par5 | 8, 4);
             }
