--- a/minecraft_server/net/minecraft/src/BlockHalfSlab.java
+++ b/minecraft_server/net/minecraft/src/BlockHalfSlab.java
@@ -15,18 +15,26 @@
         if (par2)
         {
             opaqueCubeLookup[par1] = true;
+            
+            // FCMOD: Added
+            InitBlockBounds( 0D, 0D, 0D, 1D, 1D, 1D );
+            // END FCMOD
         }
         else
         {
-            this.setBlockBounds(0.0F, 0.0F, 0.0F, 1.0F, 0.5F, 1.0F);
+        	// FCMOD: Changed
+            //this.setBlockBounds(0.0F, 0.0F, 0.0F, 1.0F, 0.5F, 1.0F);
+            InitBlockBounds( 0D, 0D, 0D, 1D, 0.5D, 1D );
+            
+            useNeighborBrightness[par1] = true;
+            // END FCMOD
         }
 
         this.setLightOpacity(255);
     }
 
-    /**
-     * Updates the blocks bounds based on its current state. Args: world, x, y, z
-     */
+    // FCMOD: Removed
+    /*
     public void setBlockBoundsBasedOnState(IBlockAccess par1IBlockAccess, int par2, int par3, int par4)
     {
         if (this.isDoubleSlab)
@@ -48,9 +56,6 @@
         }
     }
 
-    /**
-     * Sets the block's bounds for rendering it as an item
-     */
     public void setBlockBoundsForItemRender()
     {
         if (this.isDoubleSlab)
@@ -63,15 +68,13 @@
         }
     }
 
-    /**
-     * Adds all intersecting collision boxes to a list. (Be sure to only add boxes to the list if they intersect the
-     * mask.) Parameters: World, X, Y, Z, mask, list, colliding entity
-     */
     public void addCollisionBoxesToList(World par1World, int par2, int par3, int par4, AxisAlignedBB par5AxisAlignedBB, List par6List, Entity par7Entity)
     {
         this.setBlockBoundsBasedOnState(par1World, par2, par3, par4);
         super.addCollisionBoxesToList(par1World, par2, par3, par4, par5AxisAlignedBB, par6List, par7Entity);
     }
+    */
+    // END FCMOD
 
     /**
      * Is this block (a) opaque and (b) a full 1m cube?  This determines whether or not to render the shared face of two
@@ -126,4 +129,108 @@
     {
         return super.getDamageValue(par1World, par2, par3, par4) & 7;
     }
+    
+	// FCMOD: Added New
+    @Override
+    public AxisAlignedBB GetBlockBoundsFromPoolBasedOnState( 
+    	IBlockAccess blockAccess, int i, int j, int k )
+    {
+    	AxisAlignedBB bounds;
+    	
+        if ( isDoubleSlab )
+        {
+            bounds = AxisAlignedBB.getAABBPool().getAABB( 0F, 0F, 0F, 1F, 1F, 1F );
 }
+        else
+        {
+            if ( GetIsUpsideDown( blockAccess, i, j, k ) )
+            {
+                bounds = AxisAlignedBB.getAABBPool().getAABB( 0F, 0.5F, 0F, 1F, 1F, 1F );
+            }
+            else
+            {
+            	bounds = AxisAlignedBB.getAABBPool().getAABB( 0F, 0F, 0F, 1F, 0.5F, 1F );
+            }
+        }
+        
+        return bounds;
+    }
+    
+    @Override
+	public boolean HasLargeCenterHardPointToFacing( IBlockAccess blockAccess, int i, int j, int k, int iFacing, boolean bIgnoreTransparency )
+	{
+    	if ( !isDoubleSlab )
+    	{
+            boolean bIsUpsideDown = GetIsUpsideDown( blockAccess, i, j, k );
+            
+            if ( iFacing == 0  ) 
+            {
+            	if ( !bIsUpsideDown )
+            	{
+            		return true;
+            	}
+            }
+            else if ( iFacing == 1 )
+            {
+            	if ( bIsUpsideDown )
+            	{
+            		return true;
+            	}
+            }
+    	}
+    	
+		return super.HasLargeCenterHardPointToFacing( blockAccess, i, j, k, iFacing, bIgnoreTransparency );
+	}
+
+    @Override
+    public float GroundCoverRestingOnVisualOffset( IBlockAccess blockAccess, int i, int j, int k )
+    {
+        if ( !isDoubleSlab )
+        {
+            if ( ( blockAccess.getBlockMetadata( i, j, k ) & 8 ) == 0 )
+            {
+            	return -0.5F;
+            }
+        }
+        
+    	return 0F;
+    }
+    
+    @Override
+    public boolean CanGroundCoverRestOnBlock( World world, int i, int j, int k )
+    {
+    	return true;
+    }
+    
+    @Override
+    public boolean CanMobsSpawnOn( World world, int i, int j, int k )
+    {
+        return blockMaterial.GetMobsCanSpawnOn( world.provider.dimensionId );
+    }
+
+    @Override
+    public float MobSpawnOnVerticalOffset( World world, int i, int j, int k )
+    {
+        if ( !isDoubleSlab && !GetIsUpsideDown( world, i, j, k ) )
+        {
+        	return -0.5F;
+        }
+        
+    	return 0F;
+    }
+    
+    //------------- Class Specific Methods ------------//
+    
+    public boolean GetIsUpsideDown( IBlockAccess blockAccess, int i, int j, int k )
+    {
+    	return GetIsUpsideDown( blockAccess.getBlockMetadata( i, j, k ) );
+    }
+    
+    public boolean GetIsUpsideDown( int iMetadata )
+    {
+    	return ( iMetadata & 8 ) > 0;
+    }
+    
+	//----------- Client Side Functionality -----------//
+	// END FCMOD
+}
