--- a/minecraft_server/net/minecraft/src/BlockRedstoneLogic.java
+++ b/minecraft_server/net/minecraft/src/BlockRedstoneLogic.java
@@ -11,7 +11,10 @@
     {
         super(par1, Material.circuits);
         this.isRepeaterPowered = par2;
-        this.setBlockBounds(0.0F, 0.0F, 0.0F, 1.0F, 0.125F, 1.0F);
+        // FCMOD: Changed
+        //this.setBlockBounds(0.0F, 0.0F, 0.0F, 1.0F, 0.125F, 1.0F);
+        InitBlockBounds( 0.0F, 0.0F, 0.0F, 1.0F, 0.125F, 1.0F );
+        // END FCMOD
     }
 
     /**
@@ -346,4 +349,8 @@
     {
         return this.func_94487_f(par1);
     }
+    
+    // FCMOD: Added New
+	//----------- Client Side Functionality -----------//
+    // END FCMOD
 }
