--- a/minecraft_server/net/minecraft/src/BlockSand.java
+++ b/minecraft_server/net/minecraft/src/BlockSand.java
@@ -59,7 +59,7 @@
             {
                 if (!par1World.isRemote)
                 {
-                    EntityFallingSand var9 = new EntityFallingSand(par1World, (double)((float)par2 + 0.5F), (double)((float)par3 + 0.5F), (double)((float)par4 + 0.5F), this.blockID, par1World.getBlockMetadata(par2, par3, par4));
+                	EntityFallingSand var9 = (EntityFallingSand) EntityList.createEntityOfType(EntityFallingSand.class, par1World, (double)((float)par2 + 0.5F), (double)((float)par3 + 0.5F), (double)((float)par4 + 0.5F), this.blockID, par1World.getBlockMetadata(par2, par3, par4));
                     this.onStartFalling(var9);
                     par1World.spawnEntityInWorld(var9);
                 }
