--- a/minecraft_server/net/minecraft/src/BlockDoor.java
+++ b/minecraft_server/net/minecraft/src/BlockDoor.java
@@ -365,4 +365,11 @@
             par1World.setBlockToAir(par2, par3 - 1, par4);
         }
     }
+    
+    // FCMOD: Code added
+    public void OnAIOpenDoor( World world, int i, int j, int k, boolean bOpen )
+    {
+    	onPoweredBlockChange( world, i, j, k, bOpen );    	
 }
+    // END FCMOD
+}
