--- a/minecraft_server/net/minecraft/src/CommandGive.java
+++ b/minecraft_server/net/minecraft/src/CommandGive.java
@@ -49,6 +49,9 @@
                 }
 
                 ItemStack var7 = new ItemStack(var4, var5, var6);
+                // FCMOD: Added
+                var7.getItem().InitializeStackOnGiveCommand( var3.worldObj.rand, var7 );
+                // END FCMOD
                 EntityItem var8 = var3.dropPlayerItem(var7);
                 var8.delayBeforeCanPickup = 0;
                 notifyAdmins(par1ICommandSender, "commands.give.success", new Object[] {Item.itemsList[var4].func_77653_i(var7), Integer.valueOf(var4), Integer.valueOf(var5), var3.getEntityName()});
