--- a/minecraft_server/net/minecraft/src/ComponentStrongholdLibrary.java
+++ b/minecraft_server/net/minecraft/src/ComponentStrongholdLibrary.java
@@ -111,6 +111,8 @@
                 this.placeBlockAtCurrentPosition(par1World, Block.fence.blockID, 0, 9, 6, 11, par3StructureBoundingBox);
                 this.placeBlockAtCurrentPosition(par1World, Block.fence.blockID, 0, 8, 6, 11, par3StructureBoundingBox);
                 this.placeBlockAtCurrentPosition(par1World, Block.fence.blockID, 0, 9, 6, 10, par3StructureBoundingBox);
+                // FCMOD: Changed
+                /*
                 var7 = this.getMetadataWithOffset(Block.ladder.blockID, 3);
                 this.placeBlockAtCurrentPosition(par1World, Block.ladder.blockID, var7, 10, 1, 13, par3StructureBoundingBox);
                 this.placeBlockAtCurrentPosition(par1World, Block.ladder.blockID, var7, 10, 2, 13, par3StructureBoundingBox);
@@ -119,6 +121,18 @@
                 this.placeBlockAtCurrentPosition(par1World, Block.ladder.blockID, var7, 10, 5, 13, par3StructureBoundingBox);
                 this.placeBlockAtCurrentPosition(par1World, Block.ladder.blockID, var7, 10, 6, 13, par3StructureBoundingBox);
                 this.placeBlockAtCurrentPosition(par1World, Block.ladder.blockID, var7, 10, 7, 13, par3StructureBoundingBox);
+                */
+                int iFacing = getMetadataWithOffset( Block.ladder.blockID, 3 ); // legacy ladder metadata is straight facing
+                int iMetadata = FCBetterThanWolves.fcBlockLadder.SetFacing( 0, iFacing );                
+                
+                placeBlockAtCurrentPosition(par1World, FCBetterThanWolves.fcBlockLadder.blockID, iMetadata, 10, 1, 13, par3StructureBoundingBox);
+                placeBlockAtCurrentPosition(par1World, FCBetterThanWolves.fcBlockLadder.blockID, iMetadata, 10, 2, 13, par3StructureBoundingBox);
+                placeBlockAtCurrentPosition(par1World, FCBetterThanWolves.fcBlockLadder.blockID, iMetadata, 10, 3, 13, par3StructureBoundingBox);
+                placeBlockAtCurrentPosition(par1World, FCBetterThanWolves.fcBlockLadder.blockID, iMetadata, 10, 4, 13, par3StructureBoundingBox);
+                placeBlockAtCurrentPosition(par1World, FCBetterThanWolves.fcBlockLadder.blockID, iMetadata, 10, 5, 13, par3StructureBoundingBox);
+                placeBlockAtCurrentPosition(par1World, FCBetterThanWolves.fcBlockLadder.blockID, iMetadata, 10, 6, 13, par3StructureBoundingBox);
+                placeBlockAtCurrentPosition(par1World, FCBetterThanWolves.fcBlockLadder.blockID, iMetadata, 10, 7, 13, par3StructureBoundingBox);
+                // END FCMOD
                 byte var8 = 7;
                 byte var9 = 7;
                 this.placeBlockAtCurrentPosition(par1World, Block.fence.blockID, 0, var8 - 1, 9, var9, par3StructureBoundingBox);
