--- a/minecraft_server/net/minecraft/src/ComponentVillageHouse4_Garden.java
+++ b/minecraft_server/net/minecraft/src/ComponentVillageHouse4_Garden.java
@@ -58,9 +58,27 @@
         this.fillWithBlocks(par1World, par3StructureBoundingBox, 0, 1, 1, 0, 3, 3, Block.planks.blockID, Block.planks.blockID, false);
         this.fillWithBlocks(par1World, par3StructureBoundingBox, 4, 1, 1, 4, 3, 3, Block.planks.blockID, Block.planks.blockID, false);
         this.fillWithBlocks(par1World, par3StructureBoundingBox, 1, 1, 4, 3, 3, 4, Block.planks.blockID, Block.planks.blockID, false);
+        // FCMOD: Code change to remove glass from abandoned villages
+        /*
         this.placeBlockAtCurrentPosition(par1World, Block.thinGlass.blockID, 0, 0, 2, 2, par3StructureBoundingBox);
         this.placeBlockAtCurrentPosition(par1World, Block.thinGlass.blockID, 0, 2, 2, 4, par3StructureBoundingBox);
         this.placeBlockAtCurrentPosition(par1World, Block.thinGlass.blockID, 0, 4, 2, 2, par3StructureBoundingBox);
+        */
+        int iAbandonmentLevel = startPiece.GetAbandonmentLevel( par1World );
+        
+        if ( iAbandonmentLevel == 0 )
+    	{
+            this.placeBlockAtCurrentPosition(par1World, Block.thinGlass.blockID, 0, 0, 2, 2, par3StructureBoundingBox);
+            this.placeBlockAtCurrentPosition(par1World, Block.thinGlass.blockID, 0, 2, 2, 4, par3StructureBoundingBox);
+            this.placeBlockAtCurrentPosition(par1World, Block.thinGlass.blockID, 0, 4, 2, 2, par3StructureBoundingBox);
+    	}
+        else
+        {
+            this.placeBlockAtCurrentPosition(par1World, 0, 0, 0, 2, 2, par3StructureBoundingBox);
+            this.placeBlockAtCurrentPosition(par1World, 0, 0, 2, 2, 4, par3StructureBoundingBox);
+            this.placeBlockAtCurrentPosition(par1World, 0, 0, 4, 2, 2, par3StructureBoundingBox);
+        }
+        // END FCMOD
         this.placeBlockAtCurrentPosition(par1World, Block.planks.blockID, 0, 1, 1, 0, par3StructureBoundingBox);
         this.placeBlockAtCurrentPosition(par1World, Block.planks.blockID, 0, 1, 2, 0, par3StructureBoundingBox);
         this.placeBlockAtCurrentPosition(par1World, Block.planks.blockID, 0, 1, 3, 0, par3StructureBoundingBox);
@@ -100,14 +118,31 @@
 
         if (this.isRoofAccessible)
         {
+            // FCMOD: Changed
+        	/*
             var4 = this.getMetadataWithOffset(Block.ladder.blockID, 3);
             this.placeBlockAtCurrentPosition(par1World, Block.ladder.blockID, var4, 3, 1, 3, par3StructureBoundingBox);
             this.placeBlockAtCurrentPosition(par1World, Block.ladder.blockID, var4, 3, 2, 3, par3StructureBoundingBox);
             this.placeBlockAtCurrentPosition(par1World, Block.ladder.blockID, var4, 3, 3, 3, par3StructureBoundingBox);
             this.placeBlockAtCurrentPosition(par1World, Block.ladder.blockID, var4, 3, 4, 3, par3StructureBoundingBox);
+            */
+            int iFacing = getMetadataWithOffset( Block.ladder.blockID, 3 ); // legacy ladder metadata is straight facing
+            int iMetadata = FCBetterThanWolves.fcBlockLadder.SetFacing( 0, iFacing );                
+            
+            placeBlockAtCurrentPosition( par1World, FCBetterThanWolves.fcBlockLadder.blockID, iMetadata, 3, 1, 3, par3StructureBoundingBox );
+            placeBlockAtCurrentPosition( par1World, FCBetterThanWolves.fcBlockLadder.blockID, iMetadata, 3, 2, 3, par3StructureBoundingBox );
+            placeBlockAtCurrentPosition( par1World, FCBetterThanWolves.fcBlockLadder.blockID, iMetadata, 3, 3, 3, par3StructureBoundingBox );
+            placeBlockAtCurrentPosition( par1World, FCBetterThanWolves.fcBlockLadder.blockID, iMetadata, 3, 4, 3, par3StructureBoundingBox );
+            // END FCMOD
         }
 
-        this.placeBlockAtCurrentPosition(par1World, Block.torchWood.blockID, 0, 2, 3, 1, par3StructureBoundingBox);
+        // FCMOD: Changed
+        //this.placeBlockAtCurrentPosition(par1World, Block.torchWood.blockID, 0, 2, 3, 1, par3StructureBoundingBox);
+        if ( startPiece.GetAbandonmentLevel( par1World ) <= 1 )
+        {
+            this.placeBlockAtCurrentPosition(par1World, FCBetterThanWolves.fcBlockTorchFiniteUnlit.blockID, 8, 2, 3, 1, par3StructureBoundingBox);
+        }
+        // END FCMOD
 
         for (var4 = 0; var4 < 5; ++var4)
         {
