--- a/minecraft_server/net/minecraft/src/ComponentVillage.java
+++ b/minecraft_server/net/minecraft/src/ComponentVillage.java
@@ -107,6 +107,30 @@
      */
     protected void spawnVillagers(World par1World, StructureBoundingBox par2StructureBoundingBox, int par3, int par4, int par5, int par6)
     {
+    	// FCMOD: Code added
+    	int iAbandonmentLevel = startPiece.GetAbandonmentLevel( par1World );
+    	
+    	boolean bDirtyPeasants = false;
+    	
+    	if ( iAbandonmentLevel > 1 )
+    	{
+    		if ( par1World.rand.nextInt( 20 ) != 0 )
+    		{
+    			// only a small chance villagers will spawn in fully abandoned villages
+    			
+    			return;
+    		}    		
+    		else if ( par6 > 1 )
+    		{
+    			// and in small dirty numbers if they do
+    			
+    			par6 = 1;    			
+    		}
+    		
+			bDirtyPeasants = true;
+    	}
+    	// END FCMOD
+    	
         if (this.villagersSpawned < par6)
         {
             for (int var7 = this.villagersSpawned; var7 < par6; ++var7)
@@ -121,7 +145,45 @@
                 }
 
                 ++this.villagersSpawned;
+                
+            	// FCMOD: Change
+                /*
                 EntityVillager var11 = new EntityVillager(par1World, this.getVillagerType(var7));
+                */
+                EntityLiving var11;
+                
+                int iVillagerType = this.getVillagerType(var7);
+                
+                if ( iVillagerType != 0 && iAbandonmentLevel > 0 )
+                {
+                	if ( iAbandonmentLevel > 1 || ( iVillagerType != 3 && iVillagerType != 4 ) ) // fully abandoned or not blacksmith or butcher
+                	{
+                		continue;
+                	}
+                }
+                
+                if ( par1World.getWorldInfo().getGameType() != EnumGameType.CREATIVE )
+                {
+                	FCEntityZombie zombieVillager = new FCEntityZombie( par1World );
+                	
+                	zombieVillager.m_iVillagerClass = iVillagerType;
+                	zombieVillager.SetPersistent( true );                	
+            		zombieVillager.setVillager( true );
+                	
+                	var11 = zombieVillager;
+                }
+                else
+                {
+                	FCEntityVillager villager = new FCEntityVillager( par1World, iVillagerType ); 
+                    
+                    if ( bDirtyPeasants )
+                    {
+                    	villager.SetDirtyPeasant( 1 );
+                    }
+                    
+                    var11 = villager;
+                }
+                // END FCMOD
                 var11.setLocationAndAngles((double)var8 + 0.5D, (double)var9, (double)var10 + 0.5D, 0.0F, 0.0F);
                 par1World.spawnEntityInWorld(var11);
             }
