--- a/minecraft_server/net/minecraft/src/ContainerBrewingStand.java
+++ b/minecraft_server/net/minecraft/src/ContainerBrewingStand.java
@@ -105,7 +105,10 @@
                         return null;
                     }
                 }
-                else if (!this.mergeItemStack(var5, 4, 40, false))
+                // FCMOD: Change to give priority to player hotbar
+                //else if (!this.mergeItemStack(var5, 4, 40, false))
+                else if (!this.mergeItemStack(var5, 4, 40, true))
+            	// END FCMOD
                 {
                     return null;
                 }
