--- a/minecraft_server/net/minecraft/src/DispenserBehaviorFireball.java
+++ b/minecraft_server/net/minecraft/src/DispenserBehaviorFireball.java
@@ -19,7 +19,7 @@
         double var13 = var12.nextGaussian() * 0.05D + (double)var3.getFrontOffsetX();
         double var15 = var12.nextGaussian() * 0.05D + (double)var3.getFrontOffsetY();
         double var17 = var12.nextGaussian() * 0.05D + (double)var3.getFrontOffsetZ();
-        var11.spawnEntityInWorld(new EntitySmallFireball(var11, var5, var7, var9, var13, var15, var17));
+        var11.spawnEntityInWorld(EntityList.createEntityOfType(EntitySmallFireball.class, var11, var5, var7, var9, var13, var15, var17));
         par2ItemStack.splitStack(1);
         return par2ItemStack;
     }
