--- a/minecraft_server/net/minecraft/src/DispenserBehaviorSnowball.java
+++ b/minecraft_server/net/minecraft/src/DispenserBehaviorSnowball.java
@@ -7,6 +7,6 @@
      */
     protected IProjectile getProjectileEntity(World par1World, IPosition par2IPosition)
     {
-        return new EntitySnowball(par1World, par2IPosition.getX(), par2IPosition.getY(), par2IPosition.getZ());
+    	return (IProjectile) EntityList.createEntityOfType(EntitySnowball.class, par1World, par2IPosition.getX(), par2IPosition.getY(), par2IPosition.getZ());
     }
 }
