--- a/minecraft_server/net/minecraft/src/EntityFireworkRocket.java
+++ b/minecraft_server/net/minecraft/src/EntityFireworkRocket.java
@@ -21,6 +21,10 @@
         this.dataWatcher.addObjectByDataType(8, 5);
     }
 
+    public EntityFireworkRocket(World par1World, double par2, double par4, double par6) {
+    	this(par1World, par2, par4, par6, null);
+    }
+
     public EntityFireworkRocket(World par1World, double par2, double par4, double par6, ItemStack par8ItemStack)
     {
         super(par1World);
