--- a/minecraft_server/net/minecraft/src/EntityCow.java
+++ b/minecraft_server/net/minecraft/src/EntityCow.java
@@ -1,5 +1,12 @@
+// FCMOD: Class changes deprecated 07/23/2018
+
 package net.minecraft.src;
 
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.util.Iterator;
+import java.util.List;
+
 public class EntityCow extends EntityAnimal
 {
     public EntityCow(World par1World)
@@ -138,7 +145,7 @@
      */
     public EntityCow spawnBabyAnimal(EntityAgeable par1EntityAgeable)
     {
-        return new EntityCow(this.worldObj);
+    	return (EntityCow) EntityList.createEntityOfType(EntityCow.class, this.worldObj);
     }
 
     public EntityAgeable createChild(EntityAgeable par1EntityAgeable)
