--- a/minecraft_server/net/minecraft/src/EntityAITasks.java
+++ b/minecraft_server/net/minecraft/src/EntityAITasks.java
@@ -178,4 +178,48 @@
     {
         return (par1EntityAITaskEntry.action.getMutexBits() & par2EntityAITaskEntry.action.getMutexBits()) == 0;
     }
+    
+    // FCMOD: Added New
+    public void RemoveAllTasksOfClass( Class classToRemove )
+    {
+        Iterator<EntityAITaskEntry> entries = taskEntries.iterator();
+
+        while ( entries.hasNext())
+        {
+            EntityAITaskEntry tempEntry = entries.next();
+            EntityAIBase tempAction = tempEntry.action;
+
+            if ( classToRemove.isAssignableFrom( tempAction.getClass() ) )
+            {
+                if ( executingTaskEntries.contains( tempEntry ) )
+                {
+                    tempAction.resetTask();
+                    
+                    executingTaskEntries.remove( tempEntry );
 }
+
+                entries.remove();
+            }
+        }
+    }
+    
+    public void RemoveAllTasks()
+    {
+        Iterator<EntityAITaskEntry> entries = taskEntries.iterator();
+
+        while ( entries.hasNext())
+        {
+            EntityAITaskEntry tempEntry = entries.next();
+
+            if ( executingTaskEntries.contains( tempEntry ) )
+            {
+                tempEntry.action.resetTask();
+                
+                executingTaskEntries.remove( tempEntry );
+            }
+
+            entries.remove();
+        }
+    }
+    // END FCMOD
+}
