--- a/minecraft_server/net/minecraft/src/EntityPig.java
+++ b/minecraft_server/net/minecraft/src/EntityPig.java
@@ -1,5 +1,10 @@
+// FCMOD: Class changes deprecated 07/23/2018
+
 package net.minecraft.src;
 
+import java.util.Iterator;
+import java.util.List;
+
 public class EntityPig extends EntityAnimal
 {
     /** AI task for player control. */
@@ -191,7 +196,7 @@
     {
         if (!this.worldObj.isRemote)
         {
-            EntityPigZombie var2 = new EntityPigZombie(this.worldObj);
+        	EntityPigZombie var2 = (EntityPigZombie) EntityList.createEntityOfType(EntityPigZombie.class, this.worldObj);
             var2.setLocationAndAngles(this.posX, this.posY, this.posZ, this.rotationYaw, this.rotationPitch);
             this.worldObj.spawnEntityInWorld(var2);
             this.setDead();
@@ -216,7 +221,7 @@
      */
     public EntityPig spawnBabyAnimal(EntityAgeable par1EntityAgeable)
     {
-        return new EntityPig(this.worldObj);
+    	return (EntityPig) EntityList.createEntityOfType(EntityPig.class, this.worldObj);
     }
 
     /**
