--- a/minecraft_server/net/minecraft/src/EntityMooshroom.java
+++ b/minecraft_server/net/minecraft/src/EntityMooshroom.java
@@ -1,6 +1,9 @@
 package net.minecraft.src;
 
-public class EntityMooshroom extends EntityCow
+// FCMOD: Changed
+//public class EntityMooshroom extends EntityCow
+public class EntityMooshroom extends FCEntityCow
+// END FCMOD
 {
     public EntityMooshroom(World par1World)
     {
@@ -12,6 +15,8 @@
     /**
      * Called when a player interacts with a mob. e.g. gets milk from a cow, gets into the saddle on a pig.
      */
+    // FCMOD: Removed and replaced
+    /*
     public boolean interact(EntityPlayer par1EntityPlayer)
     {
         ItemStack var2 = par1EntityPlayer.inventory.getCurrentItem();
@@ -57,6 +62,8 @@
             return super.interact(par1EntityPlayer);
         }
     }
+    */
+    // END FCMOD    
 
     public EntityMooshroom func_94900_c(EntityAgeable par1EntityAgeable)
     {
@@ -66,7 +73,10 @@
     /**
      * This function is used when two same-species animals in 'love mode' breed to generate the new baby animal.
      */
-    public EntityCow spawnBabyAnimal(EntityAgeable par1EntityAgeable)
+    // FCMOD: Changed
+    //public EntityCow spawnBabyAnimal(EntityAgeable par1EntityAgeable)
+    public FCEntityCow spawnBabyAnimal(EntityAgeable par1EntityAgeable)
+    // END FCMOD
     {
         return this.func_94900_c(par1EntityAgeable);
     }
@@ -75,4 +85,78 @@
     {
         return this.func_94900_c(par1EntityAgeable);
     }
+
+    // FCMOD: Added New
+    @Override
+    public void CheckForGrazeSideEffects( int i, int j, int k )
+    {
+    	// override to get rid of mooshroom conversion on eat mycellium
 }
+    
+    @Override
+    public void ConvertToMooshroom()
+    {
+    }
+    
+    @Override
+    public void onLivingUpdate()
+    {
+        if ( !worldObj.isRemote )
+        {
+            CheckForMyceliumSpread();
+        }
+
+        super.onLivingUpdate();
+    }
+    
+    @Override
+    public boolean interact( EntityPlayer player )
+    {
+        ItemStack heldStack = player.inventory.getCurrentItem();
+
+        if ( heldStack != null && heldStack.itemID == Item.bowlEmpty.itemID && GotMilk() )
+        {
+            attackEntityFrom( DamageSource.generic, 0 );
+            
+            if ( !worldObj.isRemote )
+        	{
+            	SetGotMilk( false );
+            	
+		        worldObj.playAuxSFX( FCBetterThanWolves.m_iCowMilkedAuxFXID,
+	                MathHelper.floor_double( posX ), (int)posY,
+	                MathHelper.floor_double( posZ ), 0 );
+            }
+            
+            if ( heldStack.stackSize == 1 )
+            {
+                player.inventory.setInventorySlotContents( player.inventory.currentItem, 
+                	new ItemStack( Item.bowlSoup ) );
+            }
+            else if ( player.inventory.addItemStackToInventory( new ItemStack( Item.bowlSoup ) ) ) 
+            {
+                player.inventory.decrStackSize(player.inventory.currentItem, 1);
+            }
+            
+            return true;
+        }
+        
+        // skip over parent to avoid vanilla milking
+        
+        return EntityAnimalInteract( player ); 
+    }
+    
+	//------------- Class Specific Methods ------------//
+    
+    private void CheckForMyceliumSpread()
+    {
+    	if ( worldObj.provider.dimensionId != 1 && rand.nextInt( 1000 ) == 0 )
+    	{
+    		FCBlockMycelium.CheckForMyceliumSpreadToRandomBlockAround( worldObj,
+    			MathHelper.floor_double( posX ), (int)posY - 1, 
+    			MathHelper.floor_double( posZ ) );
+    	}
+    }
+    
+	//----------- Client Side Functionality -----------//
+    // END FCMOD    
+}
