--- a/minecraft_server/net/minecraft/src/EntityAIOwnerHurtByTarget.java
+++ b/minecraft_server/net/minecraft/src/EntityAIOwnerHurtByTarget.java
@@ -17,6 +17,12 @@
      */
     public boolean shouldExecute()
     {
+    	// FCMOD: Code added to check if sitting before attacking    	
+    	if ( theDefendingTameable.isSitting() )
+    	{
+    		return false;
+    	}
+    	// END FCMOD
         if (!this.theDefendingTameable.isTamed())
         {
             return false;
@@ -45,4 +51,17 @@
         this.taskOwner.setAttackTarget(this.theOwnerAttacker);
         super.startExecuting();
     }
+
+    // FCMOD: Inherited function added
+    @Override
+    public boolean continueExecuting()
+    {
+    	if ( theDefendingTameable.isSitting() )
+    	{
+    		return false;
 }
+    	
+    	return super.continueExecuting();
+    }
+    // END FCMOD
+}
