--- a/minecraft_server/net/minecraft/src/EntityMagmaCube.java
+++ b/minecraft_server/net/minecraft/src/EntityMagmaCube.java
@@ -1,6 +1,9 @@
 package net.minecraft.src;
 
-public class EntityMagmaCube extends EntitySlime
+// FCMOD: Changed
+//public class EntityMagmaCube extends EntitySlime
+public class EntityMagmaCube extends FCEntitySlime
+// END FCMOD
 {
     public EntityMagmaCube(World par1World)
     {
