--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCAddOnHandler.java
@@ -0,0 +1,413 @@
+// FCMOD
+
+package net.minecraft.src;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.channels.ReadPendingException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.logging.FileHandler;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.logging.SimpleFormatter;
+
+public class FCAddOnHandler
+{
+	public static List<FCAddOn> m_ModList = new LinkedList();
+	public static boolean m_bModsInitialized = false;
+
+	public static final Logger m_Logger = Logger.getLogger("BetterThanWolves");
+
+	// Client version
+	//private static final File m_LogFile = new File(Minecraft.getMinecraftDir(), "BTWLog.txt");    
+	// Server version
+	private static final File m_LogFile = new File(new File("."), "BTWLog.txt");
+
+	private static FileHandler m_LogFileHandler = null;
+
+	private static NetServerHandler netServerHandler;
+	private static ArrayList<String> ackCheckFails = new ArrayList<String>();
+	
+	public static Map<Class<? extends FCAddOn>, FCAddOnUtilsWorldData> addonWorldDataMap;
+	
+	private static ArrayList<FCAddOn> loadOrder = new ArrayList();
+
+	public static void AddMod(FCAddOn mod)
+	{
+		m_ModList.add(mod);
+	}
+	
+	public static URL getResource(URL url) {
+		try {
+			if (url.getPath().endsWith(".jar") || url.getPath().endsWith("zip")) {
+				return new URL("jar:" + url + "!/");
+			}
+		} catch (MalformedURLException e) {
+			e.printStackTrace();
+		}
+		return url;
+	}
+	
+	public static ClassLoader getModLoader() {
+		final Class domain = FCAddOnHandler.class;
+		final ClassLoader classLoader = domain.getClassLoader();
+		try {
+			final File[] files = new File(m_LogFile.getParentFile().getCanonicalPath() + "/mods").listFiles();
+			final List<URL> urls = new ArrayList<URL>();
+			final URL baseURL = getResource(domain.getProtectionDomain().getCodeSource().getLocation());
+			urls.add(baseURL);
+			for (int i = 1; i <= files.length; i++) {
+				File file = files[i - 1];
+				if (file.isDirectory())
+					urls.add(file.toURI().toURL());
+				else if (file.getParent().endsWith(".jar"))
+					urls.add(new URL("jar:" + file.toURI().toURL() + "!/"));
+			}
+
+			return new URLClassLoader(urls.toArray(new URL[urls.size()]));
+		} catch (Exception e) {
+			e.printStackTrace();
+		}
+		return classLoader;
+	}
+
+	public static void loadAddOns() {
+		final ClassLoader classLoader = getModLoader();
+		try {
+			final String[] list = FCAddOnFinder.getClasses(classLoader);
+			for (final String className : list) {
+				final Class<?> classDescriptor = 
+						classLoader.loadClass(className);
+				if (FCAddOn.class.isAssignableFrom(classDescriptor)) {
+					Class<FCAddOn> addon = (Class<FCAddOn>) Class.forName(className, true, classLoader);
+					LogMessage("Found Add-on: " + addon.getSimpleName());
+				}
+			}
+		} catch (final Exception e) {
+			e.printStackTrace();
+		}
+	}
+
+	public static void InitializeMods()
+	{
+		if (!m_bModsInitialized)
+		{
+			InitializeLogger();
+
+			LogMessage("...Add-On Handler Initializing...");
+			
+			loadAddOns();
+			
+			//buildLoadOrder();
+			
+			addonWorldDataMap = initWorldDataForAddon();
+
+			PreInitializeMods();
+			
+			loadModConfigs();
+
+			Iterator modIterator = m_ModList.iterator();
+
+			while (modIterator.hasNext())
+			{
+				FCAddOn mod = (FCAddOn)modIterator.next();
+
+				mod.Initialize();
+			}
+
+			PostInitializeMods();
+
+			m_bModsInitialized = true;
+
+			OnLanguageLoaded(StringTranslate.getInstance());
+
+			LogMessage("...Add-On Handler Initialization Complete...");
+		}
+	}
+	/*
+	private static void buildLoadOrder() {
+		buildDependencies();
+	}
+	
+	private static void addModToLoadOrder(FCAddOn mod, Set<FCAddOn> modsInProgress) {
+		if (modsInProgress.contains(mod)) {
+			throw new ExceptionInInitializerError("Dependency cycle detected when loading addons");
+		}
+		
+		modsInProgress.add(mod);
+		
+		for (FCAddOn dependency : mod.dependents) {
+			if (!dependency.hasBeenInitialized) {
+				addModToLoadOrder(dependency, modsInProgress);
+			}
+		}
+		
+		loadOrder.add(mod);
+		
+		modsInProgress.remove(mod);
+	}
+	
+	private static void buildDependencies() {
+		for (FCAddOn mod : m_ModList) {
+			for (FCAddOn modForCheck : m_ModList) {
+				DependencyType dependencyType = mod.getDependencyForAddon(modForCheck);
+				
+				switch (dependencyType) {
+				case INCOMPATIBLE:
+					throw new ExceptionInInitializerError("Cannot load incompatible mods: " + mod.getName() + ", " + modForCheck.getName());
+				case LOAD_BEFORE:
+					mod.dependents.add(modForCheck);
+					break;
+				case LOAD_AFTER:
+					modForCheck.dependents.add(mod);
+					break;
+				default:
+					break;
+				}
+			}
+		}
+	}
+	*/
+
+	public static void InitializeLogger()
+	{
+		try
+		{
+			if ((m_LogFile.exists() || m_LogFile.createNewFile()) && m_LogFile.canWrite())
+			{
+				m_LogFileHandler = new FileHandler(m_LogFile.getPath());
+				m_LogFileHandler.setFormatter(new SimpleFormatter());
+				m_Logger.addHandler(m_LogFileHandler);
+
+				m_Logger.setLevel(Level.FINE);
+			}
+		}
+		catch (Throwable error)
+		{
+			throw new RuntimeException(error);
+		}
+	}
+
+	public static void LogMessage(String string)
+	{
+		System.out.println(string);
+
+		if (net.minecraft.server.MinecraftServer.getServer() != null)
+		{
+			// client
+			//net.minecraft.server.MinecraftServer.getServer().getLogAgent().logInfo(string);
+			// server
+			net.minecraft.server.MinecraftServer.getServer().getLogAgent().func_98233_a(string);
+		}
+
+		m_Logger.fine(string);
+	}
+
+	public static void LogWarning(String string)
+	{
+		System.out.println(string);
+
+		if (net.minecraft.server.MinecraftServer.getServer() != null)
+		{
+			// client
+			//net.minecraft.server.MinecraftServer.getServer().getLogAgent().logWarning(string);
+			// server
+			net.minecraft.server.MinecraftServer.getServer().getLogAgent().func_98236_b(string);
+		}
+
+		m_Logger.fine(string);
+	}
+
+	public static void PreInitializeMods()
+	{
+		Iterator modIterator = m_ModList.iterator();
+
+		while (modIterator.hasNext())
+		{
+			FCAddOn mod = (FCAddOn)modIterator.next();
+
+			mod.PreInitialize();
+		}
+	}
+
+	public static void PostInitializeMods()
+	{
+		Iterator modIterator = m_ModList.iterator();
+
+		while (modIterator.hasNext())
+		{
+			FCAddOn mod = (FCAddOn)modIterator.next();
+
+			mod.PostInitialize();
+		}
+	}
+
+	public static void OnLanguageLoaded(StringTranslate translator) {
+		// only call on language loaded after mods have been initialized to prevent funkiness due to static instance variable of 
+		// StringTranslate creating ambiguous initialization order.
+
+		if (m_bModsInitialized) {
+			Iterator modIterator = m_ModList.iterator();
+
+			while (modIterator.hasNext()) {
+				FCAddOn mod = (FCAddOn)modIterator.next();
+
+				mod.OnLanguageLoaded(translator);
+
+				String prefix = mod.GetLanguageFilePrefix();
+
+				if (prefix != null) {
+					LogMessage("...Add-On Handler Loading Custom Language File With Prefix: " + prefix + "..." );
+
+					translator.LoadAddonLanguageExtension(prefix);
+				}
+			}	    	
+		}		
+	}
+
+	public static void ServerCustomPacketReceived(NetServerHandler handler, Packet250CustomPayload packet)
+	{
+		Iterator modIterator = m_ModList.iterator();
+
+		while (modIterator.hasNext())
+		{
+			FCAddOn mod = (FCAddOn)modIterator.next();
+
+			if (mod.ServerCustomPacketReceived(handler, packet))
+			{
+				return;
+			}    		
+		}    	
+	}
+
+
+	public static void serverPlayerConnectionInitialized(NetServerHandler serverHandler, EntityPlayerMP playerMP) {
+		netServerHandler = serverHandler;
+
+		for (Object mod : FCAddOnHandler.m_ModList) {
+			((FCAddOn) mod).serverPlayerConnectionInitialized(serverHandler, playerMP);
+			
+			if (((FCAddOn) mod).shouldVersionCheck)
+				((FCAddOn) mod).sendVersionCheckToClient(serverHandler, playerMP);
+		}
+	}
+
+	public static boolean getAwaitingLoginAck() {
+		for (Object mod : FCAddOnHandler.m_ModList) {
+			if (((FCAddOn) mod).getAwaitingLoginAck()) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	public static void incrementTicksSinceAckRequested() {
+		for (Object mod : FCAddOnHandler.m_ModList) {
+			((FCAddOn) mod).incrementTicksSinceAckRequested();
+		}
+	}
+
+	public static void handleAckCheck() {
+		for (Object mod : FCAddOnHandler.m_ModList) {
+			if (!((FCAddOn) mod).handleAckCheck()) {
+				ackCheckFails.add(((FCAddOn) mod).getName());
+			}
+		}
+
+		if (!ackCheckFails.isEmpty()) {
+			String message = "WARNING: Client missing the following addons, or very high latency connection: ";
+
+			for (int i = 0; i < ackCheckFails.size(); i++) {
+				if (i > 0)
+					message += ", ";
+				message += ackCheckFails.get(i);
+			}
+
+			FCUtilsWorld.SendPacketToPlayer(netServerHandler, new Packet3Chat(message));
+		}
+	}
+	
+	public static Map<Class<? extends FCAddOn>, FCAddOnUtilsWorldData> initWorldDataForAddon() {
+		Map<Class<? extends FCAddOn>, FCAddOnUtilsWorldData> worldDataMap = new HashMap();
+		
+		for (Object mod : FCAddOnHandler.m_ModList) {
+			FCAddOnUtilsWorldData worldData = ((FCAddOn) mod).createWorldData();
+			
+			if (worldData != null) {
+				worldDataMap.put(((FCAddOn) mod).getClass(), worldData);
+			}
+		}
+		
+		return worldDataMap;
+	}
+	
+	public static void loadModConfigs() {
+		for (Object mod : FCAddOnHandler.m_ModList) {
+			FCAddOn addon = (FCAddOn) mod;
+			
+			Map<String, String> addonConfigProperties = addon.loadConfigProperties();
+			
+			if (addonConfigProperties != null) {
+				addon.repopulateConfigFile(addonConfigProperties);
+				addon.handleConfigProperties(addonConfigProperties);
+			}
+		}
+	}
+	
+	public static FCAddOnUtilsWorldData getWorldDataForMod(Class<? extends FCAddOn> mod) {
+		if (FCAddOnHandler.addonWorldDataMap.containsKey(mod))
+			return FCAddOnHandler.addonWorldDataMap.get(mod);
+		else
+			return null;
+	}
+
+	public static void setWorldDataForMod(Class<? extends FCAddOn> mod, FCAddOnUtilsWorldData data) {
+		FCAddOnHandler.addonWorldDataMap.put(mod, data);
+	}
+	
+	public static void decorateWorld(FCIBiomeDecorator decorator, World world, Random rand, int x, int y, BiomeGenBase biome) {
+		for (Object mod : FCAddOnHandler.m_ModList) {
+			FCAddOn addon = (FCAddOn) mod;
+
+			addon.decorateWorld(decorator, world, rand, x, y, biome);
+		}
+	}
+	
+	public static boolean isModInstalled(String name) {
+		for (FCAddOn mod : m_ModList) {
+			if (mod.getName().equals(name)) {
+				return true;
+			}
+		}
+		
+		return false;
+	}
+	
+	public static FCAddOn getModByName(String name) {
+		for (FCAddOn mod : m_ModList) {
+			if (mod.getName().equals(name)) {
+				return mod;
+			}
+		}
+		
+		return null;
+	}
+	
+	public static enum DependencyType {
+		LOAD_BEFORE,
+		LOAD_AFTER,
+		INCOMPATIBLE,
+		NO_DEPENDENCY
+	}
+}
\ No newline at end of file
