--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockStoneBrickLoose.java
@@ -0,0 +1,44 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCBlockStoneBrickLoose extends FCBlockLavaReceiver
+{
+    public FCBlockStoneBrickLoose( int iBlockID )
+    {
+        super( iBlockID, Material.rock );
+        
+        setHardness( 1F ); // setHardness( 2.25F ); regular stone brick
+        setResistance( 5F ); // setResistance( 10F ); regular stone brick
+        
+        SetPicksEffectiveOn();
+        
+        setStepSound( soundStoneFootstep );
+        
+        setUnlocalizedName( "fcBlockStoneBrickLoose" );        
+        
+		setCreativeTab( CreativeTabs.tabBlock );
+    }
+    
+	@Override
+	public int damageDropped(int metadata) {
+		return getStrata(metadata) << 2; // this block stores strata in last 2 bits
+	}
+
+    @Override
+    public boolean OnMortarApplied( World world, int i, int j, int k ) {
+		world.setBlockAndMetadataWithNotify( i, j, k, Block.stoneBrick.blockID,  getStrata( world, i, j, k) << 2);
+		
+		return true;
+    }
+    
+    @Override
+    public int getDamageValue(World world, int x, int y, int z) {
+		// used only by pick block
+		return world.getBlockMetadata(x, y, z);
+    }
+    
+    //------------- Class Specific Methods ------------//
+
+	//------------ Client Side Functionality ----------//    
+}
\ No newline at end of file
