--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockCarrotFlowers.java
@@ -0,0 +1,18 @@
+package net.minecraft.src;
+
+public class FCBlockCarrotFlowers extends FCBlockCarrotBase {
+	public FCBlockCarrotFlowers(int iBlockID) {
+		super(iBlockID);
+		this.setUnlocalizedName("fcBlockCarrotFlowers");
+	}
+
+	@Override
+	protected int GetCropItemID() {
+		return FCBetterThanWolves.fcItemCarrotSeeds.itemID;
+	}
+
+	@Override
+	protected int GetSeedItemID() {
+		return FCBetterThanWolves.fcItemCarrotSeeds.itemID;
+	}
+}
\ No newline at end of file
