--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockButtonWood.java
@@ -0,0 +1,18 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCBlockButtonWood extends FCBlockButton
+{
+    protected FCBlockButtonWood( int iBlockID )
+    {
+        super( iBlockID, true );
+        
+        SetAxesEffectiveOn( true );        
+        SetBuoyant();
+    }
+
+    //------------- Class Specific Methods ------------//
+    
+	//----------- Client Side Functionality -----------//
+}
