--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockStoneBrickLooseSlab.java
@@ -0,0 +1,62 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCBlockStoneBrickLooseSlab extends FCBlockMortarReceiverSlab
+{
+    public FCBlockStoneBrickLooseSlab( int iBlockID )
+    {
+        super( iBlockID, Material.rock );
+        
+        setHardness( 1F ); // setHardness( 2.25F ); regular stone brick
+        setResistance( 5F ); // setResistance( 10F ); regular stone brick
+        
+        SetPicksEffectiveOn();
+        
+        setStepSound( soundStoneFootstep );
+        
+        setUnlocalizedName( "fcBlockStoneBrickLooseSlab" );
+        
+		setCreativeTab( CreativeTabs.tabBlock );
+    }
+    
+	@Override
+	public int GetCombinedBlockID( int iMetadata )
+	{
+		return FCBetterThanWolves.fcBlockStoneBrickLoose.blockID;
+	}
+
+	@Override
+	public int GetCombinedMetadata( int iMetadata )
+	{
+		return getStrata(iMetadata);
+	}
+	
+	@Override
+	public boolean OnMortarApplied(World world, int i, int j, int k) {
+		int iNewMetadata = getStrata(world, i, j, k);
+
+		if (GetIsUpsideDown(world, i, j, k)) {
+			iNewMetadata |= 8;
+		}
+
+		world.setBlockAndMetadataWithNotify(i, j, k, FCBetterThanWolves.fcBlockStoneBrickSlabSingle.blockID, iNewMetadata);
+
+		return true;
+	}
+
+    
+    //------------- Class Specific Methods ------------//    
+    public int getStrata( IBlockAccess blockAccess, int i, int j, int k )
+    {
+		return getStrata( blockAccess.getBlockMetadata( i, j, k ) );
+    }
+
+    public int getStrata( int iMetadata )
+    {
+    	return (iMetadata & 12) >>> 2;
+    }
+
+
+	//----------- Client Side Functionality -----------//
+}
\ No newline at end of file
