--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockReedRoots.java
@@ -0,0 +1,26 @@
+package net.minecraft.src;
+
+import java.util.Random;
+
+public class FCBlockReedRoots extends FCBlockReedBase {
+	public FCBlockReedRoots(int id) {
+		super(id);
+		this.setUnlocalizedName("fcBlockReedRoots");
+	}
+
+	@Override
+	public int idDropped(int par1, Random rand, int par3) {
+        return FCBetterThanWolves.fcItemReedRoots.itemID;
+    }
+    
+	@Override
+    public boolean canPlaceBlockAt(World world, int x, int y, int z) {
+        int blockBelowID = world.getBlockId(x, y - 1, z);
+        Block blockBelow = Block.blocksList[blockBelowID];
+
+    	return blockBelow != null && 
+    		!(blockBelow instanceof FCBlockReedBase) &&
+    		!(blockBelow instanceof FCBlockReedLegacy) &&
+    		super.canPlaceBlockAt(world, x, y, z);
+    }
+}
\ No newline at end of file
