--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockBucketMilk.java
@@ -0,0 +1,38 @@
+// FCMOD
+
+package net.minecraft.src;
+
+import java.util.Random;
+
+public class FCBlockBucketMilk extends FCBlockBucketFull
+{
+    public FCBlockBucketMilk( int iBlockID )
+    {
+        super( iBlockID );
+    	
+    	setUnlocalizedName( "fcBlockBucketMilk" );
+    }
+    
+	@Override
+    public int idDropped( int iMetadata, Random rand, int iFortuneMod )
+    {
+		return Item.bucketMilk.itemID;
+    }
+	
+	//------------- Class Specific Methods ------------//
+	
+	@Override
+    public boolean AttemptToSpillIntoBlock( World world, int i, int j, int k )
+    {
+        if ( ( world.isAirBlock( i, j, k ) || !world.getBlockMaterial( i, j, k ).isSolid() ) )
+        {     
+    		world.setBlockWithNotify( i, j, k, FCBetterThanWolves.fcBlockMilk.blockID );
+            
+            return true;
+        }
+        
+        return false;
+    }
+	
+	//----------- Client Side Functionality -----------//
+}
