--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockFullBlock.java
@@ -0,0 +1,13 @@
+// FCMOD
+// special cased full block with standard attributes (stuff like stone and dirt) to speed rendering
+
+package net.minecraft.src;
+
+public class FCBlockFullBlock extends Block
+{
+    protected FCBlockFullBlock( int iBlockID, Material material )
+    {
+    	super( iBlockID, material );
+    }
+    
+}
