--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockGrassSlab.java
@@ -0,0 +1,149 @@
+package net.minecraft.src;
+
+import java.util.List;
+import java.util.Random;
+
+public class FCBlockGrassSlab extends FCBlockSlabAttached {
+	protected FCBlockGrassSlab(int blockID) {
+		super(blockID, Material.ground);
+
+		this.setHardness(0.5F);
+		this.SetShovelsEffectiveOn(true);
+
+		this.setStepSound(soundGrassFootstep);
+
+		this.setUnlocalizedName("fcBlockSlabDirt");
+
+		this.setTickRandomly(true);
+
+		this.setCreativeTab(CreativeTabs.tabBlock);
+
+		this.setUnlocalizedName("fcBlockGrassSlab");
+	}
+
+	@Override
+	public void updateTick(World world, int x, int y, int z, Random rand) {
+		if (!FCBlockGrass.canGrassSurviveAtLocation(world, x, y, z)) {
+			this.revertToDirt(world, x, y, z);
+		}
+		else if (FCBlockGrass.canGrassSpreadFromLocation(world, x, y, z)) {
+			if (rand.nextFloat() <= FCBlockGrass.GROWTH_CHANCE) {
+				FCBlockGrass.checkForGrassSpreadFromLocation(world, x, y, z);
+			}
+
+			if (isSparse(world, x, y, z) && rand.nextInt(FCBlockGrass.SELF_GROWTH_CHANCE) == 0) {
+				this.setFullyGrown(world, x, y, z);
+			}
+		}
+	}
+    
+    @Override
+    public int idDropped(int metadata, Random random, int fortuneModifier) {
+        return FCBetterThanWolves.fcBlockDirtLooseSlab.blockID;
+    }
+    
+	@Override
+	public boolean DropComponentItemsOnBadBreak(World world, int x, int y, int z, int iMetadata, float fChanceOfDrop) {
+		DropItemsIndividualy(world, x, y, z, FCBetterThanWolves.fcItemPileDirt.itemID, 3, 0, fChanceOfDrop);
+		return true;
+	}
+
+	@Override
+	protected void OnAnchorBlockLost(World world, int i, int j, int k) {
+		world.setBlock(i, j, k, FCBetterThanWolves.fcBlockDirtLooseSlab.blockID, world.getBlockMetadata(i, j, k) & 3, 2);
+	}
+
+	@Override
+	public int GetCombinedBlockID(int iMetadata) {
+		return Block.grass.blockID;
+	}
+	
+	@Override
+    public boolean AttemptToCombineWithFallingEntity(World world, int x, int y, int z, EntityFallingSand entity) {
+		if (entity.blockID == FCBetterThanWolves.fcBlockDirtLooseSlab.blockID) {
+			if (!GetIsUpsideDown(world, x, y, z)) {
+				world.setBlockWithNotify(x, y, z, FCBetterThanWolves.fcBlockDirtLoose.blockID);	
+						
+				return true;
+			}
+		}
+		
+		return super.AttemptToCombineWithFallingEntity(world, x, y, z, entity);
+	}
+
+	@Override
+	public boolean CanBePistonShoveled(World world, int x, int y, int z) {
+		return true;
+	}
+
+	@Override
+	public boolean CanBeGrazedOn(IBlockAccess blockAccess, int x, int y, int z, EntityAnimal animal) {
+		return !isSparse(blockAccess, x, y, z) || animal.IsStarving() || animal.GetDisruptsEarthOnGraze();
+	}
+
+	@Override
+	public void OnGrazed(World world, int x, int y, int z, EntityAnimal animal) {
+		if (!animal.GetDisruptsEarthOnGraze()) {
+			if (isSparse(world, x, y, z)) {
+				world.setBlockWithNotify(x, y, z, FCBetterThanWolves.fcBlockDirtSlab.blockID);
+			}
+			else {
+				setSparse(world, x, y, z);
+			}
+		}
+		else {
+			world.setBlockWithNotify(x, y, z, FCBetterThanWolves.fcBlockDirtLooseSlab.blockID);
+			NotifyNeighborsBlockDisrupted(world, x, y, z);
+		}
+	}
+
+	@Override
+	public void OnVegetationAboveGrazed(World world, int x, int y, int z, EntityAnimal animal) {
+		if (animal.GetDisruptsEarthOnGraze()) {
+			world.setBlockWithNotify(x, y, z, FCBetterThanWolves.fcBlockDirtLoose.blockID);
+			NotifyNeighborsBlockDisrupted(world, x, y, z);
+		}
+	}
+    
+    @Override
+    public void OnBlockDestroyedWithImproperTool(World world, EntityPlayer player, int i, int j, int k, int iMetadata) {
+    	super.OnBlockDestroyedWithImproperTool(world, player, i, j, k, iMetadata);
+    	
+		OnDirtSlabDugWithImproperTool(world, i, j, k, GetIsUpsideDown(iMetadata));
+    }
+    
+	@Override
+    public void onBlockDestroyedByExplosion(World world, int i, int j, int k, Explosion explosion) {
+		super.onBlockDestroyedByExplosion(world, i, j, k, explosion);
+    	
+		OnDirtSlabDugWithImproperTool(world, i, j, k, GetIsUpsideDown(world, i, j, k));
+    }
+
+	//------ Class specific methods ------//
+
+	public void revertToDirt(World world, int x, int y, int z) {
+		boolean isUpsideDown = this.GetIsUpsideDown(world, x, y, z);
+
+		world.setBlockWithNotify(x, y, z, FCBetterThanWolves.fcBlockDirtSlab.blockID);
+		FCBetterThanWolves.fcBlockDirtSlab.SetSubtype(world, x, y, z, FCBlockDirtSlab.m_iSubtypeDirt);
+		FCBetterThanWolves.fcBlockDirtSlab.SetIsUpsideDown(world, x, y, z, isUpsideDown);
+	}
+
+	public boolean isSparse(IBlockAccess blockAccess, int x, int y, int z) {
+		return isSparse(blockAccess.getBlockMetadata(x, y, z));
+	}
+
+	public boolean isSparse(int metadata) {
+		return (metadata & -1) == 2;
+	}
+
+	public void setSparse(World world, int x, int y, int z) {
+		int metadata = world.getBlockMetadata(x, y, z);
+		world.setBlockMetadataWithNotify(x, y, z, metadata | 2);
+	}
+
+	public void setFullyGrown(World world, int x, int y, int z) {
+		int metadata = world.getBlockMetadata(x, y, z);
+		world.setBlockMetadataWithNotify(x, y, z, metadata & 1);
+	}
+}
\ No newline at end of file
