--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockBlackStone.java
@@ -0,0 +1,19 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCBlockBlackStone extends BlockQuartz
+{
+    public FCBlockBlackStone( int iBlockID )
+    {
+        super( iBlockID );
+        
+        setHardness( 2F );
+        
+        setStepSound( soundStoneFootstep );
+        
+        setUnlocalizedName( "quartzBlock" );
+    }
+    
+	//----------- Client Side Functionality -----------//
+}
