--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockWaterFlowing.java
@@ -0,0 +1,29 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCBlockWaterFlowing extends BlockFlowing
+{
+    protected FCBlockWaterFlowing( int iBlockID, Material material)
+    {
+        super( iBlockID, material );
+    }
+    
+    @Override
+    public boolean CanPathThroughBlock( IBlockAccess blockAccess, int i, int j, int k, Entity entity, PathFinder pathFinder )
+    {
+    	return pathFinder.CanPathThroughWater();
+    }
+    
+    @Override
+    public int GetWeightOnPathBlocked( IBlockAccess blockAccess, int i, int j, int k )
+    {
+    	return -1;
+    }
+
+    @Override
+    public int AdjustPathWeightOnNotBlocked( int iPreviousWeight )
+    {
+    	return 2;
+    }
+}
