--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockSandstoneMouldingAndDecorative.java
@@ -0,0 +1,13 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCBlockSandstoneMouldingAndDecorative extends FCBlockMouldingAndDecorative
+{
+	protected FCBlockSandstoneMouldingAndDecorative( int iBlockID, int iMatchingCornerBlockID )
+	{
+		super( iBlockID, Material.rock, "fcBlockDecorativeSandstone_top", "fcBlockColumnSandstone_side", iMatchingCornerBlockID, 
+			0.8F, 1.34F, Block.soundStoneFootstep, "fcSandstoneMoulding" );
+	}
+	
+}
\ No newline at end of file
