--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockSnowLoose.java
@@ -0,0 +1,203 @@
+// FCMOD
+
+package net.minecraft.src;
+
+import java.util.Random;
+
+public class FCBlockSnowLoose extends FCBlockFalling
+{
+	public static final float m_fHardness = 0.2F;
+	
+	public static final int m_iChanceOfHardeningIncrement = 1;
+	public static final int m_iChanceOfMeltingIncrement = 1;
+	
+    public FCBlockSnowLoose( int iBlockID )
+    {
+        super( iBlockID, Material.craftedSnow );
+        
+    	setHardness( m_fHardness );        
+        SetShovelsEffectiveOn();
+        
+    	SetBuoyant();
+    	
+    	setStepSound( soundSnowFootstep );
+        
+        setUnlocalizedName( "fcBlockSnowLoose" );
+        
+        setTickRandomly( true );
+		
+        setLightOpacity( 4 );
+        Block.useNeighborBrightness[iBlockID] = true;
+        
+		setCreativeTab( CreativeTabs.tabBlock );
+    }
+    
+	@Override
+    public int idDropped( int iMetadata, Random rand, int iFortuneModifier )
+    {
+        return Item.snowball.itemID;
+    }
+
+    public int quantityDropped( Random rand )
+    {
+        return 8;
+    }
+
+    @Override
+    public void OnBlockDestroyedWithImproperTool( World world, EntityPlayer player, int i, int j, int k, int iMetadata )
+    {
+        dropBlockAsItem( world, i, j, k, iMetadata, 0 );
+    }
+    
+    @Override
+    public void onBlockAdded( World world, int i, int j, int k )
+    {
+        if ( world.provider.isHellWorld )
+        {
+        	// melt instantly in the nether
+        	
+        	world.setBlockToAir( i, j, k );
+        	
+            world.playSoundEffect( (double)i + 0.5D, (double)j + 0.5D, (double)k + 0.5D, 
+            	"random.fizz", 0.5F, 2.6F + ( world.rand.nextFloat() - world.rand.nextFloat() ) * 0.8F );
+            
+            for ( int iTempCount = 0; iTempCount < 8; iTempCount++ )
+            {
+                world.spawnParticle( "largesmoke", (double)i + Math.random(), (double)j + Math.random(), (double)k + Math.random(), 
+                	0D, 0D, 0D );
+            }
+        }
+        else
+        {
+        	ScheduleCheckForFall( world, i, j, k );
+        }
+    }
+    
+    @Override
+    public void updateTick( World world, int i, int j, int k, Random rand ) 
+    {
+    	if ( !HasStickySnowNeighborInContact( world, i, j, k ) || HasFallingBlockRestingOn( world, i, j, k ) )
+    	{
+    		CheckForFall( world, i, j, k );
+    	}
+    }
+    
+	@Override
+    public void RandomUpdateTick( World world, int i, int j, int k, Random rand )
+    {
+		if ( FCUtilsMisc.IsIKInColdBiome( world, i, k ) )
+		{
+			if ( rand.nextInt( m_iChanceOfHardeningIncrement ) == 0 )
+			{
+				int iHardeningLevel = GetHardeningLevel( world, i, j, k );
+				
+				if ( iHardeningLevel < 7 )
+				{
+					SetHardeningLevel( world, i, j, k, iHardeningLevel + 1 );
+				}
+				else
+				{
+					ConvertToSolidSnow( world, i, j, k );
+				}
+			}
+		}
+		else
+		{
+			if ( rand.nextInt( m_iChanceOfMeltingIncrement ) == 0 )
+			{
+				int iHardeningLevel = GetHardeningLevel( world, i, j, k );
+				
+				if ( iHardeningLevel > 0 )
+				{
+					SetHardeningLevel( world, i, j, k, iHardeningLevel - 1 );
+				}
+				else
+				{
+					Melt( world, i, j, k );
+				}
+			}
+		}
+    }
+	
+	@Override
+    public void OnPlayerWalksOnBlock( World world, int i, int j, int k, EntityPlayer player )
+    {
+		CheckForFall( world, i, j, k );
+    }
+    
+    @Override
+    public float GetMovementModifier( World world, int i, int j, int k )
+    {
+    	return 0.8F;
+    }
+    
+    @Override
+    public boolean GetCanBeSetOnFireDirectly( IBlockAccess blockAccess, int i, int j, int k )
+    {
+    	return true;
+    }
+    
+    @Override
+    public boolean GetCanBeSetOnFireDirectlyByItem( IBlockAccess blockAccess, int i, int j, int k )
+    {
+    	return false;
+    }
+    
+    @Override
+    public boolean SetOnFireDirectly( World world, int i, int j, int k )
+    {
+		Melt( world, i, j, k );
+		
+    	return true;
+    }
+    
+    @Override
+    public int GetChanceOfFireSpreadingDirectlyTo( IBlockAccess blockAccess, int i, int j, int k )
+    {
+		return 60; // same chance as leaves and other highly flammable objects
+    }
+    
+    @Override
+    public boolean CanBePistonShoveled( World world, int i, int j, int k )
+    {
+    	return true;
+    }
+    
+    //------------- Class Specific Methods ------------//
+    
+    public int GetHardeningLevel( IBlockAccess blockAccess, int i, int j, int k )
+    {
+    	return GetHardeningLevel( blockAccess.getBlockMetadata( i, j, k ) );
+    }
+    
+    public int GetHardeningLevel( int iMetadata )
+    {
+    	return ( iMetadata & 14 ) >> 1; 
+    }
+
+    public void SetHardeningLevel( World world, int i, int j, int k, int iLevel )
+    {
+    	int iMetadata = SetHardeningLevel( world.getBlockMetadata( i, j, k ), iLevel );
+    	
+    	world.setBlockMetadataWithNotify( i, j, k, iMetadata );
+    }
+    	
+    public int SetHardeningLevel( int iMetadata, int iLevel )
+    {
+    	iMetadata &= ~14;
+    	
+    	return iMetadata | ( iLevel << 1 );
+    }
+    
+	private void ConvertToSolidSnow( World world, int i, int j, int k )
+	{
+		world.setBlockWithNotify( i, j, k, FCBetterThanWolves.fcBlockSnowSolid.blockID );
+	}
+	
+	private void Melt( World world, int i, int j, int k )
+	{
+		FCUtilsMisc.PlaceNonPersistantWaterMinorSpread( world, i, j, k );
+	}
+	
+	//------------ Client Side Functionality ----------//    
+}
\ No newline at end of file
