--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockStoneBrickLooseStairs.java
@@ -0,0 +1,28 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCBlockStoneBrickLooseStairs extends FCBlockMortarReceiverStairs
+{
+    protected FCBlockStoneBrickLooseStairs( int iBlockID )
+    {
+        super( iBlockID, FCBetterThanWolves.fcBlockStoneBrickLoose, 0 );
+        
+        SetPicksEffectiveOn();
+        
+        setUnlocalizedName( "fcBlockStoneBrickLooseStairs" );        
+    }
+	
+    @Override
+    public boolean OnMortarApplied( World world, int i, int j, int k )
+    {
+		world.setBlockAndMetadataWithNotify( i, j, k, Block.stairsStoneBrick.blockID, 
+			world.getBlockMetadata( i, j, k ) );
+		
+		return true;
+    }
+    
+    //------------- Class Specific Methods ------------//
+	
+	//----------- Client Side Functionality -----------//
+}
\ No newline at end of file
