--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockChunkOreIron.java
@@ -0,0 +1,21 @@
+// FCMOD
+
+package net.minecraft.src;
+
+import java.util.Random;
+
+public class FCBlockChunkOreIron extends FCBlockChunkOre
+{
+    protected FCBlockChunkOreIron( int iBlockID )
+    {
+        super( iBlockID );
+        
+        setUnlocalizedName( "fcBlockChunkOreIron" );
+    }
+    
+	@Override
+    public int idDropped( int iMetadata, Random random, int iFortuneModifier )
+    {
+		return FCBetterThanWolves.fcItemChunkIronOre.itemID;
+    }
+}
\ No newline at end of file
