--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCBlockNetherBrickMouldingAndDecorative.java
@@ -0,0 +1,13 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCBlockNetherBrickMouldingAndDecorative extends FCBlockMouldingAndDecorative
+{
+	protected FCBlockNetherBrickMouldingAndDecorative( int iBlockID, int iMatchingCornerBlockID )
+	{
+		super( iBlockID, FCBetterThanWolves.fcMaterialNetherRock, 
+			"fcBlockDecorativeNetherBrick", "fcBlockColumnNetherBrick_side", 
+			iMatchingCornerBlockID, 2.0F, 10F, Block.soundStoneFootstep, "fcNetherBrickMoulding" );		
+	}
+}
\ No newline at end of file
