--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCCraftingManagerCrucible.java
@@ -0,0 +1,18 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCCraftingManagerCrucible extends FCCraftingManagerBulk
+{
+    private static final FCCraftingManagerCrucible instance = new FCCraftingManagerCrucible();
+    
+    public static final FCCraftingManagerCrucible getInstance()
+    {
+        return instance;
+    }
+
+    private FCCraftingManagerCrucible()
+    {
+    	super();    	
+    }
+}
