--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCCraftingManagerPistonPackingRecipe.java
@@ -0,0 +1,52 @@
+package net.minecraft.src;
+
+public class FCCraftingManagerPistonPackingRecipe {
+	private final Block output;
+	private final int outputMetadata;
+	private final ItemStack[] input;
+	
+	public FCCraftingManagerPistonPackingRecipe(Block output, int outputMetadata, ItemStack[] input) {
+		this.output = output;
+		this.outputMetadata = outputMetadata;
+		this.input = input;
+	}
+	
+	public boolean matchesRecipe(FCCraftingManagerPistonPackingRecipe recipe) {
+		if (output.blockID == recipe.output.blockID &&
+				outputMetadata == recipe.outputMetadata)
+		{
+			return input.equals(recipe.input);
+		}
+		
+		return false;
+	}
+	
+	public boolean matchesInputs(ItemStack[] inputToMatch) {
+		if (this.input.length == inputToMatch.length) {
+			for (int i = 0; i < this.input.length; i++) {
+				if (this.input[i].getItem().itemID == inputToMatch[i].getItem().itemID) {
+					if (this.input[i].getItemDamage() != inputToMatch[i].getItemDamage() &&
+							this.input[i].getItemDamage() != FCUtilsInventory.m_iIgnoreMetadata) {
+						return false;
+					}
+				}
+			}
+			
+			return true;
+		}
+		
+		return false;
+	}
+
+	public Block getOutput() {
+		return output;
+	}
+
+	public int getOutputMetadata() {
+		return outputMetadata;
+	}
+
+	public ItemStack[] getInput() {
+		return input;
+	}
+}
\ No newline at end of file
