--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCCraftingManagerHopperFilterRecipe.java
@@ -0,0 +1,49 @@
+package net.minecraft.src;
+
+public class FCCraftingManagerHopperFilterRecipe {
+	private final ItemStack hopperOutput;
+	private final ItemStack filteredOutput;
+	private final ItemStack input;
+	private final ItemStack filterUsed;
+	private final boolean containsSouls;
+	
+	public FCCraftingManagerHopperFilterRecipe(ItemStack hopperOutput, ItemStack filteredOutput, ItemStack input, ItemStack filterUsed, boolean containsSouls) {
+		this.hopperOutput = hopperOutput;
+		this.filteredOutput = filteredOutput;
+		this.input = input;
+		this.filterUsed = filterUsed;
+		this.containsSouls = containsSouls;
+	}
+	
+	public boolean matchesRecipe(FCCraftingManagerHopperFilterRecipe recipe) {
+		return this.hopperOutput.equals(recipe.hopperOutput) &&
+				this.filteredOutput.equals(recipe.filteredOutput) &&
+				this.input.isItemEqual(recipe.input) &&
+				this.filterUsed.isItemEqual(recipe.filterUsed );
+	}
+	
+	public boolean matchesInputs(ItemStack inputToCheck, ItemStack filterToCheck) {
+		return this.input.isItemEqual(inputToCheck) &&
+				this.filterUsed.isItemEqual(filterToCheck);
+	}
+
+	public ItemStack getHopperOutput() {
+		return hopperOutput;
+	}
+
+	public ItemStack getFilteredOutput() {
+		return filteredOutput;
+	}
+
+	public ItemStack getInput() {
+		return input;
+	}
+
+	public ItemStack getFilterUsed() {
+		return filterUsed;
+	}
+	
+	public boolean getContainsSouls() {
+		return containsSouls;
+	}
+}
\ No newline at end of file
