--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCCraftingManagerCauldron.java
@@ -0,0 +1,19 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCCraftingManagerCauldron  extends FCCraftingManagerBulk
+{
+    private static final FCCraftingManagerCauldron instance = new FCCraftingManagerCauldron();
+    
+    public static final FCCraftingManagerCauldron getInstance()
+    {
+        return instance;
+    }
+
+    private FCCraftingManagerCauldron()
+    {
+    	super();    	
+    }
+}
+
