--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCIBlockFluidSource.java
@@ -0,0 +1,11 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public interface FCIBlockFluidSource
+{
+	/*
+	 * Returns the height level of the source (0 to 8) if a valid source for the fluid block, -1 otherwise
+	 */
+	public int IsSourceToFluidBlockAtFacing( World world, int i, int j, int k, int iFacing  );
+}
\ No newline at end of file
