--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCIBlockMechanical.java
@@ -0,0 +1,18 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public interface FCIBlockMechanical
+{
+	public boolean CanOutputMechanicalPower();
+	
+	public boolean CanInputMechanicalPower();
+	
+	public boolean IsInputtingMechanicalPower( World world, int i, int j, int k );
+	
+	public boolean IsOutputtingMechanicalPower( World world, int i, int j, int k );
+	
+	public boolean CanInputAxlePowerToFacing( World world, int i, int j, int k, int iFacing );
+	
+	public void Overpower( World world, int i, int j, int k );
+}
\ No newline at end of file
