--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCItemWickerPaneLegacy.java
@@ -0,0 +1,27 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCItemWickerPaneLegacy extends FCItemPlacesAsBlock
+{
+    public FCItemWickerPaneLegacy( int iItemID )
+    {
+    	super( iItemID, FCBetterThanWolves.fcAestheticNonOpaque.blockID, 
+    		FCBlockAestheticNonOpaque.m_iSubtypeWicker, "fcItemWicker" );
+    	
+    	SetBuoyant();
+    	SetIncineratedInCrucible();    	
+    }
+	
+    @Override
+    public boolean CanItemPassIfFilter( ItemStack filteredItem )
+    {
+    	int iFilterableProperties = filteredItem.getItem().GetFilterableProperties( filteredItem ); 
+		
+    	return ( iFilterableProperties & Item.m_iFilterable_Fine ) != 0;
+    }
+    
+    //------------- Class Specific Methods ------------//
+	
+	//----------- Client Side Functionality -----------//
+}
