--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCItemStone.java
@@ -0,0 +1,20 @@
+//FCMOD
+
+package net.minecraft.src;
+
+public class FCItemStone extends Item
+{
+    public FCItemStone( int iItemID )
+    {
+        super( iItemID );
+        
+    	setUnlocalizedName( "fcItemStone" );
+    	
+    	SetFilterableProperties( Item.m_iFilterable_Small );
+    	
+    	setCreativeTab( CreativeTabs.tabMaterials );	    
+    }
+    
+    //------------- Class Specific Methods ------------//
+}
+    
