--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCItemChunkOreIron.java
@@ -0,0 +1,17 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCItemChunkOreIron extends FCItemPlacesAsBlock
+{
+    public FCItemChunkOreIron( int iItemID )
+    {
+        super( iItemID, FCBetterThanWolves.fcBlockChunkOreIron.blockID );
+        
+        SetFilterableProperties( Item.m_iFilterable_Small );
+        
+        setUnlocalizedName( "fcItemChunkIronOre" );
+        
+        setCreativeTab( CreativeTabs.tabMaterials );
+    }
+}
