--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCItemBlockLever.java
@@ -0,0 +1,23 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCItemBlockLever extends ItemBlock
+{
+    public FCItemBlockLever( int iItemID )
+    {
+        super( iItemID );
+    }
+    
+    @Override
+	public int GetTargetFacingPlacedByBlockDispenser( int iDispenserFacing )
+    {
+    	// always place facing upwards
+    	
+    	return 1;
+    }
+    
+    //------------- Class Specific Methods ------------//
+    
+	//------------ Client Side Functionality ----------//
+}
