--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCItemBlockBloodWood.java
@@ -0,0 +1,19 @@
+// FCMOD 
+
+package net.minecraft.src;
+
+public class FCItemBlockBloodWood extends ItemBlock
+{
+    public FCItemBlockBloodWood( int iItemID )
+    {
+        super( iItemID );
+    }
+    
+    @Override
+    public int GetCampfireBurnTime( int iItemDamage )
+    {
+    	// logs can't be burned directly in a campfire without being split first
+    	
+    	return 0;
+    }    
+}
