--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCItemBlockGrassSlab.java
@@ -0,0 +1,50 @@
+package net.minecraft.src;
+
+public class FCItemBlockGrassSlab extends FCItemBlockSlab {
+	public FCItemBlockGrassSlab(int id) {
+		super(id);
+	}
+
+	@Override
+	public boolean canCombineWithBlock(World world, int x, int y, int z, int itemDamage) {
+		int blockID = world.getBlockId(x, y, z);
+
+		if (blockID == FCBetterThanWolves.fcBlockDirtSlab.blockID) {
+			int targetSubtype = ((FCBlockDirtSlab)FCBetterThanWolves.fcBlockDirtSlab).GetSubtype(world, x, y, z);
+
+			if (targetSubtype != FCBlockDirtSlab.m_iSubtypePackedEarth) {
+				return true;
+			}
+		}
+		else if (blockID == FCBetterThanWolves.fcBlockGrassSlab.blockID) {
+			return true;
+		}
+
+		return false;
+	}
+
+	@Override
+	public boolean convertToFullBlock(World world, int x, int y, int z) {
+		int blockID = world.getBlockId(x, y, z);
+
+		if (blockID == FCBetterThanWolves.fcBlockDirtSlab.blockID) {
+			FCBlockDirtSlab slabBlock = (FCBlockDirtSlab)(FCBetterThanWolves.fcBlockDirtSlab);
+
+			boolean isTargetUpsideDown = slabBlock.GetIsUpsideDown(world, x, y, z);
+			int targetSubType = slabBlock.GetSubtype(world, x, y, z);
+
+			if (isTargetUpsideDown) {
+				if (targetSubType == FCBlockDirtSlab.m_iSubtypeDirt) {
+					return world.setBlockWithNotify(x, y, z, Block.dirt.blockID);
+				}
+			}
+
+			return world.setBlockWithNotify(x, y, z, Block.grass.blockID);
+		}
+		else if (blockID == FCBetterThanWolves.fcBlockGrassSlab.blockID) {
+			return world.setBlockWithNotify(x, y, z, Block.grass.blockID);
+		}
+
+		return false;
+	}
+}
\ No newline at end of file
