--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCItemArrowBroadhead.java
@@ -0,0 +1,34 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCItemArrowBroadhead extends FCItemArrow
+{
+    public FCItemArrowBroadhead( int iItemID )
+    {
+    	super( iItemID );
+		
+    	SetNeutralBuoyant();
+    	
+    	// neutralize parent properties
+		SetBellowsBlowDistance( 0 );		
+    	SetNotIncineratedInCrucible();
+    	SetFurnaceBurnTime( FCEnumFurnaceBurnTime.NONE );
+    	
+    	setUnlocalizedName( "fcItemArrowBroadhead" );
+    }
+
+    @Override
+    EntityArrow GetFiredArrowEntity( World world, double dXPos, double dYPos, double dZPos )
+    {
+    	EntityArrow entity = (EntityArrow) EntityList.createEntityOfType(FCEntityBroadheadArrow.class, world, dXPos, dYPos, dZPos );
+        
+        entity.canBePickedUp = 1;
+        
+        return entity;
+    }
+    
+    //------------- Class Specific Methods ------------//
+    
+	//------------ Client Side Functionality ----------//
+}
