--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCItemWaterWheel.java
@@ -0,0 +1,78 @@
+//FCMOD
+
+package net.minecraft.src;
+
+public class FCItemWaterWheel extends Item
+{
+    public FCItemWaterWheel( int iItemID )
+    {
+        super( iItemID );
+        
+        maxStackSize = 1;
+        
+        SetBuoyant();
+        
+        setUnlocalizedName( "fcItemWaterWheel" );        
+		
+        setCreativeTab( CreativeTabs.tabRedstone );
+    }
+    
+    @Override
+    public boolean onItemUse( ItemStack itemStack, EntityPlayer player, World world, int i, int j, int k, int iFacing, float fClickX, float fClickY, float fClickZ )    
+    {
+        int iTargetBlockID = world.getBlockId( i, j, k );
+        
+        if ( iTargetBlockID == FCBetterThanWolves.fcBlockAxle.blockID )
+        {
+        	int iAxisAlignment = ((FCBlockAxle)(FCBetterThanWolves.fcBlockAxle)).
+        		GetAxisAlignment( world, i, j, k );
+        	
+        	if ( iAxisAlignment != 0 )
+        	{
+        		boolean bIAligned = false;
+        		
+        		if ( iAxisAlignment == 2 )
+        		{
+	            	bIAligned = true;
+        		}
+
+        		FCEntityWaterWheel waterWheel = (FCEntityWaterWheel) EntityList.createEntityOfType(FCEntityWaterWheel.class, world,
+            		(float)i + 0.5F, (float)j + 0.5F, (float)k + 0.5F, bIAligned );
+
+        		if ( waterWheel.ValidateAreaAroundDevice() )
+        		{
+        			if ( waterWheel.IsClearOfBlockingEntities() )
+        			{
+	                    if( !world.isRemote )
+	                    {
+	                        waterWheel.setRotationSpeed( waterWheel.ComputeRotation() );
+	
+			                world.spawnEntityInWorld( waterWheel );
+	                    }
+		                
+		                itemStack.stackSize--;
+		                
+		                return true;
+        			}
+        			else
+        			{
+                        if( world.isRemote )
+                        {
+                        	player.addChatMessage( this.getUnlocalizedName() + ".placementObstructed" );
+                        }
+        			}
+        		}
+        		else
+        		{
+                    if( world.isRemote )
+                    {
+                    	player.addChatMessage( this.getUnlocalizedName() + ".notEnoughRoom" );
+                    }
+        		}
+        	}
+        }        
+        
+        return false;
+    }    
+}
+    
