--- /dev/null
+++ b/minecraft_server/net/minecraft/src/FCStepSoundSquish.java
@@ -0,0 +1,23 @@
+// FCMOD
+
+package net.minecraft.src;
+
+class FCStepSoundSquish extends StepSound
+{
+	FCStepSoundSquish( String s, float f, float f1 )
+    {
+        super( s, f, f1 );
+    }
+    
+    @Override
+    public String getStepSound()
+    {
+        return "mob.slime.attack";
+    }    
+
+    @Override
+    public String getBreakSound()
+    {
+        return "mob.slime.small";
+    }
+}
