--- a/minecraft_server/net/minecraft/src/InventoryPlayer.java
+++ b/minecraft_server/net/minecraft/src/InventoryPlayer.java
@@ -407,13 +407,23 @@
     /**
      * Gets the strength of the current item (tool) against the specified block, 1.0f if not holding anything.
      */
+    // FCMOD: Code change
+    /*
     public float getStrVsBlock(Block par1Block)
+    */
+    public float getStrVsBlock( World world, Block par1Block, int i, int j, int k )
+    // END FCMOD
     {
         float var2 = 1.0F;
 
         if (this.mainInventory[this.currentItem] != null)
         {
+            // FCMOD: Code change
+            /*
             var2 *= this.mainInventory[this.currentItem].getStrVsBlock(par1Block);
+            */
+            var2 *= this.mainInventory[this.currentItem].getStrVsBlock( world, par1Block, i, j, k );
+        	// END FCMOD
         }
 
         return var2;
@@ -544,7 +554,12 @@
     /**
      * Returns whether the current item (tool) can harvest from the specified block (actually get a result).
      */
+    // FCMOD: Code change
+    /*
     public boolean canHarvestBlock(Block par1Block)
+    */
+    public boolean canHarvestBlock(World world, Block par1Block, int i, int j, int k)
+    // END FCMOD
     {
         if (par1Block.blockMaterial.isToolNotRequired())
         {
@@ -553,7 +568,12 @@
         else
         {
             ItemStack var2 = this.getStackInSlot(this.currentItem);
+            // FCMOD: Code change
+            /*
             return var2 != null ? var2.canHarvestBlock(par1Block) : false;
+            */
+            return var2 != null ? var2.canHarvestBlock(world, par1Block, i, j, k) : false;
+            // END FCMOD
         }
     }
 
