--- a/minecraft_server/net/minecraft/src/ItemBow.java
+++ b/minecraft_server/net/minecraft/src/ItemBow.java
@@ -35,7 +35,7 @@
                 var7 = 1.0F;
             }
 
-            EntityArrow var8 = new EntityArrow(par2World, par3EntityPlayer, var7 * 2.0F);
+            EntityArrow var8 = (EntityArrow) EntityList.createEntityOfType(EntityArrow.class, par2World, par3EntityPlayer, var7 * 2.0F);
 
             if (var7 == 1.0F)
             {
