--- a/minecraft_server/net/minecraft/src/ItemSkull.java
+++ b/minecraft_server/net/minecraft/src/ItemSkull.java
@@ -2,8 +2,14 @@
 
 public class ItemSkull extends Item
 {
+	// FCMOD: Changed to add Infused Skull
+	/*
     private static final String[] skullTypes = new String[] {"skeleton", "wither", "zombie", "char", "creeper"};
     public static final String[] field_94587_a = new String[] {"skull_skeleton", "skull_wither", "skull_zombie", "skull_char", "skull_creeper"};
+    */
+    private static final String[] skullTypes = new String[] {"skeleton", "wither", "zombie", "char", "creeper", "infused"};
+    public static final String[] field_94587_a = new String[] {"skull_skeleton", "skull_wither", "skull_zombie", "skull_char", "skull_creeper", "fcItemSkullInfused"};
+	// END FCMOD
 
     public ItemSkull(int par1)
     {
@@ -23,7 +29,11 @@
         {
             return false;
         }
-        else if (!par3World.getBlockMaterial(par4, par5, par6).isSolid())
+        // FCMOD: Change
+        //else if (!par3World.getBlockMaterial(par4, par5, par6).isSolid())
+        else if ( !par3World.getBlockMaterial( par4, par5, par6 ).isSolid() && 
+        	!( par7 == 1 && FCUtilsWorld.DoesBlockHaveSmallCenterHardpointToFacing( par3World, par4, par5, par6, 1 ) ) )
+    	// END FCMOD
         {
             return false;
         }
@@ -64,7 +74,10 @@
             }
             else
             {
-                par3World.setBlock(par4, par5, par6, Block.skull.blockID, par7, 2);
+            	// FCMOD: Changed to notify neighbors
+                //par3World.setBlock(par4, par5, par6, Block.skull.blockID, par7, 2);
+                par3World.setBlock(par4, par5, par6, Block.skull.blockID, par7, 3);
+                // END FCMOD
                 int var11 = 0;
 
                 if (par7 == 1)
