--- a/minecraft_server/net/minecraft/src/ItemFishingRod.java
+++ b/minecraft_server/net/minecraft/src/ItemFishingRod.java
@@ -27,7 +27,7 @@
 
             if (!par2World.isRemote)
             {
-                par2World.spawnEntityInWorld(new EntityFishHook(par2World, par3EntityPlayer));
+            	par2World.spawnEntityInWorld(EntityList.createEntityOfType(EntityFishHook.class, par2World, par3EntityPlayer));
             }
 
             par3EntityPlayer.swingItem();
