--- a/minecraft_server/net/minecraft/src/Packet41EntityEffect.java
+++ b/minecraft_server/net/minecraft/src/Packet41EntityEffect.java
@@ -12,6 +12,9 @@
     /** The effect's amplifier. */
     public byte effectAmplifier;
     public short duration;
+    // FCMOD: Code added
+    public byte m_bIsAmbient;
+    // END FCMOD
 
     public Packet41EntityEffect() {}
 
@@ -29,7 +32,15 @@
         {
             this.duration = (short)par2PotionEffect.getDuration();
         }
+
+        // FCMOD: Code added
+        m_bIsAmbient = 0;
+        if ( par2PotionEffect.getIsAmbient() )
+        {
+        	m_bIsAmbient = 1;
     }
+        // END FCMOD
+    }
 
     /**
      * Abstract. Reads the raw packet data from the data stream.
@@ -40,6 +51,9 @@
         this.effectId = par1DataInputStream.readByte();
         this.effectAmplifier = par1DataInputStream.readByte();
         this.duration = par1DataInputStream.readShort();
+        // FCMOD: Code added
+        m_bIsAmbient = par1DataInputStream.readByte();
+        // END FCMOD
     }
 
     /**
@@ -51,6 +65,9 @@
         par1DataOutputStream.writeByte(this.effectId);
         par1DataOutputStream.writeByte(this.effectAmplifier);
         par1DataOutputStream.writeShort(this.duration);
+        // FCMOD: Code added
+        par1DataOutputStream.writeByte(m_bIsAmbient);
+        // END FCMOD
     }
 
     /**
@@ -66,7 +83,12 @@
      */
     public int getPacketSize()
     {
+    	// FCMOD: Code change
+    	/*
         return 8;
+        */
+    	return 9;
+    	// END FCMOD
     }
 
     /**
