--- a/minecraft_server/net/minecraft/src/Packet19EntityAction.java
+++ b/minecraft_server/net/minecraft/src/Packet19EntityAction.java
@@ -27,7 +27,7 @@
     public void writePacketData(DataOutputStream par1DataOutputStream) throws IOException
     {
         par1DataOutputStream.writeInt(this.entityId);
-        par1DataOutputStream.writeByte(this.state);
+        par1DataOutputStream.writeInt(this.state);
     }
 
     /**
@@ -43,6 +43,6 @@
      */
     public int getPacketSize()
     {
-        return 5;
+        return 8;
     }
 }
