--- a/minecraft_server/net/minecraft/src/Packet15Place.java
+++ b/minecraft_server/net/minecraft/src/Packet15Place.java
@@ -33,9 +33,16 @@
         this.zPosition = par1DataInputStream.readInt();
         this.direction = par1DataInputStream.read();
         this.itemStack = readItemStack(par1DataInputStream);
+        // FCMOD: Code change to fix rounding errors
+        /*
         this.xOffset = (float)par1DataInputStream.read() / 16.0F;
         this.yOffset = (float)par1DataInputStream.read() / 16.0F;
         this.zOffset = (float)par1DataInputStream.read() / 16.0F;
+        */
+        this.xOffset = (float)par1DataInputStream.readShort() / 32000F;
+        this.yOffset = (float)par1DataInputStream.readShort() / 32000F;
+        this.zOffset = (float)par1DataInputStream.readShort() / 32000F;
+        // END FCMOD
     }
 
     /**
@@ -48,9 +55,16 @@
         par1DataOutputStream.writeInt(this.zPosition);
         par1DataOutputStream.write(this.direction);
         writeItemStack(this.itemStack, par1DataOutputStream);
+        // FCMOD: Code change to fix rounding errors
+        /*
         par1DataOutputStream.write((int)(this.xOffset * 16.0F));
         par1DataOutputStream.write((int)(this.yOffset * 16.0F));
         par1DataOutputStream.write((int)(this.zOffset * 16.0F));
+        */
+        par1DataOutputStream.writeShort((int)(this.xOffset * 32000F + 0.5F));
+        par1DataOutputStream.writeShort((int)(this.yOffset * 32000F + 0.5F));
+        par1DataOutputStream.writeShort((int)(this.zOffset * 32000F + 0.5F));
+        // END FCMOD        
     }
 
     /**
@@ -66,7 +80,12 @@
      */
     public int getPacketSize()
     {
+    	// FCMOD: Code change
+    	/*
         return 19;
+        */
+        return 22;
+        // END FCMOD
     }
 
     public int getXPosition()
