--- a/minecraft_server/net/minecraft/src/RecipeFireworks.java
+++ b/minecraft_server/net/minecraft/src/RecipeFireworks.java
@@ -77,8 +77,8 @@
 
         if (var4 <= 3 && var3 <= 1)
         {
-            NBTTagCompound var16;
-            NBTTagCompound var19;
+            NBTTagCompound var15;
+            NBTTagCompound var18;
 
             if (var4 >= 1 && var3 == 1 && var7 == 0)
             {
@@ -86,8 +86,8 @@
 
                 if (var6 > 0)
                 {
-                    var16 = new NBTTagCompound();
-                    var19 = new NBTTagCompound("Fireworks");
+                    var15 = new NBTTagCompound();
+                    var18 = new NBTTagCompound("Fireworks");
                     NBTTagList var25 = new NBTTagList("Explosions");
 
                     for (int var22 = 0; var22 < par1InventoryCrafting.getSizeInventory(); ++var22)
@@ -100,10 +100,10 @@
                         }
                     }
 
-                    var19.setTag("Explosions", var25);
-                    var19.setByte("Flight", (byte)var4);
-                    var16.setTag("Fireworks", var19);
-                    this.field_92102_a.setTagCompound(var16);
+                    var18.setTag("Explosions", var25);
+                    var18.setByte("Flight", (byte)var4);
+                    var15.setTag("Fireworks", var18);
+                    this.field_92102_a.setTagCompound(var15);
                 }
 
                 return true;
@@ -111,9 +111,9 @@
             else if (var4 == 1 && var3 == 0 && var6 == 0 && var5 > 0 && var8 <= 1)
             {
                 this.field_92102_a = new ItemStack(Item.fireworkCharge);
-                var16 = new NBTTagCompound();
-                var19 = new NBTTagCompound("Explosion");
-                byte var23 = 0;
+                var15 = new NBTTagCompound();
+                var18 = new NBTTagCompound("Explosion");
+                byte var21 = 0;
                 ArrayList var12 = new ArrayList();
 
                 for (int var13 = 0; var13 < par1InventoryCrafting.getSizeInventory(); ++var13)
@@ -128,27 +128,27 @@
                         }
                         else if (var14.itemID == Item.lightStoneDust.itemID)
                         {
-                            var19.setBoolean("Flicker", true);
+                            var18.setBoolean("Flicker", true);
                         }
                         else if (var14.itemID == Item.diamond.itemID)
                         {
-                            var19.setBoolean("Trail", true);
+                            var18.setBoolean("Trail", true);
                         }
                         else if (var14.itemID == Item.fireballCharge.itemID)
                         {
-                            var23 = 1;
+                            var21 = 1;
                         }
                         else if (var14.itemID == Item.feather.itemID)
                         {
-                            var23 = 4;
+                            var21 = 4;
                         }
                         else if (var14.itemID == Item.goldNugget.itemID)
                         {
-                            var23 = 2;
+                            var21 = 2;
                         }
                         else if (var14.itemID == Item.skull.itemID)
                         {
-                            var23 = 3;
+                            var21 = 3;
                         }
                     }
                 }
@@ -160,25 +160,25 @@
                     var24[var27] = ((Integer)var12.get(var27)).intValue();
                 }
 
-                var19.setIntArray("Colors", var24);
-                var19.setByte("Type", var23);
-                var16.setTag("Explosion", var19);
-                this.field_92102_a.setTagCompound(var16);
+                var18.setIntArray("Colors", var24);
+                var18.setByte("Type", var21);
+                var15.setTag("Explosion", var18);
+                this.field_92102_a.setTagCompound(var15);
                 return true;
             }
             else if (var4 == 0 && var3 == 0 && var6 == 1 && var5 > 0 && var5 == var7)
             {
-                ArrayList var15 = new ArrayList();
+                ArrayList var16 = new ArrayList();
 
-                for (int var17 = 0; var17 < par1InventoryCrafting.getSizeInventory(); ++var17)
+                for (int var20 = 0; var20 < par1InventoryCrafting.getSizeInventory(); ++var20)
                 {
-                    ItemStack var11 = par1InventoryCrafting.getStackInSlot(var17);
+                    ItemStack var11 = par1InventoryCrafting.getStackInSlot(var20);
 
                     if (var11 != null)
                     {
                         if (var11.itemID == Item.dyePowder.itemID)
                         {
-                            var15.add(Integer.valueOf(ItemDye.dyeColors[var11.getItemDamage()]));
+                            var16.add(Integer.valueOf(ItemDye.dyeColors[var11.getItemDamage()]));
                         }
                         else if (var11.itemID == Item.fireworkCharge.itemID)
                         {
@@ -188,24 +188,24 @@
                     }
                 }
 
-                int[] var18 = new int[var15.size()];
+                int[] var17 = new int[var16.size()];
 
-                for (int var20 = 0; var20 < var18.length; ++var20)
+                for (int var19 = 0; var19 < var17.length; ++var19)
                 {
-                    var18[var20] = ((Integer)var15.get(var20)).intValue();
+                    var17[var19] = ((Integer)var16.get(var19)).intValue();
                 }
 
                 if (this.field_92102_a != null && this.field_92102_a.hasTagCompound())
                 {
-                    NBTTagCompound var21 = this.field_92102_a.getTagCompound().getCompoundTag("Explosion");
+                    NBTTagCompound var23 = this.field_92102_a.getTagCompound().getCompoundTag("Explosion");
 
-                    if (var21 == null)
+                    if (var23 == null)
                     {
                         return false;
                     }
                     else
                     {
-                        var21.setIntArray("FadeColors", var18);
+                        var23.setIntArray("FadeColors", var17);
                         return true;
                     }
                 }
@@ -245,4 +245,23 @@
     {
         return this.field_92102_a;
     }
+    
+    // FCMOD: Function added
+    @Override
+    public boolean matches( IRecipe recipe )
+    {
+    	return false;
 }
+    
+    @Override
+    public boolean HasSecondaryOutput()
+    {
+    	return false;
+    }
+    
+    @Override
+    public ItemStack[] getSecondaryOutput(IInventory inventory) {
+    	return null;
+    }
+    // END FCMOD
+}
