--- a/minecraft_server/net/minecraft/src/RecipesMapCloning.java
+++ b/minecraft_server/net/minecraft/src/RecipesMapCloning.java
@@ -104,4 +104,18 @@
     {
         return null;
     }
+    
+    // FCMOD: Function added
+    @Override
+    public boolean matches( IRecipe recipe )
+    {
+    	return false;
 }
+    
+    @Override
+    public boolean HasSecondaryOutput()
+    {
+    	return false;
+    }
+    // END FCMOD
+}
