--- a/minecraft_server/net/minecraft/src/TileEntityBeacon.java
+++ b/minecraft_server/net/minecraft/src/TileEntityBeacon.java
@@ -349,4 +349,31 @@
     {
         return par2ItemStack.itemID == Item.emerald.itemID || par2ItemStack.itemID == Item.diamond.itemID || par2ItemStack.itemID == Item.ingotGold.itemID || par2ItemStack.itemID == Item.ingotIron.itemID;
     }
+    
+    // FCMOD: Code added
+    protected boolean IsOn()
+    {
+    	return isBeaconActive;
 }
+    
+    protected void SetIsOn( boolean bOn )
+    {
+    	isBeaconActive = bOn;
+    }
+    
+    protected void SetPrimaryEffect( int iEffect )
+    {
+    	primaryEffect = iEffect;
+    }
+    
+    public void setLevelsServerSafe(int par1)
+    {
+        levels = par1;
+    }
+    
+    public boolean IsHomeBeacon()
+    {
+    	return false;
+    }
+    // END FCMOD
+}
