--- a/minecraft_server/net/minecraft/src/Village.java
+++ b/minecraft_server/net/minecraft/src/Village.java
@@ -64,6 +64,8 @@
             this.updateNumIronGolems();
         }
 
+        // FCMOD: Code removed to get rid of spawning of iron golems
+        /*
         int var2 = this.numVillagers / 10;
 
         if (this.numIronGolems < var2 && this.villageDoorInfoList.size() > 20 && this.worldObj.rand.nextInt(7000) == 0)
@@ -78,6 +80,8 @@
                 ++this.numIronGolems;
             }
         }
+        */
+        // END FCMOD
     }
 
     /**
@@ -391,6 +395,14 @@
                 var4.resetDoorOpeningRestrictionCounter();
             }
 
+            // FCMOD: Added to prevent unintentionally loading chunks
+            if ( !worldObj.chunkExists( var4.posX >> 4, var4.posZ >> 4 ) )
+            {
+            	continue;
+            	
+            }
+            // END FCMOD
+
             if (!this.isBlockDoor(var4.posX, var4.posY, var4.posZ) || Math.abs(this.tickCounter - var4.lastActivityTimestamp) > 1200)
             {
                 this.centerHelper.posX -= var4.posX;
@@ -411,7 +423,11 @@
     private boolean isBlockDoor(int par1, int par2, int par3)
     {
         int var4 = this.worldObj.getBlockId(par1, par2, par3);
-        return var4 <= 0 ? false : var4 == Block.doorWood.blockID;
+        
+        // FCMOD: Changed
+        //return var4 <= 0 ? false : var4 == Block.doorWood.blockID;
+        return var4 <= 0 ? false : ( var4 == Block.doorWood.blockID || var4 == FCBetterThanWolves.fcBlockDoorWood.blockID );
+    	// END FCMOD
     }
 
     private void updateVillageRadiusAndCenter()
