--- a/minecraft_server/net/minecraft/src/VillageCollection.java
+++ b/minecraft_server/net/minecraft/src/VillageCollection.java
@@ -329,7 +329,11 @@
     private boolean isWoodenDoorAt(int par1, int par2, int par3)
     {
         int var4 = this.worldObj.getBlockId(par1, par2, par3);
-        return var4 == Block.doorWood.blockID;
+        
+        // FCMOD: Changed
+        //return var4 == Block.doorWood.blockID;
+        return var4 == Block.doorWood.blockID || var4 == FCBetterThanWolves.fcBlockDoorWood.blockID;
+        // END FCMOD
     }
 
     /**
