--- a/minecraft_server/net/minecraft/src/WorldGenDungeons.java
+++ b/minecraft_server/net/minecraft/src/WorldGenDungeons.java
@@ -81,7 +81,7 @@
                 {
                     if (var11 < 3)
                     {
-                        label113:
+                        label210:
                         {
                             var12 = par3 + par2Random.nextInt(var7 * 2 + 1) - var7;
                             int var14 = par5 + par2Random.nextInt(var8 * 2 + 1) - var8;
@@ -112,7 +112,12 @@
 
                                 if (var15 == 1)
                                 {
+                                    // FCMOD: Changed
+                                    /*
                                     par1World.setBlock(var12, par4, var14, Block.chest.blockID, 0, 2);
+                                    */
+                                    par1World.setBlock(var12, par4, var14, FCBetterThanWolves.fcBlockChest.blockID, 0, 2);
+                                	// END FCMOD
                                     TileEntityChest var16 = (TileEntityChest)par1World.getBlockTileEntity(var12, par4, var14);
 
                                     if (var16 != null)
@@ -128,7 +133,11 @@
                                         }
                                     }
 
-                                    break label113;
+                                    // FCMOD: Code added
+                                    FilterChestContentsForDepth( par1World, var12, par4, var14 );
+                                    // END FCMOD
+
+                                    break label210;
                                 }
                             }
 
@@ -179,4 +188,62 @@
         int var2 = par1Random.nextInt(4);
         return var2 == 0 ? "Skeleton" : (var2 == 1 ? "Zombie" : (var2 == 2 ? "Zombie" : (var2 == 3 ? "Spider" : "")));
     }
+    
+    // FCMOD: Code added
+    private void FilterChestContentsForDepth( World world, int iChestI, int iChestJ, int iChestK )
+    {
+        TileEntityChest chestEnt = (TileEntityChest)world.getBlockTileEntity( iChestI, iChestJ, iChestK );
+        
+        if ( chestEnt != null )
+        {
+            for ( int iSlot = 0; iSlot < chestEnt.getSizeInventory(); iSlot++ )
+            {
+                ItemStack tempStack = chestEnt.getStackInSlot( iSlot );
+                
+                if ( tempStack != null )
+                {
+                	int iItemID = tempStack.itemID;
+
+                	if ( iItemID == Item.ingotIron.itemID )
+                	{
+                		if ( iChestJ > 36 )
+                		{
+                    		chestEnt.setInventorySlotContents( iSlot, null );
 }
+                		else
+                		{
+                			tempStack.stackSize = 1;
+                		}
+                	}
+                	else if ( iItemID == Item.bucketEmpty.itemID )
+                	{
+                		chestEnt.setInventorySlotContents( iSlot, null );
+                	}
+                	else if ( iItemID == Item.gunpowder.itemID )
+                	{
+                		if ( iChestJ > 36 )
+                		{
+                    		chestEnt.setInventorySlotContents( iSlot, null );
+                		}
+                		else
+                		{
+                			tempStack.stackSize = 1;
+                		}
+                	}
+                	else if ( iItemID == Item.redstone.itemID )
+                	{
+                		if ( iChestJ > 24 )
+                		{
+                    		chestEnt.setInventorySlotContents( iSlot, null );
+                		}
+                	}
+                	else if ( iItemID == Item.wheat.itemID )
+                	{
+                		chestEnt.setInventorySlotContents( iSlot, null );
+                	}
+                }
+            }
+        }                	
+    }
+    // END FCMOD
+}
