--- a/minecraft_server/net/minecraft/server/MinecraftServer.java
+++ b/minecraft_server/net/minecraft/server/MinecraftServer.java
@@ -454,7 +454,7 @@
                     var50 += var7;
                     var1 = var5;
 
-                    if (this.worldServers[0].areAllPlayersAsleep())
+                    if (this.areAllPlayersAsleep())
                     {
                         this.tick();
                         var50 = 0L;
@@ -1428,4 +1428,19 @@
     {
         return par0MinecraftServer.serverConfigManager;
     }
+    
+    public boolean areAllPlayersAsleep() {
+    	boolean allPlayersSleeping = true;
+
+    	for (int i = 0; i < this.worldServers.length; i++) {
+    		WorldServer world = this.worldServers[i];
+
+    		if (world != null && !world.areAllPlayersAsleep()) {
+    			allPlayersSleeping = false;
+    			break;
 }
+    	}
+
+    	return allPlayersSleeping;
+    }
+}
