--- a/minecraft/net/minecraft/src/BehaviorDispenseMinecart.java
+++ b/minecraft/net/minecraft/src/BehaviorDispenseMinecart.java
@@ -7,6 +7,8 @@
     /**
      * Dispense the specified stack, play the dispense sound and spawn particles.
      */
+	// FCMOD: Code removed to prevent usage
+    /*
     public ItemStack dispenseStack(IBlockSource par1IBlockSource, ItemStack par2ItemStack)
     {
         EnumFacing var3 = BlockDispenser.getFacing(par1IBlockSource.getBlockMetadata());
@@ -39,12 +41,18 @@
         par2ItemStack.splitStack(1);
         return par2ItemStack;
     }
+    */
+    // END FCMOD
 
     /**
      * Play the dispense sound from the specified block.
      */
+	// FCMOD: Code removed
+    /*
     protected void playDispenseSound(IBlockSource par1IBlockSource)
     {
         par1IBlockSource.getWorld().playAuxSFX(1000, par1IBlockSource.getXInt(), par1IBlockSource.getYInt(), par1IBlockSource.getZInt(), 0);
     }
+    */
+    // END FCMOD
 }
