--- a/minecraft/net/minecraft/src/BlockVine.java
+++ b/minecraft/net/minecraft/src/BlockVine.java
@@ -1,5 +1,6 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.cc.ColorizeBlock;
 import java.util.Random;
 
 public class BlockVine extends Block
@@ -205,7 +206,7 @@
 
     public int getBlockColor()
     {
-        return ColorizerFoliage.getFoliageColorBasic();
+    	return ColorizeBlock.colorizeBlock(this) ? ColorizeBlock.blockColor : ColorizerFoliage.getFoliageColorBasic();
     }
 
     /**
@@ -213,7 +214,7 @@
      */
     public int getRenderColor(int par1)
     {
-        return ColorizerFoliage.getFoliageColorBasic();
+    	return ColorizeBlock.colorizeBlock(this, par1) ? ColorizeBlock.blockColor : ColorizerFoliage.getFoliageColorBasic();
     }
 
     /**
@@ -222,7 +223,7 @@
      */
     public int colorMultiplier(IBlockAccess par1IBlockAccess, int par2, int par3, int par4)
     {
-        return par1IBlockAccess.getBiomeGenForCoords(par2, par4).getBiomeFoliageColor();
+    	return ColorizeBlock.colorizeBlock(this, par1IBlockAccess, par2, par3, par4) ? ColorizeBlock.blockColor : par1IBlockAccess.getBiomeGenForCoords(par2, par4).getBiomeFoliageColor();
     }
 
     /**
