--- a/minecraft/net/minecraft/src/BlockDispenser.java
+++ b/minecraft/net/minecraft/src/BlockDispenser.java
@@ -233,7 +233,7 @@
                         }
 
                         var9.stackSize -= var13;
-                        EntityItem var14 = new EntityItem(par1World, (double)((float)par2 + var10), (double)((float)par3 + var11), (double)((float)par4 + var12), new ItemStack(var9.itemID, var13, var9.getItemDamage()));
+                        EntityItem var14 = (EntityItem) EntityList.createEntityOfType(EntityItem.class, par1World, (double)((float)par2 + var10), (double)((float)par3 + var11), (double)((float)par4 + var12), new ItemStack(var9.itemID, var13, var9.getItemDamage()));
 
                         if (var9.hasTagCompound())
                         {
