--- a/minecraft/net/minecraft/src/BlockJukeBox.java
+++ b/minecraft/net/minecraft/src/BlockJukeBox.java
@@ -75,7 +75,7 @@
                     double var10 = (double)(par1World.rand.nextFloat() * var7) + (double)(1.0F - var7) * 0.2D + 0.6D;
                     double var12 = (double)(par1World.rand.nextFloat() * var7) + (double)(1.0F - var7) * 0.5D;
                     ItemStack var14 = var6.copy();
-                    EntityItem var15 = new EntityItem(par1World, (double)par2 + var8, (double)par3 + var10, (double)par4 + var12, var14);
+                    EntityItem var15 = (EntityItem) EntityList.createEntityOfType(EntityItem.class, par1World, (double)par2 + var8, (double)par3 + var10, (double)par4 + var12, var14);
                     var15.delayBeforeCanPickup = 10;
                     par1World.spawnEntityInWorld(var15);
                 }
