--- a/minecraft/net/minecraft/src/BlockLog.java
+++ b/minecraft/net/minecraft/src/BlockLog.java
@@ -162,4 +162,13 @@
             this.iconArray[var2] = par1IconRegister.registerIcon(treeTextureTypes[var2]);
         }
     }
+    
+    // FCMOD: Code added
+    protected BlockLog( int iBlockID, Material material )
+    {
+        super(iBlockID, material );
+        
+        setCreativeTab( CreativeTabs.tabBlock );
 }
+    // END FCMOD
+}
