--- a/minecraft/net/minecraft/src/BlockPistonBase.java
+++ b/minecraft/net/minecraft/src/BlockPistonBase.java
@@ -5,16 +5,28 @@
 public class BlockPistonBase extends Block
 {
     /** This pistons is the sticky one? */
-    private final boolean isSticky;
+	// FCMOD: Changed
+    //private final boolean isSticky;
+    protected final boolean isSticky;
+    // END FCMOD
 
     /** Only visible when piston is extended */
-    private Icon innerTopIcon;
+	// FCMOD: Changed (client only)
+    //private Icon innerTopIcon;
+    protected Icon innerTopIcon;
+    // END FCMOD
 
     /** Bottom side texture */
-    private Icon bottomIcon;
+	// FCMOD: Changed (client only)
+    //private Icon bottomIcon;
+    protected Icon bottomIcon;
+    // END FCMOD
 
     /** Top icon of piston depends on (either sticky or normal) */
-    private Icon topIcon;
+	// FCMOD: Changed (client only)
+    //private Icon topIcon;
+    protected Icon topIcon;
+    // END FCMOD
 
     public BlockPistonBase(int par1, boolean par2)
     {
@@ -131,7 +143,10 @@
     /**
      * handles attempts to extend or retract the piston.
      */
-    private void updatePistonState(World par1World, int par2, int par3, int par4)
+    // FCMOD: Changed
+    //private void updatePistonState(World par1World, int par2, int par3, int par4)
+    protected void updatePistonState(World par1World, int par2, int par3, int par4)
+    // END FCMOD
     {
         int var5 = par1World.getBlockMetadata(par2, par3, par4);
         int var6 = getOrientation(var5);
@@ -234,8 +249,18 @@
                     }
                 }
 
-                if (!var13 && var11 > 0 && canPushBlock(var11, par1World, var8, var9, var10, false) && (Block.blocksList[var11].getMobilityFlag() == 0 || var11 == Block.pistonBase.blockID || var11 == Block.pistonStickyBase.blockID))
+                // FCMOD: Changed
+                //if (!var13 && var11 > 0 && canPushBlock(var11, par1World, var8, var9, var10, false) && (Block.blocksList[var11].getMobilityFlag() == 0 || var11 == Block.pistonBase.blockID || var11 == Block.pistonStickyBase.blockID))
+                Block targetBlock = Block.blocksList[var11];
+                
+                if ( !var13 && targetBlock != null && 
+                	targetBlock.CanBlockBePulledByPiston( par1World, var8, var9, var10, Block.GetOppositeFacing( par6 ) ) )
+                // END FCMOD                	
                 {
+                    // FCMOD: Added
+            		var12 =  targetBlock.AdjustMetadataForPistonMove( var12 );
+                    // END FCMOD
+                    
                     par2 += Facing.offsetsXForSide[par6];
                     par3 += Facing.offsetsYForSide[par6];
                     par4 += Facing.offsetsZForSide[par6];
@@ -379,6 +404,8 @@
     /**
      * returns true if the piston can push the specified block
      */
+    // FCMOD: Removed and replaced in child class
+    /* 
     private static boolean canPushBlock(int par0, World par1World, int par2, int par3, int par4, boolean par5)
     {
         if (par0 == Block.obsidian.blockID)
@@ -417,10 +444,14 @@
             return !(Block.blocksList[par0] instanceof ITileEntityProvider);
         }
     }
+    */
+    // END FCMOD
 
     /**
      * checks to see if this piston could push the blocks in front of it.
      */
+    // FCMOD: Removed and replaced in child class
+    /* 
     private static boolean canExtend(World par0World, int par1, int par2, int par3, int par4)
     {
         int var5 = par1 + Facing.offsetsXForSide[par4];
@@ -465,10 +496,14 @@
             return true;
         }
     }
+    */
+    // END FCMOD
 
     /**
      * attempts to extend the piston. returns false if impossible.
      */
+    // FCMOD: Removed and replaced in child class
+    /* 
     private boolean tryExtend(World par1World, int par2, int par3, int par4, int par5)
     {
         int var6 = par2 + Facing.offsetsXForSide[par5];
@@ -565,4 +600,22 @@
             return true;
         }
     }
+    */
+    // END FCMOD
+    
+    // FCMOD: Added New    
+    protected boolean canExtend( World world, int i, int j, int k, int iFacing )
+    {
+    	// intended to be overriden by child class
+    	
+    	return false;
 }
+    
+    protected boolean tryExtend( World world, int i, int j, int k, int iFacing )
+    {
+    	// intended to be overriden by child class
+    	
+    	return false;
+    }
+    // END FCMOD
+}
