--- a/minecraft/net/minecraft/src/BlockFurnace.java
+++ b/minecraft/net/minecraft/src/BlockFurnace.java
@@ -10,15 +10,31 @@
     private final Random furnaceRand = new Random();
 
     /** True if this is an active furnace, false if idle */
+    // FCMOD: Changed to protected
+    /*
     private final boolean isActive;
+    */
+    protected final boolean isActive;
+    // END FCMOD
 
+
     /**
      * This flag is used to prevent the furnace inventory to be dropped upon block removal, is used internally when the
      * furnace block changes from idle to active and vice-versa.
      */
-    private static boolean keepFurnaceInventory = false;
+    // FCMOD: Changed
+    //private static boolean keepFurnaceInventory = false;
+    protected static boolean keepFurnaceInventory = false;
+    // END FCMOD
+    
+    // FCMOD: Changed (client only)
+    /*
     private Icon furnaceIconTop;
     private Icon furnaceIconFront;
+    */
+    protected Icon furnaceIconTop;
+    protected Icon furnaceIconFront;
+    // END FCMOD
 
     protected BlockFurnace(int par1, boolean par2)
     {
@@ -107,6 +123,9 @@
         if (this.isActive)
         {
             int var6 = par1World.getBlockMetadata(par2, par3, par4);
+            //FCMOD: Added to filter out inventory state
+            var6 = var6 & 7;
+            // END FCMOD
             float var7 = (float)par2 + 0.5F;
             float var8 = (float)par3 + 0.0F + par5Random.nextFloat() * 6.0F / 16.0F;
             float var9 = (float)par4 + 0.5F;
@@ -161,7 +180,12 @@
     /**
      * Update which block ID the furnace is using depending on whether or not it is burning
      */
+    // FCMOD: Changed
+    /*
     public static void updateFurnaceBlockState(boolean par0, World par1World, int par2, int par3, int par4)
+    */
+    public void updateFurnaceBlockState(boolean par0, World par1World, int par2, int par3, int par4, boolean bHasContents)
+    // END FCMOD
     {
         int var5 = par1World.getBlockMetadata(par2, par3, par4);
         TileEntity var6 = par1World.getBlockTileEntity(par2, par3, par4);
@@ -177,6 +201,16 @@
         }
 
         keepFurnaceInventory = false;
+        // FCMOD: Code added
+        if ( !bHasContents )
+        {
+        	var5 = var5 & 7;
+        }
+        else
+        {
+        	var5 = var5 | 8;
+        }
+        // END FCMOD        
         par1World.setBlockMetadataWithNotify(par2, par3, par4, var5, 2);
 
         if (var6 != null)
@@ -258,7 +292,7 @@
                             }
 
                             var9.stackSize -= var13;
-                            EntityItem var14 = new EntityItem(par1World, (double)((float)par2 + var10), (double)((float)par3 + var11), (double)((float)par4 + var12), new ItemStack(var9.itemID, var13, var9.getItemDamage()));
+                            EntityItem var14 = (EntityItem) EntityList.createEntityOfType(EntityItem.class, par1World, (double)((float)par2 + var10), (double)((float)par3 + var11), (double)((float)par4 + var12), new ItemStack(var9.itemID, var13, var9.getItemDamage()));
 
                             if (var9.hasTagCompound())
                             {
