--- a/minecraft/net/minecraft/src/BlockTallGrass.java
+++ b/minecraft/net/minecraft/src/BlockTallGrass.java
@@ -1,5 +1,6 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.cc.ColorizeBlock;
 import java.util.List;
 import java.util.Random;
 
@@ -30,17 +31,24 @@
 
     public int getBlockColor()
     {
+    	if (ColorizeBlock.colorizeBlock(this))
+        {
+            return ColorizeBlock.blockColor;
+        }
+        else
+        {
         double var1 = 0.5D;
         double var3 = 1.0D;
         return ColorizerGrass.getGrassColor(var1, var3);
     }
+    }
 
     /**
      * Returns the color this block should be rendered. Used by leaves.
      */
     public int getRenderColor(int par1)
     {
-        return par1 == 0 ? 16777215 : ColorizerFoliage.getFoliageColorBasic();
+    	return ColorizeBlock.colorizeBlock(this, par1) ? ColorizeBlock.blockColor : (par1 == 0 ? 16777215 : ColorizerFoliage.getFoliageColorBasic());
     }
 
     /**
@@ -49,9 +57,16 @@
      */
     public int colorMultiplier(IBlockAccess par1IBlockAccess, int par2, int par3, int par4)
     {
+    	if (ColorizeBlock.colorizeBlock(this, par1IBlockAccess, par2, par3, par4))
+        {
+            return ColorizeBlock.blockColor;
+        }
+        else
+        {
         int var5 = par1IBlockAccess.getBlockMetadata(par2, par3, par4);
         return var5 == 0 ? 16777215 : par1IBlockAccess.getBiomeGenForCoords(par2, par4).getBiomeGrassColor();
     }
+    }
 
     /**
      * Returns the ID of the items to drop on destruction.
