--- a/minecraft/net/minecraft/src/BlockCauldron.java
+++ b/minecraft/net/minecraft/src/BlockCauldron.java
@@ -137,7 +137,7 @@
 
                             if (!par5EntityPlayer.inventory.addItemStackToInventory(var12))
                             {
-                                par1World.spawnEntityInWorld(new EntityItem(par1World, (double)par2 + 0.5D, (double)par3 + 1.5D, (double)par4 + 0.5D, var12));
+                                par1World.spawnEntityInWorld(EntityList.createEntityOfType(EntityItem.class, par1World, (double)par2 + 0.5D, (double)par3 + 1.5D, (double)par4 + 0.5D, var12));
                             }
                             else if (par5EntityPlayer instanceof EntityPlayerMP)
                             {
