--- a/minecraft/net/minecraft/src/BlockChest.java
+++ b/minecraft/net/minecraft/src/BlockChest.java
@@ -370,7 +370,7 @@
                         }
 
                         var9.stackSize -= var13;
-                        var14 = new EntityItem(par1World, (double)((float)par2 + var10), (double)((float)par3 + var11), (double)((float)par4 + var12), new ItemStack(var9.itemID, var13, var9.getItemDamage()));
+                        var14 = (EntityItem) EntityList.createEntityOfType(EntityItem.class, par1World, (double)((float)par2 + var10), (double)((float)par3 + var11), (double)((float)par4 + var12), new ItemStack(var9.itemID, var13, var9.getItemDamage()));
                         float var15 = 0.05F;
                         var14.motionX = (double)((float)this.random.nextGaussian() * var15);
                         var14.motionY = (double)((float)this.random.nextGaussian() * var15 + 0.2F);
