--- a/minecraft/net/minecraft/src/BlockDragonEgg.java
+++ b/minecraft/net/minecraft/src/BlockDragonEgg.java
@@ -46,7 +46,7 @@
 
             if (!BlockSand.fallInstantly && par1World.checkChunksExist(par2 - var5, par3 - var5, par4 - var5, par2 + var5, par3 + var5, par4 + var5))
             {
-                EntityFallingSand var6 = new EntityFallingSand(par1World, (double)((float)par2 + 0.5F), (double)((float)par3 + 0.5F), (double)((float)par4 + 0.5F), this.blockID);
+                EntityFallingSand var6 = (EntityFallingSand) EntityList.createEntityOfType(EntityFallingSand.class, par1World, (double)((float)par2 + 0.5F), (double)((float)par3 + 0.5F), (double)((float)par4 + 0.5F), this.blockID);
                 par1World.spawnEntityInWorld(var6);
             }
             else
