--- a/minecraft/net/minecraft/src/ComponentMineshaftCorridor.java
+++ b/minecraft/net/minecraft/src/ComponentMineshaftCorridor.java
@@ -197,8 +197,11 @@
         if (par2StructureBoundingBox.isVecInside(var9, var10, var11) && par1World.getBlockId(var9, var10, var11) == 0)
         {
             par1World.setBlock(var9, var10, var11, Block.rail.blockID, this.getMetadataWithOffset(Block.rail.blockID, par3Random.nextBoolean() ? 1 : 0), 2);
-            EntityMinecartChest var12 = new EntityMinecartChest(par1World, (double)((float)var9 + 0.5F), (double)((float)var10 + 0.5F), (double)((float)var11 + 0.5F));
+            EntityMinecartChest var12 = (EntityMinecartChest) EntityList.createEntityOfType(EntityMinecartChest.class, par1World, (double)((float)var9 + 0.5F), (double)((float)var10 + 0.5F), (double)((float)var11 + 0.5F));
             WeightedRandomChestContent.generateChestContents(par3Random, par7ArrayOfWeightedRandomChestContent, var12, par8);
+            // FCMOD: Added
+            FilterChestMinecartContents( var12 );
+            // END FCMOD
             par1World.spawnEntityInWorld(var12);
             return true;
         }
@@ -226,7 +229,11 @@
 
             if (this.hasSpiders)
             {
-                this.randomlyFillWithBlocks(par1World, par3StructureBoundingBox, par2Random, 0.6F, 0, 0, 0, 2, 1, var8, Block.web.blockID, 0, false);
+            	// FCMOD: Changed
+                //this.randomlyFillWithBlocks(par1World, par3StructureBoundingBox, par2Random, 0.6F, 0, 0, 0, 2, 1, var8, Block.web.blockID, 0, false);
+                this.randomlyFillWithBlocks(par1World, par3StructureBoundingBox, par2Random, 0.6F, 0, 0, 0, 2, 1, var8, 
+                	FCBetterThanWolves.fcBlockWeb.blockID, 0, false);
+                // END FCMOD            	
             }
 
             int var9;
@@ -236,19 +243,42 @@
             for (var9 = 0; var9 < this.sectionCount; ++var9)
             {
                 var10 = 2 + var9 * 5;
+                // FCMOD: Code change to replace fences by logs
+                /*
                 this.fillWithBlocks(par1World, par3StructureBoundingBox, 0, 0, var10, 0, 1, var10, Block.fence.blockID, 0, false);
                 this.fillWithBlocks(par1World, par3StructureBoundingBox, 2, 0, var10, 2, 1, var10, Block.fence.blockID, 0, false);
+                */
+                this.fillWithMetadataBlocks(par1World, par3StructureBoundingBox, 0, 0, var10, 0, 1, var10, Block.wood.blockID, 0, 0, 0, false);
+                this.fillWithMetadataBlocks(par1World, par3StructureBoundingBox, 2, 0, var10, 2, 1, var10, Block.wood.blockID, 0, 0, 0, false);
+                // END FCMOD
 
+            	// FCMOD: Code added
+                int iHorizontalLogMetadata = getMetadataWithOffset( Block.wood.blockID, 4 );
+                // END FCMOD
+                
                 if (par2Random.nextInt(4) == 0)
                 {
+                	// FCMOD: Code change
+                	/*
                     this.fillWithBlocks(par1World, par3StructureBoundingBox, 0, 2, var10, 0, 2, var10, Block.planks.blockID, 0, false);
                     this.fillWithBlocks(par1World, par3StructureBoundingBox, 2, 2, var10, 2, 2, var10, Block.planks.blockID, 0, false);
+                    */
+                    this.fillWithMetadataBlocks(par1World, par3StructureBoundingBox, 0, 2, var10, 0, 2, var10, Block.wood.blockID, iHorizontalLogMetadata, 0, 0, false);
+                    this.fillWithMetadataBlocks(par1World, par3StructureBoundingBox, 2, 2, var10, 2, 2, var10, Block.wood.blockID, iHorizontalLogMetadata, 0, 0, false);
+                	// END FCMOD
                 }
                 else
                 {
+                	// FCMOD: Code change
+                	/*
                     this.fillWithBlocks(par1World, par3StructureBoundingBox, 0, 2, var10, 2, 2, var10, Block.planks.blockID, 0, false);
+                    */
+                    this.fillWithMetadataBlocks(par1World, par3StructureBoundingBox, 0, 2, var10, 2, 2, var10, Block.wood.blockID, iHorizontalLogMetadata, 0, 0, false);
+                	// END FCMOD
                 }
 
+                // FCMOD: Changed
+                /*
                 this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.1F, 0, 2, var10 - 1, Block.web.blockID, 0);
                 this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.1F, 2, 2, var10 - 1, Block.web.blockID, 0);
                 this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.1F, 0, 2, var10 + 1, Block.web.blockID, 0);
@@ -259,6 +289,18 @@
                 this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.05F, 2, 2, var10 + 2, Block.web.blockID, 0);
                 this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.05F, 1, 2, var10 - 1, Block.torchWood.blockID, 0);
                 this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.05F, 1, 2, var10 + 1, Block.torchWood.blockID, 0);
+                */
+                this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.1F, 0, 2, var10 - 1, FCBetterThanWolves.fcBlockWeb.blockID, 0);
+                this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.1F, 2, 2, var10 - 1, FCBetterThanWolves.fcBlockWeb.blockID, 0);
+                this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.1F, 0, 2, var10 + 1, FCBetterThanWolves.fcBlockWeb.blockID, 0);
+                this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.1F, 2, 2, var10 + 1, FCBetterThanWolves.fcBlockWeb.blockID, 0);
+                this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.05F, 0, 2, var10 - 2, FCBetterThanWolves.fcBlockWeb.blockID, 0);
+                this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.05F, 2, 2, var10 - 2, FCBetterThanWolves.fcBlockWeb.blockID, 0);
+                this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.05F, 0, 2, var10 + 2, FCBetterThanWolves.fcBlockWeb.blockID, 0);
+                this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.05F, 2, 2, var10 + 2, FCBetterThanWolves.fcBlockWeb.blockID, 0);
+                this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.05F, 1, 2, var10 - 1, FCBetterThanWolves.fcBlockTorchFiniteUnlit.blockID, 8);
+                this.randomlyPlaceBlock(par1World, par3StructureBoundingBox, par2Random, 0.05F, 1, 2, var10 + 1, FCBetterThanWolves.fcBlockTorchFiniteUnlit.blockID, 8);                
+                // END FCMOD
 
                 if (par2Random.nextInt(100) == 0)
                 {
@@ -320,4 +362,89 @@
             return true;
         }
     }
+    
+    // FCMOD: Added
+    private void FilterChestMinecartContents( EntityMinecartChest minecart )
+    {
+        for ( int iSlot = 0; iSlot < minecart.getSizeInventory(); iSlot++ )
+        {
+            ItemStack tempStack = minecart.getStackInSlot( iSlot );
+            
+            if ( tempStack != null )
+            {
+            	int iItemID = tempStack.itemID;
+
+            	if ( iItemID == Item.ingotIron.itemID )
+            	{
+            		if ( minecart.posY > 36F )
+            		{
+                		minecart.setInventorySlotContents( iSlot, null );
 }
+            		else
+            		{
+            			tempStack.stackSize = 1;
+            		}
+            	}
+            	else if ( iItemID == Item.diamond.itemID )
+            	{
+            		if ( minecart.posY > 24F )
+            		{
+                		minecart.setInventorySlotContents( iSlot, null );
+            		}
+            		else
+            		{
+            			tempStack.stackSize = 1;
+            		}
+            	}
+            	else if ( iItemID == Item.pickaxeIron.itemID )
+            	{
+            		minecart.setInventorySlotContents( iSlot, null );
+            	}
+            	else if ( iItemID == Item.redstone.itemID )
+            	{
+            		if ( minecart.posY > 24F )
+            		{
+                		minecart.setInventorySlotContents( iSlot, null );
+            		}
+            	}
+            	else if ( iItemID == Item.pumpkinSeeds.itemID )
+            	{
+            		minecart.setInventorySlotContents( iSlot, null );
+            	}            	
+            }
+        }
+    }
+
+    @Override
+    protected int getMetadataWithOffset( int iBlockID, int iMetadata )
+    {
+    	// override to handle horizontal log placement in supports
+    	
+    	if ( iBlockID == Block.wood.blockID )
+    	{
+            if (this.coordBaseMode == 1 || this.coordBaseMode == 3)
+            {
+            	int iLogType = iMetadata & 3;
+            	int iDirection = iMetadata & 12;
+            	
+            	if ( iDirection == 4 )
+            	{
+            		iDirection = 8;
+            	}
+            	else if ( iDirection == 8 )
+            	{
+            		iDirection = 4;
+            	}
+            	
+            	iMetadata = iLogType | iDirection;
+            }
+            
+    		return iMetadata;
+    	}
+    	else
+    	{
+    		return super.getMetadataWithOffset( iBlockID, iMetadata );
+    	}
+    }    
+    // END FCMOD
+}
