--- a/minecraft/net/minecraft/src/ComponentStrongholdRoomCrossing.java
+++ b/minecraft/net/minecraft/src/ComponentStrongholdRoomCrossing.java
@@ -139,9 +139,19 @@
                         this.placeBlockAtCurrentPosition(par1World, Block.planks.blockID, 0, 8, 3, var4, par3StructureBoundingBox);
                     }
 
+                    // FCMOD: Changed
+                    /*
                     this.placeBlockAtCurrentPosition(par1World, Block.ladder.blockID, this.getMetadataWithOffset(Block.ladder.blockID, 4), 9, 1, 3, par3StructureBoundingBox);
                     this.placeBlockAtCurrentPosition(par1World, Block.ladder.blockID, this.getMetadataWithOffset(Block.ladder.blockID, 4), 9, 2, 3, par3StructureBoundingBox);
                     this.placeBlockAtCurrentPosition(par1World, Block.ladder.blockID, this.getMetadataWithOffset(Block.ladder.blockID, 4), 9, 3, 3, par3StructureBoundingBox);
+	                */
+	                int iFacing = getMetadataWithOffset( Block.ladder.blockID, 4 ); // legacy ladder metadata is straight facing
+	                int iMetadata = FCBetterThanWolves.fcBlockLadder.SetFacing( 0, iFacing );                
+	                
+                    placeBlockAtCurrentPosition( par1World, FCBetterThanWolves.fcBlockLadder.blockID, iMetadata, 9, 1, 3, par3StructureBoundingBox );
+                    placeBlockAtCurrentPosition( par1World, FCBetterThanWolves.fcBlockLadder.blockID, iMetadata, 9, 2, 3, par3StructureBoundingBox );
+                    placeBlockAtCurrentPosition( par1World, FCBetterThanWolves.fcBlockLadder.blockID, iMetadata, 9, 3, 3, par3StructureBoundingBox );
+	                // END FCMOD
                     this.generateStructureChestContents(par1World, par3StructureBoundingBox, par2Random, 3, 4, 8, WeightedRandomChestContent.func_92080_a(strongholdRoomCrossingChestContents, new WeightedRandomChestContent[] {Item.enchantedBook.func_92114_b(par2Random)}), 1 + par2Random.nextInt(4));
             }
 
