--- a/minecraft/net/minecraft/src/ContainerEnchantment.java
+++ b/minecraft/net/minecraft/src/ContainerEnchantment.java
@@ -202,6 +202,10 @@
                     }
 
                     this.onCraftMatrixChanged(this.tableInventory);
+
+                    // FCMOD: Code added
+                    worldPointer.playSoundAtEntity( par1EntityPlayer, "random.levelup", 0.25F, worldPointer.rand.nextFloat() * 0.1F + 0.5F );
+                    // END FCMOD
                 }
             }
 
