--- a/minecraft/net/minecraft/src/DispenserBehaviorArrow.java
+++ b/minecraft/net/minecraft/src/DispenserBehaviorArrow.java
@@ -7,7 +7,7 @@
      */
     protected IProjectile getProjectileEntity(World par1World, IPosition par2IPosition)
     {
-        EntityArrow var3 = new EntityArrow(par1World, par2IPosition.getX(), par2IPosition.getY(), par2IPosition.getZ());
+        EntityArrow var3 = (EntityArrow) EntityList.createEntityOfType(EntityArrow.class, par1World, par2IPosition.getX(), par2IPosition.getY(), par2IPosition.getZ());
         var3.canBePickedUp = 1;
         return var3;
     }
