--- a/minecraft/net/minecraft/src/DispenserBehaviorTNT.java
+++ b/minecraft/net/minecraft/src/DispenserBehaviorTNT.java
@@ -5,6 +5,8 @@
     /**
      * Dispense the specified stack, play the dispense sound and spawn particles.
      */
+	// FCMOD: Code removed to prevent usage
+	/*
     protected ItemStack dispenseStack(IBlockSource par1IBlockSource, ItemStack par2ItemStack)
     {
         EnumFacing var3 = BlockDispenser.getFacing(par1IBlockSource.getBlockMetadata());
@@ -17,4 +19,6 @@
         --par2ItemStack.stackSize;
         return par2ItemStack;
     }
+    */
+    // END FCMOD
 }
