--- a/minecraft/net/minecraft/src/DispenserBehaviorFireworks.java
+++ b/minecraft/net/minecraft/src/DispenserBehaviorFireworks.java
@@ -11,7 +11,7 @@
         double var4 = par1IBlockSource.getX() + (double)var3.getFrontOffsetX();
         double var6 = (double)((float)par1IBlockSource.getYInt() + 0.2F);
         double var8 = par1IBlockSource.getZ() + (double)var3.getFrontOffsetZ();
-        EntityFireworkRocket var10 = new EntityFireworkRocket(par1IBlockSource.getWorld(), var4, var6, var8, par2ItemStack);
+        EntityFireworkRocket var10 = (EntityFireworkRocket) EntityList.createEntityOfType(EntityFireworkRocket.class, par1IBlockSource.getWorld(), var4, var6, var8, par2ItemStack);
         par1IBlockSource.getWorld().spawnEntityInWorld(var10);
         par2ItemStack.splitStack(1);
         return par2ItemStack;
