--- a/minecraft/net/minecraft/src/DispenserBehaviorEgg.java
+++ b/minecraft/net/minecraft/src/DispenserBehaviorEgg.java
@@ -7,6 +7,6 @@
      */
     protected IProjectile getProjectileEntity(World par1World, IPosition par2IPosition)
     {
-        return new EntityEgg(par1World, par2IPosition.getX(), par2IPosition.getY(), par2IPosition.getZ());
+        return (IProjectile) EntityList.createEntityOfType(EntityEgg.class, par1World, par2IPosition.getX(), par2IPosition.getY(), par2IPosition.getZ());
     }
 }
