--- a/minecraft/net/minecraft/src/DispenserBehaviorFire.java
+++ b/minecraft/net/minecraft/src/DispenserBehaviorFire.java
@@ -7,6 +7,8 @@
     /**
      * Dispense the specified stack, play the dispense sound and spawn particles.
      */
+    // FCMOD: Code removed to prevent usage
+    /*
     protected ItemStack dispenseStack(IBlockSource par1IBlockSource, ItemStack par2ItemStack)
     {
         EnumFacing var3 = BlockDispenser.getFacing(par1IBlockSource.getBlockMetadata());
@@ -36,10 +38,14 @@
 
         return par2ItemStack;
     }
+    */
+    // END FCMOD
 
     /**
      * Play the dispense sound from the specified block.
      */
+    // FCMOD: Code removed
+    /*
     protected void playDispenseSound(IBlockSource par1IBlockSource)
     {
         if (this.field_96466_b)
@@ -51,4 +57,6 @@
             par1IBlockSource.getWorld().playAuxSFX(1001, par1IBlockSource.getXInt(), par1IBlockSource.getYInt(), par1IBlockSource.getZInt(), 0);
         }
     }
+    */
+    // END FCMOD
 }
