--- a/minecraft/net/minecraft/src/DispenserBehaviorFilledBucket.java
+++ b/minecraft/net/minecraft/src/DispenserBehaviorFilledBucket.java
@@ -7,6 +7,8 @@
     /**
      * Dispense the specified stack, play the dispense sound and spawn particles.
      */
+    // FCMOD: Code removed to prevent usage
+    /*
     public ItemStack dispenseStack(IBlockSource par1IBlockSource, ItemStack par2ItemStack)
     {
         ItemBucket var3 = (ItemBucket)par2ItemStack.getItem();
@@ -26,4 +28,6 @@
             return this.defaultDispenserItemBehavior.dispense(par1IBlockSource, par2ItemStack);
         }
     }
+    */
+    // END FCMOD
 }
