--- a/minecraft/net/minecraft/src/EntityWither.java
+++ b/minecraft/net/minecraft/src/EntityWither.java
@@ -248,7 +248,11 @@
                 {
                     this.field_82223_h[var1 - 1] = this.ticksExisted + 10 + this.rand.nextInt(10);
 
+                    // FCMOD: Code removed so that skulls launch on easy difficulty
+                    /*
                     if (this.worldObj.difficultySetting >= 2)
+                    */
+                    // END FCMOD
                     {
                         int var10001 = var1 - 1;
                         int var10003 = this.field_82224_i[var1 - 1];
@@ -332,28 +336,34 @@
                     var1 = MathHelper.floor_double(this.posY);
                     var12 = MathHelper.floor_double(this.posX);
                     int var15 = MathHelper.floor_double(this.posZ);
-                    boolean var17 = false;
+                    boolean var18 = false;
 
-                    for (int var18 = -1; var18 <= 1; ++var18)
+                    for (int var17 = -1; var17 <= 1; ++var17)
                     {
                         for (int var19 = -1; var19 <= 1; ++var19)
                         {
                             for (int var7 = 0; var7 <= 3; ++var7)
                             {
-                                int var20 = var12 + var18;
+                                int var20 = var12 + var17;
                                 int var9 = var1 + var7;
                                 int var10 = var15 + var19;
                                 int var11 = this.worldObj.getBlockId(var20, var9, var10);
 
+                                // FCMOD: Change
+                                /*
                                 if (var11 > 0 && var11 != Block.bedrock.blockID && var11 != Block.endPortal.blockID && var11 != Block.endPortalFrame.blockID)
+                                */                                	
+                                if (var11 > 0 && var11 != Block.bedrock.blockID && var11 != Block.endPortal.blockID && var11 != Block.endPortalFrame.blockID &&
+                                	var11 != FCBetterThanWolves.fcSoulforgedSteelBlock.blockID )
+                            	// END FCMOD
                                 {
-                                    var17 = this.worldObj.destroyBlock(var20, var9, var10, true) || var17;
+                                    var18 = this.worldObj.destroyBlock(var20, var9, var10, true) || var18;
                                 }
                             }
                         }
                     }
 
-                    if (var17)
+                    if (var18)
                     {
                         this.worldObj.playAuxSFXAtEntity((EntityPlayer)null, 1012, (int)this.posX, (int)this.posY, (int)this.posZ, 0);
                     }
@@ -450,7 +460,10 @@
         double var15 = par2 - var9;
         double var17 = par4 - var11;
         double var19 = par6 - var13;
-        EntityWitherSkull var21 = new EntityWitherSkull(this.worldObj, this, var15, var17, var19);
+        // FCMOD: Changed
+        //EntityWitherSkull var21 = new EntityWitherSkull(this.worldObj, this, var15, var17, var19);
+        FCEntityWitherSkull var21 = (FCEntityWitherSkull) EntityList.createEntityOfType(FCEntityWitherSkull.class, this.worldObj, this, var15, var17, var19);
+        // END FCMOD
 
         if (par8)
         {
