--- a/minecraft/net/minecraft/src/EntitySplashFX.java
+++ b/minecraft/net/minecraft/src/EntitySplashFX.java
@@ -1,5 +1,8 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.cc.ColorizeBlock;
+import com.prupe.mcpatcher.cc.Colorizer;
+
 public class EntitySplashFX extends EntityRainFX
 {
     public EntitySplashFX(World par1World, double par2, double par4, double par6, double par8, double par10, double par12)
@@ -14,5 +17,12 @@
             this.motionY = par10 + 0.1D;
             this.motionZ = par12;
         }
+        
+        if (ColorizeBlock.computeWaterColor(false, (int)this.posX, (int)this.posY, (int)this.posZ))
+        {
+            this.particleRed = Colorizer.setColor[0];
+            this.particleGreen = Colorizer.setColor[1];
+            this.particleBlue = Colorizer.setColor[2];
     }
 }
+}
