--- a/minecraft/net/minecraft/src/EntitySnowman.java
+++ b/minecraft/net/minecraft/src/EntitySnowman.java
@@ -1,3 +1,5 @@
+// FCMOD: Class changes deprecated 12/27/2018
+
 package net.minecraft.src;
 
 public class EntitySnowman extends EntityGolem implements IRangedAttackMob
@@ -89,7 +91,7 @@
      */
     public void attackEntityWithRangedAttack(EntityLiving par1EntityLiving, float par2)
     {
-        EntitySnowball var3 = new EntitySnowball(this.worldObj, this);
+        EntitySnowball var3 = (EntitySnowball) EntityList.createEntityOfType(EntitySnowball.class, this.worldObj, this);
         double var4 = par1EntityLiving.posX - this.posX;
         double var6 = par1EntityLiving.posY + (double)par1EntityLiving.getEyeHeight() - 1.100000023841858D - var3.posY;
         double var8 = par1EntityLiving.posZ - this.posZ;
