--- a/minecraft/net/minecraft/src/EntityChicken.java
+++ b/minecraft/net/minecraft/src/EntityChicken.java
@@ -1,3 +1,5 @@
+// FCMOD: Class changes deprecated 08/10/2018
+
 package net.minecraft.src;
 
 public class EntityChicken extends EntityAnimal
@@ -158,7 +160,7 @@
      */
     public EntityChicken spawnBabyAnimal(EntityAgeable par1EntityAgeable)
     {
-        return new EntityChicken(this.worldObj);
+        return (EntityChicken) EntityList.createEntityOfType(EntityChicken.class, this.worldObj);
     }
 
     /**
