--- a/minecraft/net/minecraft/src/EntityAIOpenDoor.java
+++ b/minecraft/net/minecraft/src/EntityAIOpenDoor.java
@@ -26,7 +26,12 @@
     public void startExecuting()
     {
         this.field_75360_j = 20;
+        // FCMOD: Code change
+        /*
         this.targetDoor.onPoweredBlockChange(this.theEntity.worldObj, this.entityPosX, this.entityPosY, this.entityPosZ, true);
+        */        
+        this.targetDoor.OnAIOpenDoor(this.theEntity.worldObj, this.entityPosX, this.entityPosY, this.entityPosZ, true);
+        // END FCMOD
     }
 
     /**
@@ -36,7 +41,12 @@
     {
         if (this.field_75361_i)
         {
+            // FCMOD: Code change
+            /*
             this.targetDoor.onPoweredBlockChange(this.theEntity.worldObj, this.entityPosX, this.entityPosY, this.entityPosZ, false);
+            */
+            this.targetDoor.OnAIOpenDoor(this.theEntity.worldObj, this.entityPosX, this.entityPosY, this.entityPosZ, false);
+        	// END FCMOD
         }
     }
 
