--- a/minecraft/net/minecraft/src/EntityHanging.java
+++ b/minecraft/net/minecraft/src/EntityHanging.java
@@ -198,7 +198,12 @@
 
                 var11 = (Entity)var10.next();
             }
+        	// FCMOD: Code change so that paintings and canvas can't overlap
+            /*
             while (!(var11 instanceof EntityHanging));
+            */
+            while (!(var11 instanceof EntityHanging) && !(var11 instanceof FCEntityCanvas));
+            // END FCMOD
 
             return false;
         }
@@ -348,4 +353,12 @@
      * Drop the item currently on this item frame.
      */
     public abstract void dropItemStack();
+    
+    // FCMOD: Code added
+    @Override
+    protected boolean ShouldSetPositionOnLoad()
+    {
+    	return false;
 }
+    // END FCMOD
+}
