--- a/minecraft/net/minecraft/src/EntityAIAttackOnCollide.java
+++ b/minecraft/net/minecraft/src/EntityAIAttackOnCollide.java
@@ -80,6 +80,12 @@
      */
     public void resetTask()
     {
+    	// FCMOD: Added
+    	if ( attacker.getAttackTarget() == entityTarget )
+    	{
+    		attacker.setAttackTarget( null );
+    	}
+    	// END FCMOD
         this.entityTarget = null;
         this.attacker.getNavigator().clearPathEntity();
     }
@@ -98,8 +104,21 @@
         }
 
         this.attackTick = Math.max(this.attackTick - 1, 0);
+        // FCMOD: Code change to take width of target into account as the original code assumes it's the same as the attacker
+        /*
         double var1 = (double)(this.attacker.width * 2.0F * this.attacker.width * 2.0F);
+        */
+        double dCombinedWidth = attacker.width + entityTarget.width;
+        double var1 = dCombinedWidth * dCombinedWidth;
 
+        if ( entityTarget == attacker.riddenByEntity )
+        {
+        	// stop stuff like wolves biting squid on their backs
+        	
+        	return;
+        }
+        // END FCMOD
+
         if (this.attacker.getDistanceSq(this.entityTarget.posX, this.entityTarget.boundingBox.minY, this.entityTarget.posZ) <= var1)
         {
             if (this.attackTick <= 0)
