--- a/minecraft/net/minecraft/src/EntityGhast.java
+++ b/minecraft/net/minecraft/src/EntityGhast.java
@@ -1,3 +1,5 @@
+// FCMOD: Class changes deprecated 06/15/2018
+
 package net.minecraft.src;
 
 public class EntityGhast extends EntityFlying implements IMob
@@ -143,7 +145,7 @@
                 if (this.attackCounter == 20)
                 {
                     this.worldObj.playAuxSFXAtEntity((EntityPlayer)null, 1008, (int)this.posX, (int)this.posY, (int)this.posZ, 0);
-                    EntityLargeFireball var17 = new EntityLargeFireball(this.worldObj, this, var11, var13, var15);
+                    EntityLargeFireball var17 = (EntityLargeFireball) EntityList.createEntityOfType(EntityLargeFireball.class, this.worldObj, this, var11, var13, var15);
                     var17.field_92057_e = this.explosionStrength;
                     double var18 = 4.0D;
                     Vec3 var20 = this.getLook(1.0F);
