--- a/minecraft/net/minecraft/src/EntityAIDoorInteract.java
+++ b/minecraft/net/minecraft/src/EntityAIDoorInteract.java
@@ -106,6 +106,10 @@
     private BlockDoor findUsableDoor(int par1, int par2, int par3)
     {
         int var4 = this.theEntity.worldObj.getBlockId(par1, par2, par3);
-        return var4 != Block.doorWood.blockID ? null : (BlockDoor)Block.blocksList[var4];
+        // FCMOD: Changed
+        //return var4 != Block.doorWood.blockID ? null : (BlockDoor)Block.blocksList[var4];
+        return ( var4 != Block.doorWood.blockID && var4 != FCBetterThanWolves.fcBlockDoorWood.blockID ) ? 
+        	null : (BlockDoor)Block.blocksList[var4];
+    	// END FCMOD
     }
 }
