--- a/minecraft/net/minecraft/src/EntityPigZombie.java
+++ b/minecraft/net/minecraft/src/EntityPigZombie.java
@@ -2,7 +2,10 @@
 
 import java.util.List;
 
-public class EntityPigZombie extends EntityZombie
+// FCMOD: Changed
+//public class EntityPigZombie extends EntityZombie
+public class EntityPigZombie extends FCEntityZombie
+// END FCMOD
 {
     /** Above zero if this PigZombie is Angry. */
     private int angerLevel = 0;
@@ -122,7 +125,10 @@
     /**
      * Causes this PigZombie to become angry at the supplied Entity (which will be a player).
      */
-    private void becomeAngryAt(Entity par1Entity)
+    // FCMOD: Changed
+    //private void becomeAngryAt(Entity par1Entity)
+    protected void becomeAngryAt(Entity par1Entity)
+    // END FCMOD
     {
         this.entityToAttack = par1Entity;
         this.angerLevel = 400 + this.rand.nextInt(400);
