--- a/minecraft/net/minecraft/src/EntitySheep.java
+++ b/minecraft/net/minecraft/src/EntitySheep.java
@@ -1,5 +1,9 @@
+// FCMOD: Class changes deprecated 07/23/2018
+
 package net.minecraft.src;
 
+import java.util.Iterator;
+import java.util.List;
 import java.util.Random;
 
 public class EntitySheep extends EntityAnimal
@@ -259,7 +263,7 @@
     public EntitySheep func_90015_b(EntityAgeable par1EntityAgeable)
     {
         EntitySheep var2 = (EntitySheep)par1EntityAgeable;
-        EntitySheep var3 = new EntitySheep(this.worldObj);
+        EntitySheep var3 = (EntitySheep) EntityList.createEntityOfType(EntitySheep.class, this.worldObj);
         int var4 = this.func_90014_a(this, var2);
         var3.setFleeceColor(15 - var4);
         return var3;
