--- a/minecraft/net/minecraft/src/EntityRainFX.java
+++ b/minecraft/net/minecraft/src/EntityRainFX.java
@@ -1,5 +1,8 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.cc.ColorizeBlock;
+import com.prupe.mcpatcher.cc.Colorizer;
+
 public class EntityRainFX extends EntityFX
 {
     public EntityRainFX(World par1World, double par2, double par4, double par6)
@@ -8,9 +11,20 @@
         this.motionX *= 0.30000001192092896D;
         this.motionY = (double)((float)Math.random() * 0.2F + 0.1F);
         this.motionZ *= 0.30000001192092896D;
-        this.particleRed = 1.0F;
-        this.particleGreen = 1.0F;
+        
+        if (ColorizeBlock.computeWaterColor(false, (int)this.posX, (int)this.posY, (int)this.posZ))
+        {
+            this.particleRed = Colorizer.setColor[0];
+            this.particleGreen = Colorizer.setColor[1];
+            this.particleBlue = Colorizer.setColor[2];
+        }
+        else
+        {
+            this.particleRed = 0.2F;
+            this.particleGreen = 0.3F;
         this.particleBlue = 1.0F;
+        }
+        
         this.setParticleTextureIndex(19 + this.rand.nextInt(4));
         this.setSize(0.01F, 0.01F);
         this.particleGravity = 0.06F;
