--- a/minecraft/net/minecraft/src/EntityWolf.java
+++ b/minecraft/net/minecraft/src/EntityWolf.java
@@ -1,3 +1,5 @@
+// FCMOD: Class changes deprecated 09/10/2018
+
 package net.minecraft.src;
 
 public class EntityWolf extends EntityTameable
@@ -501,7 +503,7 @@
      */
     public EntityWolf spawnBabyAnimal(EntityAgeable par1EntityAgeable)
     {
-        EntityWolf var2 = new EntityWolf(this.worldObj);
+        EntityWolf var2 = (EntityWolf) EntityList.createEntityOfType(EntityWolf.class, this.worldObj);
         String var3 = this.getOwnerName();
 
         if (var3 != null && var3.trim().length() > 0)
