--- a/minecraft/net/minecraft/src/EntityAIWander.java
+++ b/minecraft/net/minecraft/src/EntityAIWander.java
@@ -20,11 +20,20 @@
      */
     public boolean shouldExecute()
     {
+    	// FCMOD: Code removed to disable AI activation distance
+    	/*
         if (this.entity.getAge() >= 100)
         {
             return false;
         }
+        */
+    	// END FCMOD
+    	
+    	// FCMOD: Line change to get rid of else
+    	/*
         else if (this.entity.getRNG().nextInt(120) != 0)
+        */
+        if (this.entity.getRNG().nextInt(120) != 0)
         {
             return false;
         }
