--- a/minecraft/net/minecraft/src/EntityReddustFX.java
+++ b/minecraft/net/minecraft/src/EntityReddustFX.java
@@ -1,5 +1,8 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.cc.ColorizeBlock;
+import com.prupe.mcpatcher.cc.Colorizer;
+
 public class EntityReddustFX extends EntityFX
 {
     float reddustParticleScale;
@@ -19,7 +22,14 @@
         if (par9 == 0.0F)
         {
             par9 = 1.0F;
+            
+            if (ColorizeBlock.computeRedstoneWireColor(15))
+            {
+                par9 = Colorizer.setColor[0];
+                par10 = Colorizer.setColor[1];
+                par11 = Colorizer.setColor[2];
         }
+        }
 
         float var12 = (float)Math.random() * 0.4F + 0.6F;
         this.particleRed = ((float)(Math.random() * 0.20000000298023224D) + 0.8F) * par9 * var12;
