--- a/minecraft/net/minecraft/src/EntitySkeleton.java
+++ b/minecraft/net/minecraft/src/EntitySkeleton.java
@@ -1,3 +1,5 @@
+// FCMOD: Class changes deprecated 09/26/2018
+
 package net.minecraft.src;
 
 import java.util.Calendar;
@@ -317,7 +319,7 @@
      */
     public void attackEntityWithRangedAttack(EntityLiving par1EntityLiving, float par2)
     {
-        EntityArrow var3 = new EntityArrow(this.worldObj, this, par1EntityLiving, 1.6F, (float)(14 - this.worldObj.difficultySetting * 4));
+        EntityArrow var3 = (EntityArrow) EntityList.createEntityOfType(EntityArrow.class, this.worldObj, this, par1EntityLiving, 1.6F, (float)(14 - this.worldObj.difficultySetting * 4));
         int var4 = EnchantmentHelper.getEnchantmentLevel(Enchantment.power.effectId, this.getHeldItem());
         int var5 = EnchantmentHelper.getEnchantmentLevel(Enchantment.punch.effectId, this.getHeldItem());
         var3.setDamage((double)(par2 * 2.0F) + this.rand.nextGaussian() * 0.25D + (double)((float)this.worldObj.difficultySetting * 0.11F));
