--- a/minecraft/net/minecraft/src/EntityWaterMob.java
+++ b/minecraft/net/minecraft/src/EntityWaterMob.java
@@ -52,7 +52,12 @@
         int var1 = this.getAir();
         super.onEntityUpdate();
 
+        // FCMOD: Change to be more forgiving on whether the squid is in water or not 
+        /*
         if (this.isEntityAlive() && !this.isInsideOfMaterial(Material.water))
+        */
+        if (this.isEntityAlive() && !inWater)
+        // END FCMOD
         {
             --var1;
             this.setAir(var1);
