--- a/minecraft/net/minecraft/src/EntityFireworkSparkFX.java
+++ b/minecraft/net/minecraft/src/EntityFireworkSparkFX.java
@@ -115,7 +115,7 @@
 
         if (this.field_92054_ax && this.particleAge < this.particleMaxAge / 2 && (this.particleAge + this.particleMaxAge) % 2 == 0)
         {
-            EntityFireworkSparkFX var1 = new EntityFireworkSparkFX(this.worldObj, this.posX, this.posY, this.posZ, 0.0D, 0.0D, 0.0D, this.field_92047_az);
+            EntityFireworkSparkFX var1 = (EntityFireworkSparkFX) EntityList.createEntityOfType(EntityFireworkSparkFX.class, this.worldObj, this.posX, this.posY, this.posZ, 0.0D, 0.0D, 0.0D, this.field_92047_az);
             var1.setRBGColorF(this.particleRed, this.particleGreen, this.particleBlue);
             var1.particleAge = var1.particleMaxAge / 2;
 
