--- a/minecraft/net/minecraft/src/EntityAIEatGrass.java
+++ b/minecraft/net/minecraft/src/EntityAIEatGrass.java
@@ -85,7 +85,12 @@
             else if (this.theWorld.getBlockId(var1, var2 - 1, var3) == Block.grass.blockID)
             {
                 this.theWorld.playAuxSFX(2001, var1, var2 - 1, var3, Block.grass.blockID);
+                // FCMOD: Changed to send notification
+                /*
                 this.theWorld.setBlock(var1, var2 - 1, var3, Block.dirt.blockID, 0, 2);
+                */
+                this.theWorld.setBlock(var1, var2 - 1, var3, Block.dirt.blockID, 0, 3);
+                // END FCMOD
                 this.theEntity.eatGrassBonus();
             }
         }
