--- a/minecraft/net/minecraft/src/EntityMinecartContainer.java
+++ b/minecraft/net/minecraft/src/EntityMinecartContainer.java
@@ -30,9 +30,6 @@
 
             if (var3 != null)
             {
-                float var4 = this.rand.nextFloat() * 0.8F + 0.1F;
-                float var5 = this.rand.nextFloat() * 0.8F + 0.1F;
-                float var6 = this.rand.nextFloat() * 0.8F + 0.1F;
 
                 while (var3.stackSize > 0)
                 {
@@ -44,8 +41,8 @@
                     }
 
                     var3.stackSize -= var7;
-                    EntityItem var8 = new EntityItem(this.worldObj, this.posX + (double)var4, this.posY + (double)var5, this.posZ + (double)var6, new ItemStack(var3.itemID, var7, var3.getItemDamage()));
-                    float var9 = 0.05F;
+                    EntityItem var8 = (EntityItem) EntityList.createEntityOfType(EntityItem.class, this.worldObj, this.posX, this.posY, this.posZ, new ItemStack(var3.itemID, var7, var3.getItemDamage()));
+                    float var9 = 0.08F;
                     var8.motionX = (double)((float)this.rand.nextGaussian() * var9);
                     var8.motionY = (double)((float)this.rand.nextGaussian() * var9 + 0.2F);
                     var8.motionZ = (double)((float)this.rand.nextGaussian() * var9);
@@ -192,9 +189,6 @@
 
                 if (var2 != null)
                 {
-                    float var3 = this.rand.nextFloat() * 0.8F + 0.1F;
-                    float var4 = this.rand.nextFloat() * 0.8F + 0.1F;
-                    float var5 = this.rand.nextFloat() * 0.8F + 0.1F;
 
                     while (var2.stackSize > 0)
                     {
@@ -206,14 +200,14 @@
                         }
 
                         var2.stackSize -= var6;
-                        EntityItem var7 = new EntityItem(this.worldObj, this.posX + (double)var3, this.posY + (double)var4, this.posZ + (double)var5, new ItemStack(var2.itemID, var6, var2.getItemDamage()));
+                        EntityItem var7 = (EntityItem) EntityList.createEntityOfType(EntityItem.class, this.worldObj, this.posX , this.posY, this.posZ, new ItemStack(var2.itemID, var6, var2.getItemDamage()));
 
                         if (var2.hasTagCompound())
                         {
                             var7.getEntityItem().setTagCompound((NBTTagCompound)var2.getTagCompound().copy());
                         }
 
-                        float var8 = 0.05F;
+                        float var8 = 0.08F;
                         var7.motionX = (double)((float)this.rand.nextGaussian() * var8);
                         var7.motionY = (double)((float)this.rand.nextGaussian() * var8 + 0.2F);
                         var7.motionZ = (double)((float)this.rand.nextGaussian() * var8);
