--- a/minecraft/net/minecraft/src/EntityEnderEye.java
+++ b/minecraft/net/minecraft/src/EntityEnderEye.java
@@ -177,7 +177,7 @@
 
                 if (this.shatterOrDrop)
                 {
-                    this.worldObj.spawnEntityInWorld(new EntityItem(this.worldObj, this.posX, this.posY, this.posZ, new ItemStack(Item.eyeOfEnder)));
+                    this.worldObj.spawnEntityInWorld(EntityList.createEntityOfType(EntityItem.class, this.worldObj, this.posX, this.posY, this.posZ, new ItemStack(Item.eyeOfEnder)));
                 }
                 else
                 {
