--- a/minecraft/net/minecraft/src/EntityBubbleFX.java
+++ b/minecraft/net/minecraft/src/EntityBubbleFX.java
@@ -1,13 +1,27 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.cc.ColorizeBlock;
+import com.prupe.mcpatcher.cc.Colorizer;
+
 public class EntityBubbleFX extends EntityFX
 {
     public EntityBubbleFX(World par1World, double par2, double par4, double par6, double par8, double par10, double par12)
     {
         super(par1World, par2, par4, par6, par8, par10, par12);
+        
+        if (ColorizeBlock.computeWaterColor(false, (int)this.posX, (int)this.posY, (int)this.posZ))
+        {
+            this.particleRed = Colorizer.setColor[0];
+            this.particleGreen = Colorizer.setColor[1];
+            this.particleBlue = Colorizer.setColor[2];
+        }
+        else
+        {
         this.particleRed = 1.0F;
         this.particleGreen = 1.0F;
         this.particleBlue = 1.0F;
+        }
+        
         this.setParticleTextureIndex(32);
         this.setSize(0.02F, 0.02F);
         this.particleScale *= this.rand.nextFloat() * 0.6F + 0.2F;
