--- /dev/null
+++ b/minecraft/net/minecraft/src/FCBehaviorDispenseDynamite.java
@@ -0,0 +1,17 @@
+// FCMOD
+
+package net.minecraft.src;
+
+import net.minecraft.server.MinecraftServer;
+
+public class FCBehaviorDispenseDynamite extends BehaviorProjectileDispense
+{
+    public FCBehaviorDispenseDynamite()
+    {
+    }
+
+    protected IProjectile getProjectileEntity( World world, IPosition pos )
+    {
+        return new FCEntityDynamite( world, pos.getX(), pos.getY(), pos.getZ(), FCBetterThanWolves.fcItemDynamite.itemID );
+    }
+}
