--- /dev/null
+++ b/minecraft/net/minecraft/src/FCBlockMilkChocolate.java
@@ -0,0 +1,17 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCBlockMilkChocolate extends FCBlockMilk
+{
+    public FCBlockMilkChocolate( int iBlockID )
+    {
+    	super( iBlockID );
+    	
+        setUnlocalizedName( "fcBlockMilkChocolate" );
+    }
+    
+	//------------- Class Specific Methods ------------//
+	
+	//----------- Client Side Functionality -----------//
+}
