--- /dev/null
+++ b/minecraft/net/minecraft/src/FCBlockSilverfish.java
@@ -0,0 +1,111 @@
+// FCMOD
+
+package net.minecraft.src;
+
+import java.util.List;
+import java.util.Random;
+
+public class FCBlockSilverfish extends BlockSilverfish
+{
+	private static final int m_iHatchFrequency = 1200;
+    protected final Block referenceBlock;
+    protected final int referenceBlockMetadata;
+    
+	/**
+	 * 
+	 * @param iBlockID
+	 * @param reference Block
+	 * @param reference Block Metadata
+	 */
+    public FCBlockSilverfish( int iBlockID, Block refBlock, int refBlockMetadata )
+    {
+    	super( iBlockID );
+    	
+    	setHardness( 1.5F );
+        referenceBlock = refBlock;
+        referenceBlockMetadata = refBlockMetadata;
+
+    	
+    	SetPicksEffectiveOn();
+        SetChiselsEffectiveOn();
+    	
+        setTickRandomly( true );
+        
+        setUnlocalizedName( "monsterStoneEgg" );
+    }
+    
+	@Override
+    public void RandomUpdateTick( World world, int i, int j, int k, Random rand )
+    {
+		if ( world.provider.dimensionId == 1 )
+		{
+			if ( rand.nextInt( m_iHatchFrequency ) == 0 )
+			{
+				// block destroy FX
+				
+				int iMetadata = world.getBlockMetadata( i, j, k );
+				
+		        world.playAuxSFX( FCBetterThanWolves.m_iBlockDestroyRespectParticleSettingsAuxFXID, i, j, k, blockID + ( iMetadata << 12 ) );
+		        
+				world.setBlockWithNotify( i, j, k, 0 );
+				
+				int iNumSilverfish = 1;
+				
+				if ( rand.nextInt( 2 ) == 0 )
+				{
+					iNumSilverfish++;
+				}
+				
+				for ( int iTempCount = 0; iTempCount < iNumSilverfish; iTempCount++ )
+				{
+		            EntitySilverfish silverfish = (EntitySilverfish) EntityList.createEntityOfType(EntitySilverfish.class, world);
+		            
+		            silverfish.setLocationAndAngles( (double)i + 0.5D, (double)j, (double)k + 0.5D, 0.0F, 0.0F);
+		            
+		            world.spawnEntityInWorld( silverfish );
+				}
+				
+				FCUtilsItem.DropSingleItemAsIfBlockHarvested( world, i, j, k, Block.gravel.blockID, 0 );
+				
+				FCUtilsItem.DropSingleItemAsIfBlockHarvested( world, i, j, k, Item.clay.itemID, 0 );				
+			}
+		}
+    }
+    
+    @Override
+    protected ItemStack createStackedBlock( int iMetadata )
+    {
+        return new ItemStack( referenceBlock, 1, referenceBlockMetadata );
+    }
+    
+    @Override
+	public boolean isBlockInfestedBy(EntityLiving entity)
+	{
+		return (entity instanceof EntitySilverfish);
+	}
+	
+	//----------- Client Side Functionality -----------//
+	
+    @Override
+    public Icon getIcon( int iSide, int iMetadata )
+    {
+		return referenceBlock.getIcon( iSide, referenceBlockMetadata );
+    }
+
+	@Override
+    public void randomDisplayTick( World world, int i, int j, int k, Random rand )
+    {
+        if ( rand.nextInt( 32 ) == 0 )
+        {
+            world.playSound( (double)i + 0.5D, (double)j + 0.5D, (double)k + 0.5D, "mob.silverfish.step", 
+            	rand.nextFloat() * 0.05F + 0.2F, rand.nextFloat() * 1.0F + 0.5F, false );
+        }
+    }
+	
+	//override regular silverfish to prevent triple creative inventory blocks
+	@Override
+    public void getSubBlocks(int par1, CreativeTabs par2CreativeTabs, List par3List)
+    {
+		par3List.add(new ItemStack(par1, 1, 0));
+    }
+}
