--- /dev/null
+++ b/minecraft/net/minecraft/src/FCBlockCarrotBase.java
@@ -0,0 +1,66 @@
+// FCMOD
+
+package net.minecraft.src;
+
+import java.util.Random;
+
+public abstract class FCBlockCarrotBase extends FCBlockCropsDailyGrowth
+{
+    public FCBlockCarrotBase( int iBlockID )
+    {
+        super( iBlockID );
+    }
+	
+    @Override
+    public ItemStack GetStackRetrievedByBlockDispenser( World world, int i, int j, int k )
+    {
+    	int iMetadata = world.getBlockMetadata( i, j, k );
+		
+		if ( GetGrowthLevel(iMetadata) >= 7 )
+		{
+			return super.GetStackRetrievedByBlockDispenser( world, i, j, k );			
+		}
+    	
+    	return null;
+    }
+    
+    @Override
+    protected boolean requiresNaturalLight() {
+    	return false;
+    }
+    
+    //------------- Class Specific Methods ------------//
+    
+	//----------- Client Side Functionality -----------//
+	
+	private Icon[] iconArray;
+	
+	public Icon getIcon(int par1, int par2)
+    {
+		par2 = GetGrowthLevel(par2);
+		
+        if (par2 < 7)
+        {
+            if (par2 == 6)
+            {
+                par2 = 5;
+            }
+
+            return this.iconArray[par2 >> 1];
+        }
+        else
+        {
+            return this.iconArray[3];
+        }
+    }
+	
+	public void registerIcons(IconRegister par1IconRegister)
+    {
+        this.iconArray = new Icon[4];
+
+        for (int var2 = 0; var2 < this.iconArray.length; ++var2)
+        {
+            this.iconArray[var2] = par1IconRegister.registerIcon(this.getUnlocalizedName2() + "_" + var2);
+        }
+    }
+}
