--- /dev/null
+++ b/minecraft/net/minecraft/src/FCBlockOreStorage.java
@@ -0,0 +1,17 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCBlockOreStorage extends FCBlockFullBlock
+{
+    public FCBlockOreStorage( int iBlockID )
+    {
+        super( iBlockID, Material.iron );
+        
+        SetPicksEffectiveOn();
+        
+        setCreativeTab( CreativeTabs.tabBlock );
+    }
+    
+	//----------- Client Side Functionality -----------//
+}
