--- /dev/null
+++ b/minecraft/net/minecraft/src/FCBlockChunkOreStorage.java
@@ -0,0 +1,29 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public abstract class FCBlockChunkOreStorage extends FCBlockFallingFullBlock
+{
+    protected FCBlockChunkOreStorage( int iBlockID )
+    {
+        super( iBlockID, Material.rock );
+        
+        setHardness( 1F );
+        setResistance( 5F );
+        SetPicksEffectiveOn();
+        
+        setStepSound( soundStoneFootstep );
+        
+		setCreativeTab( CreativeTabs.tabBlock );
+    }
+    
+	//------------- Class Specific Methods ------------//
+	
+	//----------- Client Side Functionality -----------//
+
+    @Override
+    public void RenderBlockSecondPass( RenderBlocks renderBlocks, int i, int j, int k, boolean bFirstPassResult )
+    {
+        RenderCookingByKilnOverlay( renderBlocks, i, j, k, bFirstPassResult );
+    }
+}
