--- /dev/null
+++ b/minecraft/net/minecraft/src/FCBlockMelon.java
@@ -0,0 +1,46 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCBlockMelon extends FCBlockGourd
+{
+    protected FCBlockMelon( int iBlockID )
+    {
+        super( iBlockID );        
+    }
+    
+    //------------- FCBlockGourd ------------//
+    
+    @Override
+	protected Item ItemToDropOnExplode()
+    {
+    	return FCBetterThanWolves.fcItemMelonMashed;
+    }
+	
+    @Override
+	protected int ItemCountToDropOnExplode()
+    {
+    	return 2;
+    }
+	
+    @Override
+	protected int AuxFXIDOnExplode()
+    {
+    	return FCBetterThanWolves.m_iMelonExplodeAuxFXID;
+    }
+    
+	protected DamageSource GetFallDamageSource()
+	{
+		return FCDamageSourceCustom.m_DamageSourceMelon;
+	}
+	
+	//----------- Client Side Functionality -----------//
+    
+	@Override
+    public void registerIcons( IconRegister register )
+    {
+		blockIcon = register.registerIcon( "melon_side" );
+		
+		m_IconTop = register.registerIcon( "melon_top" );		
+    }
+}
