--- /dev/null
+++ b/minecraft/net/minecraft/src/FCBlockStoneBrickLooseStairs.java
@@ -0,0 +1,44 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCBlockStoneBrickLooseStairs extends FCBlockMortarReceiverStairs{
+	private int strata;
+	
+    protected FCBlockStoneBrickLooseStairs( int iBlockID, int strata ) {
+        super( iBlockID, FCBetterThanWolves.fcBlockStoneBrickLoose, strata << 2 );
+        
+        this.strata = strata;
+        
+        SetPicksEffectiveOn();
+        
+        setUnlocalizedName( "fcBlockStoneBrickLooseStairs" );        
+    }
+	
+    @Override
+    public boolean OnMortarApplied( World world, int i, int j, int k ) {
+    	int blockID = Block.stairsStoneBrick.blockID;
+    	
+    	if (strata != 0) {
+    		if (strata == 1) {
+    			blockID = FCBetterThanWolves.fcBlockStoneBrickStairsMidStrata.blockID;
+    		}
+    		else {
+    			blockID = FCBetterThanWolves.fcBlockStoneBrickStairsDeepStrata.blockID;
+    		}
+    	}
+    	
+		world.setBlockAndMetadataWithNotify( i, j, k, blockID, 
+			world.getBlockMetadata( i, j, k ) );
+		
+		return true;
+    }
+    
+    //------------- Class Specific Methods ------------//
+    
+    public int getStrata() {
+		return strata;
+	}
+    
+	//----------- Client Side Functionality -----------//
+}
\ No newline at end of file
