--- /dev/null
+++ b/minecraft/net/minecraft/src/FCBlockOreGold.java
@@ -0,0 +1,35 @@
+// FCMOD
+
+package net.minecraft.src;
+
+import java.util.Random;
+
+public class FCBlockOreGold extends FCBlockOreStaged
+{
+    public FCBlockOreGold( int iBlockID )
+    {
+        super( iBlockID );
+    }
+    
+    @Override
+    public int idDropped( int iMetadata, Random rand, int iFortuneModifier )
+    {
+        return FCBetterThanWolves.fcItemChunkGoldOre.itemID;
+    }
+    
+    @Override
+    public int IdDroppedOnConversion( int iMetadata )
+    {
+        return FCBetterThanWolves.fcItemPileGoldOre.itemID;
+    }
+    
+    @Override
+    public int GetRequiredToolLevelForOre( IBlockAccess blockAccess, int i, int j, int k )
+    {
+    	return 2;
+    }
+    
+    //------------- Class Specific Methods ------------//
+	
+	//----------- Client Side Functionality -----------//
+}
