--- /dev/null
+++ b/minecraft/net/minecraft/src/FCBlockNetherBrickSidingAndCornerAndDecorative.java
@@ -0,0 +1,13 @@
+//FCMOD
+
+package net.minecraft.src;
+
+public class FCBlockNetherBrickSidingAndCornerAndDecorative extends FCBlockSidingAndCornerAndDecorative
+{
+	protected FCBlockNetherBrickSidingAndCornerAndDecorative( int iBlockID )
+	{
+		super( iBlockID, FCBetterThanWolves.fcMaterialNetherRock, 
+			"fcBlockDecorativeNetherBrick", 2.0F, 10F, Block.soundStoneFootstep, 
+			"fcNetherBrickSiding" );
+	}	
+}
\ No newline at end of file
