--- /dev/null
+++ b/minecraft/net/minecraft/src/FCClientRenderEntitySoulSand.java
@@ -0,0 +1,17 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCClientRenderEntitySoulSand extends Render
+{
+    public FCClientRenderEntitySoulSand()
+    {
+    }
+
+    @Override
+    public void doRender( Entity entity, double d, double d1, double d2, 
+            float f, float f1 )
+    {
+    	// intentionally invisible renderer
+    }
+}
\ No newline at end of file
