--- /dev/null
+++ b/minecraft/net/minecraft/src/FCCraftingManagerSaw.java
@@ -0,0 +1,50 @@
+package net.minecraft.src;
+
+import java.util.ArrayList;
+
+public class FCCraftingManagerSaw {
+	public static FCCraftingManagerSaw instance = new FCCraftingManagerSaw();
+	
+	private ArrayList<FCCraftingManagerSawRecipe> recipes = new ArrayList();
+	
+	private FCCraftingManagerSaw() {}
+	
+	public void addRecipe(ItemStack[] output, Block block, int metadata) {
+		FCCraftingManagerSawRecipe recipe = new FCCraftingManagerSawRecipe(output, block, metadata);
+		
+		recipes.add(recipe);
+	}
+	
+	public boolean removeRecipe(ItemStack[] output, Block block, int metadata) {
+		FCCraftingManagerSawRecipe recipeToRemove = new FCCraftingManagerSawRecipe(output, block, metadata);
+		
+		for (FCCraftingManagerSawRecipe recipe : recipes) {
+			if (recipe.matchesRecipe(recipeToRemove)) {
+				recipes.remove(recipe);
+				return true;
+			}
+		}
+		
+		return false;
+	}
+	
+	public FCCraftingManagerSawRecipe getRecipe(Block block, int metadata) {
+		for (FCCraftingManagerSawRecipe recipe : recipes) {
+			if (recipe.matchesInputs(block, metadata)) {
+				return recipe;
+			}
+		}
+		
+		return null;
+	}
+	
+	public ItemStack[] getRecipeResult(Block block, int metadata) {
+		for (FCCraftingManagerSawRecipe recipe : recipes) {
+			if (recipe.matchesInputs(block, metadata)) {
+				return recipe.getOutput();
+			}
+		}
+		
+		return null;
+	}
+}
