--- /dev/null
+++ b/minecraft/net/minecraft/src/FCEntityVillagerButcher.java
@@ -0,0 +1,14 @@
+package net.minecraft.src;
+
+public class FCEntityVillagerButcher extends FCEntityVillager {
+	public FCEntityVillagerButcher(World world) {
+		super(world, FCEntityVillager.professionIDButcher);
+	}
+	
+	//CLIENT ONLY
+	public String getTexture() {
+		if (this.GetCurrentTradeLevel() >= 4)
+            return "/btwmodtex/fcButcherLvl.png";
+        return "/mob/villager/butcher.png";
+	}
+}
\ No newline at end of file
