--- /dev/null
+++ b/minecraft/net/minecraft/src/FCEntityVillagerBlacksmith.java
@@ -0,0 +1,12 @@
+package net.minecraft.src;
+
+public class FCEntityVillagerBlacksmith extends FCEntityVillager {
+	public FCEntityVillagerBlacksmith(World world) {
+		super(world, FCEntityVillager.professionIDBlacksmith);
+	}
+	
+	//CLIENT ONLY
+	public String getTexture() {
+        return "/mob/villager/smith.png";
+	}
+}
\ No newline at end of file
