--- /dev/null
+++ b/minecraft/net/minecraft/src/FCItemBandage.java
@@ -0,0 +1,29 @@
+package net.minecraft.src;
+
+import java.util.Iterator;
+import java.util.List;
+
+public class FCItemBandage extends FCItemPotion {
+	public FCItemBandage(int itemID) {
+		super(itemID);
+		this.setUnlocalizedName("fcItemBandage");
+	}
+    
+    @Override
+    public ItemStack onEaten(ItemStack itemStack, World world, EntityPlayer player) {
+        if (!player.capabilities.isCreativeMode) {
+            itemStack.stackSize--;
+        }
+
+        if (!world.isRemote) {
+        	player.addPotionEffect(new PotionEffect(Potion.heal.id, 1, 1));
+        }
+
+        return itemStack;
+    }
+	
+	@Override
+    public int getMaxItemUseDuration(ItemStack par1ItemStack) {
+        return 120;
+    }
+}
\ No newline at end of file
