--- /dev/null
+++ b/minecraft/net/minecraft/src/FCItemClay.java
@@ -0,0 +1,21 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCItemClay extends FCItemMortar
+{
+    public FCItemClay( int iItemID )
+    {
+    	super( iItemID );
+    	
+    	SetFilterableProperties( m_iFilterable_Small );
+    	
+    	setUnlocalizedName( "clay" );
+    	
+    	setCreativeTab( CreativeTabs.tabMaterials );
+    }
+    
+    //------------- Class Specific Methods ------------//
+    
+	//------------ Client Side Functionality ----------//
+}
