--- /dev/null
+++ b/minecraft/net/minecraft/src/FCItemPileSand.java
@@ -0,0 +1,22 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCItemPileSand extends Item
+{
+    public FCItemPileSand( int iItemID )
+    {
+        super( iItemID );
+        
+        SetBellowsBlowDistance( 2 );
+		SetFilterableProperties( m_iFilterable_Fine );
+        
+        setUnlocalizedName( "fcItemPileSand" );
+        
+        setCreativeTab( CreativeTabs.tabMaterials );
+    }
+    
+    //------------- Class Specific Methods ------------//
+    
+	//------------ Client Side Functionality ----------//
+}
