--- /dev/null
+++ b/minecraft/net/minecraft/src/FCItemDoorWood.java
@@ -0,0 +1,26 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCItemDoorWood extends FCItemDoor
+{
+    public FCItemDoorWood( int iITemID )
+    {
+        super( iITemID );
+
+        SetBuoyant();
+        SetIncineratedInCrucible();
+        
+        setUnlocalizedName( "doorWood" );
+    }
+
+    @Override
+    public Block GetDoorBlock()
+    {
+    	return FCBetterThanWolves.fcBlockDoorWood;
+    }
+    
+    //------------- Class Specific Methods ------------//
+	
+	//----------- Client Side Functionality -----------//
+}
