--- /dev/null
+++ b/minecraft/net/minecraft/src/FCItemChunkOreGold.java
@@ -0,0 +1,17 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCItemChunkOreGold extends FCItemPlacesAsBlock
+{
+    public FCItemChunkOreGold( int iItemID )
+    {
+        super( iItemID, FCBetterThanWolves.fcBlockChunkOreGold.blockID );
+        
+        SetFilterableProperties( Item.m_iFilterable_Small );
+        
+        setUnlocalizedName( "fcItemChunkGoldOre" );
+        
+        setCreativeTab( CreativeTabs.tabMaterials );
+    }
+}
