--- /dev/null
+++ b/minecraft/net/minecraft/src/FCItemNetherQuartz.java
@@ -0,0 +1,18 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCItemNetherQuartz extends Item
+{
+    protected FCItemNetherQuartz( int iItemID )
+    {
+    	super( iItemID );
+    }
+    
+	//----------- Client Side Functionality -----------//
+	
+    public void registerIcons(IconRegister par1IconRegister)
+    {
+        this.itemIcon = par1IconRegister.registerIcon( "fcItemNetherQuartz" );
+    }
+}
