--- /dev/null
+++ b/minecraft/net/minecraft/src/FCItemRefinedHoe.java
@@ -0,0 +1,40 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCItemRefinedHoe extends FCItemHoe
+{
+    public FCItemRefinedHoe( int iItemID )
+    {
+        super( iItemID, EnumToolMaterial.SOULFORGED_STEEL );
+        
+        setUnlocalizedName( "fcItemHoeRefined" );        
+    }
+
+    @Override
+    public EnumAction getItemUseAction(ItemStack itemstack)
+    {
+        return EnumAction.block;
+    }
+
+    @Override
+    public int getMaxItemUseDuration(ItemStack itemstack)
+    {
+        return 0x11940;
+    }
+
+    @Override
+    public ItemStack onItemRightClick(ItemStack itemstack, World world, EntityPlayer entityplayer)
+    {
+    	if (!entityplayer.isUsingSpecialKey())
+    	{
+    		entityplayer.setItemInUse(itemstack, getMaxItemUseDuration(itemstack));
+    	}
+    	
+        return itemstack;
+    }
+
+    //------------- Class Specific Methods ------------//
+    
+	//------------ Client Side Functionality ----------//
+}
\ No newline at end of file
