--- /dev/null
+++ b/minecraft/net/minecraft/src/FCItemKnittingNeedles.java
@@ -0,0 +1,35 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCItemKnittingNeedles extends Item
+{
+    public FCItemKnittingNeedles( int iItemID )
+    {
+    	super( iItemID );
+    	
+    	SetBuoyant();
+    	SetFurnaceBurnTime( FCEnumFurnaceBurnTime.SHAFT );
+    	SetFilterableProperties( Item.m_iFilterable_Narrow );
+    	
+    	setUnlocalizedName( "fcItemKnittingNeedles" );
+    	
+    	setCreativeTab( CreativeTabs.tabTools );
+    }
+    
+    @Override
+    public boolean GetCanBeFedDirectlyIntoCampfire( int iItemDamage )
+    {
+    	return false;
+    }
+    
+    @Override
+    public boolean GetCanBeFedDirectlyIntoBrickOven( int iItemDamage )
+    {
+    	return false;
+    }
+    
+    //------------- Class Specific Methods ------------//
+    
+	//------------ Client Side Functionality ----------//
+}
