--- /dev/null
+++ b/minecraft/net/minecraft/src/FCItemBone.java
@@ -0,0 +1,23 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCItemBone extends Item
+{
+    public FCItemBone( int iItemID )
+    {
+        super( iItemID );
+        
+        maxStackSize = 16;
+        
+        SetBuoyant();
+        SetIncineratedInCrucible();
+        SetFilterableProperties( m_iFilterable_Narrow );
+
+        setFull3D();
+        
+        setUnlocalizedName( "bone" );
+        
+        setCreativeTab( CreativeTabs.tabMisc );
+    }
+}
