--- /dev/null
+++ b/minecraft/net/minecraft/src/FCItemBreedingHarness.java
@@ -0,0 +1,45 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCItemBreedingHarness extends Item {
+    public FCItemBreedingHarness(int itemID) {
+    	super(itemID);
+    	
+    	SetBuoyant();
+    	
+    	setUnlocalizedName("fcItemHarnessBreeding");
+    	
+    	setCreativeTab(CreativeTabs.tabMaterials);
+    }
+    
+    @Override
+    public boolean itemInteractionForEntity(ItemStack itemStack, EntityLiving targetEntity) {
+    	if (targetEntity instanceof EntityAnimal) {
+	    	EntityAnimal animal = (EntityAnimal)targetEntity;
+	
+        	if (!animal.isChild() && !animal.getWearingBreedingHarness()) {
+		        if (targetEntity instanceof FCEntitySheep) {       	
+		            if (!animal.worldObj.isRemote) {
+		            	FCEntitySheep sheep = (FCEntitySheep)animal;
+		            	
+		            	sheep.setSheared(true);
+		            }
+		        }
+		        else if (!(targetEntity instanceof FCEntityPig || targetEntity instanceof FCEntityCow)) {
+		        	return false;
+		        }
+		        
+	            itemStack.stackSize--;
+
+	            if (!animal.worldObj.isRemote) {
+	            	animal.setWearingBreedingHarness(true);
+	            }
+	        	
+	            return true;
+        	}
+    	}
+        
+        return false;
+    }
+}
\ No newline at end of file
