--- /dev/null
+++ b/minecraft/net/minecraft/src/FCItemBlockVase.java
@@ -0,0 +1,22 @@
+// FCMOD 
+
+package net.minecraft.src;
+
+public class FCItemBlockVase extends ItemBlock
+{
+    public FCItemBlockVase( int i )
+    {
+        super( i );
+        
+        setMaxDamage( 0 );
+        setHasSubtypes( true );
+        
+        setUnlocalizedName( "fcBlockVase" );
+    }
+
+    @Override
+    public int getMetadata( int i )
+    {
+    	return i;
+    }    
+}
