--- /dev/null
+++ b/minecraft/net/minecraft/src/FCMaterialNetherGrowth.java
@@ -0,0 +1,19 @@
+// FCMOD
+
+package net.minecraft.src;
+
+public class FCMaterialNetherGrowth extends Material
+{
+    public FCMaterialNetherGrowth( MapColor mapColor )
+    {
+        super( mapColor );
+        setNoPushMobility(); // breaks on push by piston
+    }
+    
+    public boolean blocksMovement()
+    {
+    	// so that fluid can overwrite block
+    	
+        return false;
+    }
+}
