--- /dev/null
+++ b/minecraft/net/minecraft/src/FCTileEntityBeaconEffectDescriptor.java
@@ -0,0 +1,17 @@
+// FCMOD
+
+package net.minecraft.src;
+
+import java.util.*;
+
+public class FCTileEntityBeaconEffectDescriptor
+{
+	public int m_iBlockMetadata;
+	public int m_iEffectID;
+	
+    public FCTileEntityBeaconEffectDescriptor( int iBlockMetadata, int iEffectID )
+    {
+    	m_iBlockMetadata = iBlockMetadata;
+    	m_iEffectID = iEffectID;
+    }
+}
