--- /dev/null
+++ b/minecraft/net/minecraft/src/FCUtilsTreeGrowerForest.java
@@ -0,0 +1,54 @@
+package net.minecraft.src;
+
+import java.util.Random;
+
+public class FCUtilsTreeGrowerForest extends FCUtilsTreeGrower {
+	public int minTreeHeight;
+	public int maxTreeHeight;
+	
+	public int woodBlockID;
+	public int woodMetadata;
+	
+	public int stumpBlockID;
+	public int stumpMetadata;
+	
+	public int leavesBlockID;
+	public int leavesMetadata;
+	
+	public boolean growVines;
+	public boolean growCocoa;
+
+	public FCUtilsTreeGrowerForest(
+			int minTreeHeight, int maxTreeHeight,
+			int woodBlockID, int woodMetadata,
+			int stumpBlockID, int stumpMetadata,
+			int leavesBlockID, int leavesMetadata,
+			boolean growVines, boolean growCocoa) 
+	{
+		this.minTreeHeight = minTreeHeight;
+		this.maxTreeHeight = maxTreeHeight;
+		
+		this.woodBlockID = woodBlockID;
+		this.woodMetadata = woodMetadata;
+		
+		this.stumpBlockID = stumpBlockID;
+		this.stumpMetadata = stumpMetadata;
+		
+		this.leavesBlockID = leavesBlockID;
+		this.leavesMetadata = leavesMetadata;
+		
+		this.growVines = growVines;
+		this.growCocoa = growCocoa;
+	}
+
+	public boolean generate(World par1World, Random par2Random, int par3, int par4, int par5)
+    {
+        return false;
+    }
+
+	@Override
+	public boolean growTree(World world, Random rand, int x, int y, int z, boolean isWorldGen) {
+		// TODO Auto-generated method stub
+		return false;
+	}
+}
\ No newline at end of file
