--- a/minecraft/net/minecraft/src/GameSettings.java
+++ b/minecraft/net/minecraft/src/GameSettings.java
@@ -1,5 +1,6 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.Config;
 import java.io.BufferedReader;
 import java.io.File;
 import java.io.FileReader;
@@ -87,6 +88,7 @@
     public KeyBinding keyBindPlayerList = new KeyBinding("key.playerlist", 15);
     public KeyBinding keyBindPickBlock = new KeyBinding("key.pickItem", -98);
     public KeyBinding keyBindCommand = new KeyBinding("key.command", 53);
+    public KeyBinding keyBindSpecial = new KeyBinding("key.special", 29);
     public KeyBinding[] keyBindings;
     protected Minecraft mc;
     private File optionsFile;
@@ -127,7 +129,7 @@
 
     public GameSettings(Minecraft par1Minecraft, File par2File)
     {
-        this.keyBindings = new KeyBinding[] {this.keyBindAttack, this.keyBindUseItem, this.keyBindForward, this.keyBindLeft, this.keyBindBack, this.keyBindRight, this.keyBindJump, this.keyBindSneak, this.keyBindDrop, this.keyBindInventory, this.keyBindChat, this.keyBindPlayerList, this.keyBindPickBlock, this.keyBindCommand};
+        this.keyBindings = new KeyBinding[] {this.keyBindAttack, this.keyBindUseItem, this.keyBindForward, this.keyBindLeft, this.keyBindBack, this.keyBindRight, this.keyBindJump, this.keyBindSneak, this.keyBindDrop, this.keyBindInventory, this.keyBindChat, this.keyBindPlayerList, this.keyBindPickBlock, this.keyBindCommand, keyBindSpecial};
         this.difficulty = 2;
         this.hideGUI = false;
         this.thirdPersonView = 0;
@@ -145,13 +147,13 @@
         this.particleSetting = 0;
         this.language = "en_US";
         this.mc = par1Minecraft;
-        this.optionsFile = new File(par2File, "options.txt");
+        this.optionsFile = Config.getOptionsTxt(par2File, "options.txt");
         this.loadOptions();
     }
 
     public GameSettings()
     {
-        this.keyBindings = new KeyBinding[] {this.keyBindAttack, this.keyBindUseItem, this.keyBindForward, this.keyBindLeft, this.keyBindBack, this.keyBindRight, this.keyBindJump, this.keyBindSneak, this.keyBindDrop, this.keyBindInventory, this.keyBindChat, this.keyBindPlayerList, this.keyBindPickBlock, this.keyBindCommand};
+        this.keyBindings = new KeyBinding[] {this.keyBindAttack, this.keyBindUseItem, this.keyBindForward, this.keyBindLeft, this.keyBindBack, this.keyBindRight, this.keyBindJump, this.keyBindSneak, this.keyBindDrop, this.keyBindInventory, this.keyBindChat, this.keyBindPlayerList, this.keyBindPickBlock, this.keyBindCommand, keyBindSpecial};
         this.difficulty = 2;
         this.hideGUI = false;
         this.thirdPersonView = 0;
@@ -328,7 +330,13 @@
         if (par1EnumOptions == EnumOptions.DIFFICULTY)
         {
             this.difficulty = this.difficulty + par2 & 3;
+            // FCMOD: Code added to eliminate peaceful & easy (client only)
+            if ( difficulty < 2 )
+            {
+            	difficulty = 2;
         }
+            // END FCMOD
+        }
 
         if (par1EnumOptions == EnumOptions.GRAPHICS)
         {
@@ -585,7 +593,13 @@
                     if (var3[0].equals("difficulty"))
                     {
                         this.difficulty = Integer.parseInt(var3[1]);
+                        // FCMOD: Code added to remove peaceful and easy (client only)
+                        if ( difficulty < 2 )
+                        {
+                        	difficulty = 2;
                     }
+                        // END FCMOD
+                    }
 
                     if (var3[0].equals("fancyGraphics"))
                     {
