--- a/minecraft/net/minecraft/src/GuiInventory.java
+++ b/minecraft/net/minecraft/src/GuiInventory.java
@@ -57,6 +57,10 @@
     protected void drawGuiContainerForegroundLayer(int par1, int par2)
     {
         this.fontRenderer.drawString(StatCollector.translateToLocal("container.crafting"), 86, 16, 4210752);
+        
+        // FCMOD: Added (client only)
+        DrawSecondaryOutputIndicator();
+        // END FCMOD
     }
 
     /**
@@ -129,4 +133,29 @@
             this.mc.displayGuiScreen(new GuiStats(this, this.mc.statFileWriter));
         }
     }
+    // FCMOD: Added (client only)
+    private void DrawSecondaryOutputIndicator()
+    {
+    	if ( !mc.playerController.isInCreativeMode() )
+    	{
+	    	FCContainerPlayer container = (FCContainerPlayer)mc.thePlayer.inventoryContainer;
+	    	
+	    	if ( container != null )
+	    	{
+		        IRecipe recipe = CraftingManager.getInstance().FindMatchingRecipe(
+		        	container.craftMatrix, mc.theWorld );
+		        
+		        if ( recipe != null && recipe.HasSecondaryOutput() )
+		        {
+			        Slot outputSlot = (Slot)container.inventorySlots.get( 0 );
+			        
+			        int iDisplayX = outputSlot.xDisplayPosition + 20;
+			        int iDisplayY = outputSlot.yDisplayPosition + 5;
+			        
+			        FCClientUtilsRender.DrawSecondaryCraftingOutputIndicator( mc, iDisplayX, iDisplayY );
 }
+	    	}
+    	}
+    }
+    // END FCMOD
+}
