--- a/minecraft/net/minecraft/src/GuiSlotStatsItem.java
+++ b/minecraft/net/minecraft/src/GuiSlotStatsItem.java
@@ -79,7 +79,21 @@
     {
         StatCrafting var6 = this.func_77257_d(par1);
         int var7 = var6.getItemID();
+        // FCMOD: Code change
+        /*
         GuiStats.drawItemSprite(this.slotGuiStats, par2 + 40, par3, var7);
+        */
+        if ( var7 == Item.pocketSundial.itemID )
+        {
+        	TextureClock.m_clockTexture.UpdateInert();
+            GuiStats.drawItemSprite(this.slotGuiStats, par2 + 40, par3, var7);
+        	TextureClock.m_clockTexture.updateAnimation();        	
+        }
+        else
+        {
+            GuiStats.drawItemSprite(this.slotGuiStats, par2 + 40, par3, var7);
+        }
+        // END FCMOD
         this.func_77260_a((StatCrafting)StatList.objectBreakStats[var7], par2 + 115, par3, par1 % 2 == 0);
         this.func_77260_a((StatCrafting)StatList.objectCraftStats[var7], par2 + 165, par3, par1 % 2 == 0);
         this.func_77260_a(var6, par2 + 215, par3, par1 % 2 == 0);
