--- a/minecraft/net/minecraft/src/HttpUtilRunnable.java
+++ b/minecraft/net/minecraft/src/HttpUtilRunnable.java
@@ -49,6 +49,8 @@
 
         try
         {
+            try
+            {
             byte[] var4 = new byte[4096];
             URL var5 = new URL(this.sourceURL);
             var1 = var5.openConnection();
@@ -138,12 +140,14 @@
             if (this.feedbackHook != null)
             {
                 this.feedbackHook.onNoMoreProgress();
+                    return;
             }
         }
         catch (Throwable var26)
         {
             var26.printStackTrace();
         }
+        }
         finally
         {
             try
