--- a/minecraft/net/minecraft/src/IMerchant.java
+++ b/minecraft/net/minecraft/src/IMerchant.java
@@ -1,3 +1,5 @@
+// FCMOD: Class changes deprecated 10/02/2018
+
 package net.minecraft.src;
 
 public interface IMerchant
@@ -11,4 +13,10 @@
     void setRecipes(MerchantRecipeList var1);
 
     void useRecipe(MerchantRecipe var1);
+
+    // FCMOD: Added
+    public int GetCurrentTradeLevel(); 
+    public int GetCurrentTradeXP();
+    public int GetCurrentTradeMaxXP();
+    // END FCMOD
 }
