--- a/minecraft/net/minecraft/src/ItemEgg.java
+++ b/minecraft/net/minecraft/src/ItemEgg.java
@@ -23,7 +23,7 @@
 
         if (!par2World.isRemote)
         {
-            par2World.spawnEntityInWorld(new EntityEgg(par2World, par3EntityPlayer));
+            par2World.spawnEntityInWorld(EntityList.createEntityOfType(EntityEgg.class, par2World, par3EntityPlayer));
         }
 
         return par1ItemStack;
