--- a/minecraft/net/minecraft/src/ItemPotion.java
+++ b/minecraft/net/minecraft/src/ItemPotion.java
@@ -141,7 +141,7 @@
 
             if (!par2World.isRemote)
             {
-                par2World.spawnEntityInWorld(new EntityPotion(par2World, par3EntityPlayer, par1ItemStack));
+                par2World.spawnEntityInWorld(EntityList.createEntityOfType(EntityPotion.class, par2World, par3EntityPlayer, par1ItemStack));
             }
 
             return par1ItemStack;
