--- a/minecraft/net/minecraft/src/ItemEnderEye.java
+++ b/minecraft/net/minecraft/src/ItemEnderEye.java
@@ -163,7 +163,7 @@
 
             if (var7 != null)
             {
-                EntityEnderEye var6 = new EntityEnderEye(par2World, par3EntityPlayer.posX, par3EntityPlayer.posY + 1.62D - (double)par3EntityPlayer.yOffset, par3EntityPlayer.posZ);
+                EntityEnderEye var6 = (EntityEnderEye) EntityList.createEntityOfType(EntityEnderEye.class, par2World, par3EntityPlayer.posX, par3EntityPlayer.posY + 1.62D - (double)par3EntityPlayer.yOffset, par3EntityPlayer.posZ);
                 var6.moveTowards((double)var7.x, var7.y, (double)var7.z);
                 par2World.spawnEntityInWorld(var6);
                 par2World.playSoundAtEntity(par3EntityPlayer, "random.bow", 0.5F, 0.4F / (itemRand.nextFloat() * 0.4F + 0.8F));
