--- a/minecraft/net/minecraft/src/ModelWolf.java
+++ b/minecraft/net/minecraft/src/ModelWolf.java
@@ -28,6 +28,10 @@
     /** The wolf's mane */
     ModelRenderer wolfMane;
 
+    // FCMOD: Added (client only)
+    private float m_fHeadRotation;
+    // END FCMOD
+
     public ModelWolf()
     {
         float var1 = 0.0F;
@@ -107,9 +111,16 @@
      */
     public void setLivingAnimations(EntityLiving par1EntityLiving, float par2, float par3, float par4)
     {
+        // FCMOD: Changed (client only)
+        /*
         EntityWolf var5 = (EntityWolf)par1EntityLiving;
 
         if (var5.isAngry())
+        */
+        FCEntityWolf var5 = (FCEntityWolf)par1EntityLiving;
+
+        if ( var5.isAngry()|| var5.IsWildAndHostile() )
+        // END FCMOD
         {
             this.wolfTail.rotateAngleY = 0.0F;
         }
@@ -156,6 +167,12 @@
         this.wolfMane.rotateAngleZ = var5.getShakeAngle(par4, -0.08F);
         this.wolfBody.rotateAngleZ = var5.getShakeAngle(par4, -0.16F);
         this.wolfTail.rotateAngleZ = var5.getShakeAngle(par4, -0.2F);
+        
+        // FCMOD: Added (client only)
+        wolfHeadMain.rotationPointY = 13.5F + ((FCEntityWolf)par1EntityLiving).GetGrazeHeadVerticalOffset(par4) * 5.0F;
+        m_fHeadRotation = ((FCEntityWolf)par1EntityLiving).GetGrazeHeadRotation(par4);
+        wolfHeadMain.rotateAngleZ += var5.GetPossessionHeadRotation();
+        // END FCMOD
     }
 
     /**
@@ -169,5 +186,9 @@
         this.wolfHeadMain.rotateAngleX = par5 / (180F / (float)Math.PI);
         this.wolfHeadMain.rotateAngleY = par4 / (180F / (float)Math.PI);
         this.wolfTail.rotateAngleX = par3;
+        
+        // FCMOD: Added (client only)
+        wolfHeadMain.rotateAngleX = m_fHeadRotation;
+        // ENDD FCMOD
     }
 }
