--- a/minecraft/net/minecraft/src/ModelCow.java
+++ b/minecraft/net/minecraft/src/ModelCow.java
@@ -24,4 +24,55 @@
         --this.leg4.rotationPointZ;
         this.field_78151_h += 2.0F;
     }
+    
+    // FCMOD: Added (client only)
+    private float m_fHeadRotation;
+    
+    public void setLivingAnimations( EntityLiving entity, float par2, float par3, 
+    	float fPartialTick )
+    {
+        super.setLivingAnimations( entity, par2, par3, fPartialTick );
+
+        FCEntityCow cow = (FCEntityCow)entity;
+        
+        head.rotationPointY = 4F + cow.GetGrazeHeadVerticalOffset( fPartialTick ) * 4F;
+        
+        m_fHeadRotation = cow.GetGrazeHeadRotation( fPartialTick );
 }
+
+    public void setRotationAngles( float par1, float par2, float par3, 
+    	float par4, float par5, float par6, Entity entity )
+    {
+        super.setRotationAngles( par1, par2, par3, par4, par5, par6, entity );
+        
+        AdjustRotationAnglesForKickAttack( par1, par2, par3, par4, par5, par6, entity );
+        
+        head.rotateAngleX = m_fHeadRotation;
+    }
+    
+    private void AdjustRotationAnglesForKickAttack( float par1, float par2, float par3, 
+    	float par4, float par5, float par6, Entity entity)
+    {
+    	FCEntityCow cow = (FCEntityCow)entity;
+    	
+    	if ( cow != null )
+    	{
+	    	int iAttackProgressCounter = cow.m_iKickAttackInProgressCounter;
+	    	
+	    	if ( iAttackProgressCounter > 0 )
+	    	{
+	    		float fRotationFactor = 2F;
+	    		
+	    		if ( cow.m_iKickAttackLegUsed == 0 )
+	    		{
+	    			leg1.rotateAngleX = MathHelper.cos( (float)Math.PI * fRotationFactor ) * 1.4F;
+	    		}
+	    		else
+	    		{
+	    			leg2.rotateAngleX = MathHelper.cos( (float)Math.PI * fRotationFactor ) * 1.4F;
+	    		}
+	    	}
+    	}
+    }    
+    // END FCMOD
+}
