--- a/minecraft/net/minecraft/src/NpcMerchant.java
+++ b/minecraft/net/minecraft/src/NpcMerchant.java
@@ -34,5 +34,30 @@
         this.recipeList = par1MerchantRecipeList;
     }
 
-    public void useRecipe(MerchantRecipe par1MerchantRecipe) {}
+    public void useRecipe(MerchantRecipe par1MerchantRecipe) 
+    {
+    	// FCMOD: Added so that trades are blocked on client when they are used up
+    	par1MerchantRecipe.incrementToolUses();
+        // END FCMOD
 }
+    
+    // FCMOD: Added
+    @Override
+    public int GetCurrentTradeLevel()
+    {
+    	return 0;
+    }
+    
+    @Override
+    public int GetCurrentTradeXP()
+    {
+    	return 0;
+    }
+    
+    @Override
+    public int GetCurrentTradeMaxXP()
+    {
+    	return 0;
+    }    
+    // END FCMOD    
+}
