--- a/minecraft/net/minecraft/src/Packet15Place.java
+++ b/minecraft/net/minecraft/src/Packet15Place.java
@@ -25,11 +25,11 @@
 
     public Packet15Place() {}
 
-    public Packet15Place(int par1, int par2, int par3, int par4, ItemStack par5ItemStack, float par6, float par7, float par8)
+    public Packet15Place(int x, int y, int z, int par4, ItemStack par5ItemStack, float par6, float par7, float par8)
     {
-        this.xPosition = par1;
-        this.yPosition = par2;
-        this.zPosition = par3;
+        this.xPosition = x;
+        this.yPosition = y;
+        this.zPosition = z;
         this.direction = par4;
         this.itemStack = par5ItemStack != null ? par5ItemStack.copy() : null;
         this.xOffset = par6;
@@ -47,9 +47,16 @@
         this.zPosition = par1DataInputStream.readInt();
         this.direction = par1DataInputStream.read();
         this.itemStack = readItemStack(par1DataInputStream);
+        // FCMOD: Code change to fix rounding errors
+        /*
         this.xOffset = (float)par1DataInputStream.read() / 16.0F;
         this.yOffset = (float)par1DataInputStream.read() / 16.0F;
         this.zOffset = (float)par1DataInputStream.read() / 16.0F;
+        */
+        this.xOffset = (float)par1DataInputStream.readShort() / 32000F;
+        this.yOffset = (float)par1DataInputStream.readShort() / 32000F;
+        this.zOffset = (float)par1DataInputStream.readShort() / 32000F;
+        // END FCMOD
     }
 
     /**
@@ -62,9 +69,16 @@
         par1DataOutputStream.writeInt(this.zPosition);
         par1DataOutputStream.write(this.direction);
         writeItemStack(this.itemStack, par1DataOutputStream);
+        // FCMOD: Code change to fix rounding errors
+        /*
         par1DataOutputStream.write((int)(this.xOffset * 16.0F));
         par1DataOutputStream.write((int)(this.yOffset * 16.0F));
         par1DataOutputStream.write((int)(this.zOffset * 16.0F));
+        */
+        par1DataOutputStream.writeShort((int)(this.xOffset * 32000F + 0.5F));
+        par1DataOutputStream.writeShort((int)(this.yOffset * 32000F + 0.5F));
+        par1DataOutputStream.writeShort((int)(this.zOffset * 32000F + 0.5F));
+        // END FCMOD        
     }
 
     /**
@@ -80,7 +94,12 @@
      */
     public int getPacketSize()
     {
+    	// FCMOD: Code change
+    	/*
         return 19;
+        */
+        return 22;
+        // END FCMOD
     }
 
     public int getXPosition()
