--- a/minecraft/net/minecraft/src/RandomPositionGenerator.java
+++ b/minecraft/net/minecraft/src/RandomPositionGenerator.java
@@ -67,9 +67,14 @@
 
         for (int var16 = 0; var16 < 10; ++var16)
         {
-            int var12 = var4.nextInt(2 * par1) - par1;
-            int var17 = var4.nextInt(2 * par2) - par2;
-            int var14 = var4.nextInt(2 * par1) - par1;
+        	// FCMOD: Changed to fix bias towards the northwest
+            //int var12 = var4.nextInt(2 * par1) - par1;
+            //int var17 = var4.nextInt(2 * par2) - par2;
+            //int var14 = var4.nextInt(2 * par1) - par1;
+            int var12 = var4.nextInt( 2 * par1 + 1 ) - par1;
+            int var17 = var4.nextInt( 2 * par2 + 1 ) - par2;
+            int var14 = var4.nextInt( 2 * par1 + 1 ) - par1;
+            // END FCMOD
 
             if (par3Vec3 == null || (double)var12 * par3Vec3.xCoord + (double)var14 * par3Vec3.zCoord >= 0.0D)
             {
