--- a/minecraft/net/minecraft/src/RenderSnowball.java
+++ b/minecraft/net/minecraft/src/RenderSnowball.java
@@ -1,5 +1,6 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.cit.CITUtils;
 import org.lwjgl.opengl.GL11;
 import org.lwjgl.opengl.GL12;
 
@@ -27,7 +28,7 @@
      */
     public void doRender(Entity par1Entity, double par2, double par4, double par6, float par8, float par9)
     {
-        Icon var10 = this.field_94151_a.getIconFromDamage(this.field_94150_f);
+    	Icon var10 = CITUtils.getEntityIcon(this.field_94151_a.getIconFromDamage(this.field_94150_f), par1Entity);
 
         if (var10 != null)
         {
