--- a/minecraft/net/minecraft/src/RenderXPOrb.java
+++ b/minecraft/net/minecraft/src/RenderXPOrb.java
@@ -38,6 +38,14 @@
         var21 = (int)((MathHelper.sin(var27 + 0.0F) + 1.0F) * 0.5F * var26);
         int var22 = (int)var26;
         int var23 = (int)((MathHelper.sin(var27 + 4.1887903F) + 1.0F) * 0.1F * var26);
+        // FCMOD: Code added
+        if ( par1EntityXPOrb.m_bNotPlayerOwned )
+        {
+        	var21 = (int)((MathHelper.sin(var27 / 10.0F) + 1.0F) * 0.5F * var26);
+        	var22 = var23 = 0;
+        	var21 = ( var21 >> 2 ) + 64;
+        }
+        // END FCMOD
         int var24 = var21 << 16 | var22 << 8 | var23;
         GL11.glRotatef(180.0F - this.renderManager.playerViewY, 0.0F, 1.0F, 0.0F);
         GL11.glRotatef(-this.renderManager.playerViewX, 1.0F, 0.0F, 0.0F);
