--- a/minecraft/net/minecraft/src/RenderSpider.java
+++ b/minecraft/net/minecraft/src/RenderSpider.java
@@ -1,5 +1,9 @@
+// FCMOD: Class changes deprecated 06/15/2018
+
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.mal.resource.FakeResourceLocation;
+import com.prupe.mcpatcher.mob.MobRandomizer;
 import org.lwjgl.opengl.GL11;
 
 public class RenderSpider extends RenderLiving
@@ -26,7 +30,7 @@
         }
         else
         {
-            this.loadTexture("/mob/spider_eyes.png");
+        	this.loadTexture(FakeResourceLocation.unwrap(MobRandomizer.randomTexture((Entity)par1EntitySpider, FakeResourceLocation.wrap("/mob/spider_eyes.png"))));
             float var4 = 1.0F;
             GL11.glEnable(GL11.GL_BLEND);
             GL11.glDisable(GL11.GL_ALPHA_TEST);
