--- a/minecraft/net/minecraft/src/RenderManager.java
+++ b/minecraft/net/minecraft/src/RenderManager.java
@@ -329,4 +329,13 @@
             var3.updateIcons(par1IconRegister);
         }
     }
+    
+    // FCMOD: Added (client only)
+    public static void AddEntityRenderer( Class entityClass, Render entityRenderer )
+    {
+        instance.entityRenderMap.put( entityClass, entityRenderer );
+        
+        entityRenderer.setRenderManager( instance );
 }
+    // END FCMOD
+}
