--- a/minecraft/net/minecraft/src/RenderXPOrb.java
+++ b/minecraft/net/minecraft/src/RenderXPOrb.java
@@ -1,5 +1,6 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.cc.ColorizeEntity;
 import org.lwjgl.opengl.GL11;
 import org.lwjgl.opengl.GL12;
 
@@ -38,13 +39,21 @@
         var21 = (int)((MathHelper.sin(var27 + 0.0F) + 1.0F) * 0.5F * var26);
         int var22 = (int)var26;
         int var23 = (int)((MathHelper.sin(var27 + 4.1887903F) + 1.0F) * 0.1F * var26);
+        // FCMOD: Code added
+        if ( par1EntityXPOrb.m_bNotPlayerOwned )
+        {
+        	var21 = (int)((MathHelper.sin(var27 / 10.0F) + 1.0F) * 0.5F * var26);
+        	var22 = var23 = 0;
+        	var21 = ( var21 >> 2 ) + 64;
+        }
+        // END FCMOD
         int var24 = var21 << 16 | var22 << 8 | var23;
         GL11.glRotatef(180.0F - this.renderManager.playerViewY, 0.0F, 1.0F, 0.0F);
         GL11.glRotatef(-this.renderManager.playerViewX, 1.0F, 0.0F, 0.0F);
         float var25 = 0.3F;
         GL11.glScalef(var25, var25, var25);
         var11.startDrawingQuads();
-        var11.setColorRGBA_I(var24, 128);
+        var11.setColorRGBA_I(ColorizeEntity.colorizeXPOrb(var24, var27), 128);
         var11.setNormal(0.0F, 1.0F, 0.0F);
         var11.addVertexWithUV((double)(0.0F - var17), (double)(0.0F - var18), 0.0D, (double)var12, (double)var15);
         var11.addVertexWithUV((double)(var16 - var17), (double)(0.0F - var18), 0.0D, (double)var13, (double)var15);
