--- a/minecraft/net/minecraft/src/RenderBiped.java
+++ b/minecraft/net/minecraft/src/RenderBiped.java
@@ -45,6 +45,12 @@
             if (var5 instanceof ItemArmor)
             {
                 ItemArmor var6 = (ItemArmor)var5;
+                // FCMOD: Code added
+                if ( var5 instanceof FCItemArmorMod )
+                {
+                	return ShouldRenderPassModArmor( var4, par2, (FCItemArmorMod)var5 );
+                }
+            	// END FCMOD
                 this.loadTexture("/armor/" + bipedArmorFilenamePrefix[var6.renderIndex] + "_" + (par2 == 2 ? 2 : 1) + ".png");
                 ModelBiped var7 = par2 == 2 ? this.field_82425_h : this.field_82423_g;
                 var7.bipedHead.showModel = par2 == 0;
@@ -113,6 +119,14 @@
 
             if (var5 instanceof ItemArmor)
             {
+            	// FCMOD
+            	if ( var5 instanceof FCItemArmorMod )
+            	{
+            		LoadSecondLayerOfModArmorTexture( par2, (FCItemArmorMod)var5 );
+            		
+                	return;
+            	}
+            	// END FCMOD
                 ItemArmor var6 = (ItemArmor)var5;
                 this.loadTexture("/armor/" + bipedArmorFilenamePrefix[var6.renderIndex] + "_" + (par2 == 2 ? 2 : 1) + "_b.png");
                 float var7 = 1.0F;
@@ -160,9 +174,17 @@
             GL11.glPushMatrix();
             this.modelBipedMain.bipedHead.postRender(0.0625F);
 
-            if (var5.getItem().itemID < 256)
+            // FCMOD: Change
+            //if (var5.getItem().itemID < 256)
+            if ( var5.getItem().itemID < 4096 && Block.blocksList[var5.itemID] != null )
+        	// END FCMOD
             {
+                // FCMOD: Change
+                /*
                 if (RenderBlocks.renderItemIn3d(Block.blocksList[var5.itemID].getRenderType()))
+                */
+                if ( Block.blocksList[var5.itemID].DoesItemRenderAsBlock( var5.getItemDamage() ) )
+            	// END FCMOD
                 {
                     var6 = 0.625F;
                     GL11.glTranslatef(0.0F, -0.25F, 0.0F);
@@ -204,7 +226,12 @@
             this.modelBipedMain.bipedRightArm.postRender(0.0625F);
             GL11.glTranslatef(-0.0625F, 0.4375F, 0.0625F);
 
+            // FCMOD: Change
+            /*
             if (var4.itemID < 256 && RenderBlocks.renderItemIn3d(Block.blocksList[var4.itemID].getRenderType()))
+            */
+            if ( var4.itemID < 4096 && Block.blocksList[var4.itemID] != null && Block.blocksList[var4.itemID].DoesItemRenderAsBlock( var4.getItemDamage() ) )
+            // END FCMOD
             {
                 var6 = 0.5F;
                 GL11.glTranslatef(0.0F, 0.1875F, -0.3125F);
@@ -273,4 +300,71 @@
     {
         this.doRenderLiving((EntityLiving)par1Entity, par2, par4, par6, par8, par9);
     }
+    
+    // FCMOD: Function added
+    private int ShouldRenderPassModArmor( ItemStack stack, int iArmorSlot, FCItemArmorMod armorItem )
+    {
+        loadTexture( armorItem.GetWornTextureDirectory() + armorItem.GetWornTexturePrefix() + "_" + (iArmorSlot == 2 ? 2 : 1) + ".png");
+        
+        ModelBiped model = iArmorSlot == 2 ? this.field_82425_h : this.field_82423_g;
+        
+        model.bipedHead.showModel = iArmorSlot == 0;
+        model.bipedHeadwear.showModel = iArmorSlot == 0;
+        model.bipedBody.showModel = iArmorSlot == 1 || iArmorSlot == 2;
+        model.bipedRightArm.showModel = iArmorSlot == 1;
+        model.bipedLeftArm.showModel = iArmorSlot == 1;
+        model.bipedRightLeg.showModel = iArmorSlot == 2 || iArmorSlot == 3;
+        model.bipedLeftLeg.showModel = iArmorSlot == 2 || iArmorSlot == 3;
+        
+        setRenderPassModel(model);
+
+        if (model != null)
+        {
+            model.onGround = this.mainModel.onGround;
+            model.isRiding = this.mainModel.isRiding;
+            model.isChild = this.mainModel.isChild;
 }
+
+        if ( armorItem.HasCustomColors() )
+        {
+            int iColor = armorItem.getColor( stack );
+            
+            float fRed = (float)(iColor >> 16 & 255) / 255.0F;
+            float fGreen = (float)(iColor >> 8 & 255) / 255.0F;
+            float fBlue = (float)(iColor & 255) / 255.0F;
+            
+            GL11.glColor3f( fRed, fGreen, fBlue);
+        }
+        else
+        {
+            GL11.glColor3f( 1F, 1F, 1F);
+        }
+
+        if ( armorItem.HasSecondRenderLayerWhenWorn() )
+        {
+	        if (stack.isItemEnchanted())
+	        {
+	            return 31;
+	        }
+	
+	        return 16;
+        }
+        else
+        {        
+	        if ( stack.isItemEnchanted() )
+	        {
+	            return 15;
+	        }
+	
+	        return 1;
+        }
+    }
+    
+    private void LoadSecondLayerOfModArmorTexture( int iArmorSlot, FCItemArmorMod armorItem )
+    {
+        loadTexture( armorItem.GetWornTextureDirectory() + armorItem.GetWornTexturePrefix() + "_" + (iArmorSlot == 2 ? 2 : 1) + "_b.png");
+	    
+	    GL11.glColor3f(1F, 1F, 1F);
+    }    
+    // END FCMOD
+}
