--- a/minecraft/net/minecraft/src/RenderFish.java
+++ b/minecraft/net/minecraft/src/RenderFish.java
@@ -1,5 +1,6 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.mob.LineRenderer;
 import net.minecraft.client.Minecraft;
 import org.lwjgl.opengl.GL11;
 import org.lwjgl.opengl.GL12;
@@ -68,6 +69,9 @@
             double var37 = (double)((float)(var23 - var46));
             double var39 = (double)((float)(var25 - var33));
             double var41 = (double)((float)(var27 - var35));
+            
+            if (!LineRenderer.renderLine(0, par2, par4, par6, var37, var39, var41))
+            {
             GL11.glDisable(GL11.GL_TEXTURE_2D);
             GL11.glDisable(GL11.GL_LIGHTING);
             var12.startDrawing(3);
@@ -85,6 +89,7 @@
             GL11.glEnable(GL11.GL_TEXTURE_2D);
         }
     }
+    }
 
     /**
      * Actually renders the given argument. This is a synthetic bridge method, always casting down its argument and then
