--- a/minecraft/net/minecraft/src/RenderMooshroom.java
+++ b/minecraft/net/minecraft/src/RenderMooshroom.java
@@ -1,5 +1,7 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.mal.resource.FakeResourceLocation;
+import com.prupe.mcpatcher.mob.MobOverlay;
 import org.lwjgl.opengl.GL11;
 
 public class RenderMooshroom extends RenderLiving
@@ -7,6 +9,11 @@
     public RenderMooshroom(ModelBase par1ModelBase, float par2)
     {
         super(par1ModelBase, par2);
+        // FCMOD: Added (client only)        
+        m_ModelUdder = new FCClientModelCowUdder();
+        
+        setRenderPassModel(m_ModelUdder);
+        // END FCMOD
     }
 
     public void renderLivingMooshroom(EntityMooshroom par1EntityMooshroom, double par2, double par4, double par6, float par8, float par9)
@@ -18,27 +25,53 @@
     {
         super.renderEquippedItems(par1EntityMooshroom, par2);
 
-        if (!par1EntityMooshroom.isChild())
+        if (par1EntityMooshroom.isChild())
         {
-            this.loadTexture("/terrain.png");
+        	MobOverlay.finishMooshroom();
+        }
+        else
+        {
+            this.loadTexture(FakeResourceLocation.unwrap(MobOverlay.setupMooshroom(par1EntityMooshroom, FakeResourceLocation.wrap("/terrain.png"))));
             GL11.glEnable(GL11.GL_CULL_FACE);
             GL11.glPushMatrix();
             GL11.glScalef(1.0F, -1.0F, 1.0F);
             GL11.glTranslatef(0.2F, 0.4F, 0.5F);
             GL11.glRotatef(42.0F, 0.0F, 1.0F, 0.0F);
+            
+            if (!MobOverlay.renderMooshroomOverlay(0.0D))
+            {
             this.renderBlocks.renderBlockAsItem(Block.mushroomRed, 0, 1.0F);
+            }
+            
             GL11.glTranslatef(0.1F, 0.0F, -0.6F);
             GL11.glRotatef(42.0F, 0.0F, 1.0F, 0.0F);
+            // FCMOD: Added (client only) to not display mid mushroom if starving
+            if ( !par1EntityMooshroom.IsStarving() )
+        	// END FCMOD
+            	
+        	if (!MobOverlay.renderMooshroomOverlay(0.0D))
+            {
             this.renderBlocks.renderBlockAsItem(Block.mushroomRed, 0, 1.0F);
+            }
+            
             GL11.glPopMatrix();
             GL11.glPushMatrix();
             ((ModelQuadruped)this.mainModel).head.postRender(0.0625F);
             GL11.glScalef(1.0F, -1.0F, 1.0F);
             GL11.glTranslatef(0.0F, 0.75F, -0.2F);
             GL11.glRotatef(12.0F, 0.0F, 1.0F, 0.0F);
+            // FCMOD: Added (client only) to not display head mushroom if not full
+            if ( par1EntityMooshroom.IsFullyFed() )
+        	// END FCMOD
+            	
+        	if (!MobOverlay.renderMooshroomOverlay(0.0D))
+            {
             this.renderBlocks.renderBlockAsItem(Block.mushroomRed, 0, 1.0F);
+            }
+            
             GL11.glPopMatrix();
             GL11.glDisable(GL11.GL_CULL_FACE);
+            MobOverlay.finishMooshroom();
         }
     }
 
@@ -62,4 +95,20 @@
     {
         this.renderLivingMooshroom((EntityMooshroom)par1Entity, par2, par4, par6, par8, par9);
     }
+    
+    // FCMOD: Added New (client only)
+    FCClientModelCowUdder m_ModelUdder;
+    
+    protected int shouldRenderPass(EntityLiving par1EntityLiving, int par2, float par3)
+    {
+    	if ( par2 == 0 && ((FCEntityCow)par1EntityLiving).GotMilk() )
+    	{
+    		loadTexture( "/btwmodtex/cow_udder.png" );
+    		
+    		return 1;
 }
+    	
+        return -1;
+    }
+    // END FCMOD
+}
