--- a/minecraft/net/minecraft/src/SaveHandler.java
+++ b/minecraft/net/minecraft/src/SaveHandler.java
@@ -345,4 +345,34 @@
     {
         return this.saveDirectoryName;
     }
+    
+    // FCMOD: Added
+    @Override
+    public void LoadModSpecificData( WorldServer world )
+    {
+        world.loadWorldDataFromNBT(GetDimensionDirectory(world));
 }
+    
+    @Override
+    public void SaveModSpecificData( WorldServer world )
+    {
+        world.saveWorldDataToNBT(GetDimensionDirectory(world));
+    }
+    
+    private File GetDimensionDirectory( WorldServer world )
+    {
+        if ( world.provider.dimensionId == -1 ) // nether
+        {
+            return new File( worldDirectory, "DIM-1");
+        }
+        else if ( world.provider.dimensionId == 1 ) // the end
+        {
+        	return new File( worldDirectory, "DIM1");
+        }
+        else
+        {
+            return worldDirectory;
+        }        
+    }
+    // END FCMOD
+}
