--- a/minecraft/net/minecraft/src/SoundPool.java
+++ b/minecraft/net/minecraft/src/SoundPool.java
@@ -40,6 +40,23 @@
 
             if (this.isGetRandomSound)
             {
+            	// FCMOD: Code added to also add the file independently so that it may be called specifically (client only) 
+                if (Character.isDigit(par1Str.charAt(par1Str.length() - 1)))
+                {
+	                par1Str = par1Str.replaceAll("/", ".");
+	
+	                if (!this.nameToSoundPoolEntriesMapping.containsKey(par1Str))
+	                {
+	                    this.nameToSoundPoolEntriesMapping.put(par1Str, new ArrayList());
+	                }
+	
+	                SoundPoolEntry var4 = new SoundPoolEntry(var3, par2File.toURI().toURL());
+	                ((List)this.nameToSoundPoolEntriesMapping.get(par1Str)).add(var4);
+	                this.allSoundPoolEntries.add(var4);
+	                ++this.numberOfSoundPoolEntries;
+                }
+                // END FCMOD
+            	
                 while (Character.isDigit(par1Str.charAt(par1Str.length() - 1)))
                 {
                     par1Str = par1Str.substring(0, par1Str.length() - 1);
