--- a/minecraft/net/minecraft/src/StepSound.java
+++ b/minecraft/net/minecraft/src/StepSound.java
@@ -18,11 +18,35 @@
         return this.stepSoundVolume;
     }
 
+    public float getStepVolume() {
+    	return this.getVolume();
+    }
+    
+    public float getPlaceVolume() {
+    	return this.getVolume();
+    }
+    
+    public float getBreakVolume() {
+    	return this.getVolume();
+    }
+
     public float getPitch()
     {
         return this.stepSoundPitch;
     }
 
+    public float getStepPitch() {
+    	return this.getPitch();
+    }
+    
+    public float getPlacePitch() {
+    	return this.getPitch();
+    }
+    
+    public float getBreakPitch() {
+    	return this.getPitch();
+    }
+
     /**
      * Used when a block breaks, EXA: Player break, Shep eating grass, etc..
      */
