--- a/minecraft/net/minecraft/src/StructureScatteredFeatureStart.java
+++ b/minecraft/net/minecraft/src/StructureScatteredFeatureStart.java
@@ -1,29 +1,43 @@
 package net.minecraft.src;
 
+import java.util.ArrayList;
 import java.util.Random;
 
 public class StructureScatteredFeatureStart extends StructureStart
 {
-    public StructureScatteredFeatureStart(World par1World, Random par2Random, int par3, int par4)
+	private static ArrayList<BiomeGenBase> desertBiomeList = new ArrayList();
+	private static ArrayList<BiomeGenBase> jungleBiomeList = new ArrayList();
+	private static ArrayList<BiomeGenBase> swampBiomeList = new ArrayList();
+	
+	public static void addDesertBiome(BiomeGenBase biome) {
+		desertBiomeList.add(biome);
+	}
+	
+	public static void addJungleBiome(BiomeGenBase biome) {
+		jungleBiomeList.add(biome);
+	}
+	
+	public static void addSwampBiome(BiomeGenBase biome) {
+		swampBiomeList.add(biome);
+	}
+	
+	public StructureScatteredFeatureStart(World world, Random rand, int chunkX, int chunkZ)
     {
-        BiomeGenBase var5 = par1World.getBiomeGenForCoords(par3 * 16 + 8, par4 * 16 + 8);
+		BiomeGenBase biome = world.getBiomeGenForCoords(chunkX * 16 + 8, chunkZ * 16 + 8);
 
-        if (var5 != BiomeGenBase.jungle && var5 != BiomeGenBase.jungleHills)
+		if (swampBiomeList.contains(biome))
         {
-            if (var5 == BiomeGenBase.swampland)
-            {
-                ComponentScatteredFeatureSwampHut var7 = new ComponentScatteredFeatureSwampHut(par2Random, par3 * 16, par4 * 16);
+			ComponentScatteredFeatureSwampHut var7 = new ComponentScatteredFeatureSwampHut(rand, chunkX * 16, chunkZ * 16);
                 this.components.add(var7);
             }
-            else
+		else if (desertBiomeList.contains(biome))
             {
-                ComponentScatteredFeatureDesertPyramid var8 = new ComponentScatteredFeatureDesertPyramid(par2Random, par3 * 16, par4 * 16);
+			ComponentScatteredFeatureDesertPyramid var8 = new ComponentScatteredFeatureDesertPyramid(rand, chunkX * 16, chunkZ * 16);
                 this.components.add(var8);
             }
-        }
-        else
+		else if(jungleBiomeList.contains(biome))
         {
-            ComponentScatteredFeatureJunglePyramid var6 = new ComponentScatteredFeatureJunglePyramid(par2Random, par3 * 16, par4 * 16);
+			ComponentScatteredFeatureJunglePyramid var6 = new ComponentScatteredFeatureJunglePyramid(rand, chunkX * 16, chunkZ * 16);
             this.components.add(var6);
         }
 
