--- a/minecraft/net/minecraft/src/StructureStrongholdStones.java
+++ b/minecraft/net/minecraft/src/StructureStrongholdStones.java
@@ -24,10 +24,20 @@
             {
                 this.selectedBlockMetaData = 1;
             }
+            else if (var6 < 0.52F)
+            {
+            	this.selectedBlockId = FCBetterThanWolves.fcBlockSilverfishStoneBricks.blockID;
+                this.selectedBlockMetaData = 0;
+            }
+            else if (var6 < 0.535F)
+            {
+            	this.selectedBlockId = FCBetterThanWolves.fcBlockSilverfishStoneBricksMossy.blockID;
+                this.selectedBlockMetaData = 0;
+            }
             else if (var6 < 0.55F)
             {
-                this.selectedBlockId = Block.silverfish.blockID;
-                this.selectedBlockMetaData = 2;
+            	this.selectedBlockId = FCBetterThanWolves.fcBlockSilverfishStoneBricksCracked.blockID;
+                this.selectedBlockMetaData = 0;
             }
             else
             {
