--- a/minecraft/net/minecraft/src/TextureClock.java
+++ b/minecraft/net/minecraft/src/TextureClock.java
@@ -1,15 +1,20 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.hd.FancyDial;
 import net.minecraft.client.Minecraft;
 
 public class TextureClock extends TextureStitched
 {
-    private double field_94239_h;
+	public double field_94239_h;
     private double field_94240_i;
 
     public TextureClock()
     {
         super("compass");
+        FancyDial.setup(this);
+        // FCMOD: Code added
+        m_clockTexture = this;
+        // END FCMOD
     }
 
     public void updateAnimation()
@@ -53,6 +58,8 @@
         this.field_94240_i += var7 * 0.1D;
         this.field_94240_i *= 0.8D;
         this.field_94239_h += this.field_94240_i;
+        if (!FancyDial.update(this, false))
+        {
         int var6;
 
         for (var6 = (int)((this.field_94239_h + 1.0D) * (double)this.textureList.size()) % this.textureList.size(); var6 < 0; var6 = (var6 + this.textureList.size()) % this.textureList.size())
@@ -67,3 +74,14 @@
         }
     }
 }
+    
+    // FCMOD: Code added
+    public static TextureClock m_clockTexture;
+    
+    public void UpdateInert()
+    {
+        this.frameCounter = 0;
+        this.textureSheet.copyFrom(this.originX, this.originY, (Texture)this.textureList.get(this.frameCounter), this.rotated);
+    }
+    // END FCMOD
+}
