--- a/minecraft/net/minecraft/src/TextureStitched.java
+++ b/minecraft/net/minecraft/src/TextureStitched.java
@@ -1,5 +1,6 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.hd.Wrapper15;
 import java.io.BufferedReader;
 import java.util.ArrayList;
 import java.util.List;
@@ -36,10 +37,10 @@
 
     public static TextureStitched makeTextureStitched(String par0Str)
     {
-        return (TextureStitched)("clock".equals(par0Str) ? new TextureClock() : ("compass".equals(par0Str) ? new TextureCompass() : new TextureStitched(par0Str)));
+    	return (TextureStitched)("clock".equals(par0Str) ? new TextureClock() : ("compass".equals(par0Str) ? new TextureCompass() : Wrapper15.createSprite(par0Str)));
     }
 
-    protected TextureStitched(String par1)
+    public TextureStitched(String par1)
     {
         this.textureName = par1;
     }
@@ -233,4 +234,11 @@
             this.listAnimationTuples = var2;
         }
     }
+    
+    // FCMOD: Code added
+    public boolean IsProcedurallyAnimated()
+    {
+    	return false;
 }
+    // END FCMOD
+}
