--- a/minecraft/net/minecraft/src/TileEntitySkullRenderer.java
+++ b/minecraft/net/minecraft/src/TileEntitySkullRenderer.java
@@ -8,7 +8,11 @@
     public static TileEntitySkullRenderer skullRenderer;
     private ModelSkeletonHead field_82396_c = new ModelSkeletonHead(0, 0, 64, 32);
     private ModelSkeletonHead field_82395_d = new ModelSkeletonHead(0, 0, 64, 64);
+    // FCMOD: Added
+    private ModelSkeletonHead m_InfusedModel = new ModelSkeletonHead(0, 0, 32, 16);
+    // END FCMOD
 
+
     /**
      * Render a skull tile entity.
      */
@@ -49,7 +53,10 @@
             case 3:
                 if (par7Str != null && par7Str.length() > 0)
                 {
-                    String var9 = "http://skins.minecraft.net/MinecraftSkins/" + StringUtils.stripControlCodes(par7Str) + ".png";
+                	// FCMOD: Changed, client only        	
+                    //String var9 = "http://skins.minecraft.net/MinecraftSkins/" + StringUtils.stripControlCodes(par7Str) + ".png";
+                    String var9 = FCBetterThanWolves.fcPlayerSkinURL + StringUtils.stripControlCodes(par7Str) + ".png";
+                    // END FCMOD
 
                     if (!skullRenderer.tileEntityRenderer.renderEngine.hasImageData(var9))
                     {
@@ -67,6 +74,13 @@
 
             case 4:
                 this.bindTextureByName("/mob/creeper.png");
+            // FCMOD: Added
+                break;
+            case 5:
+                this.bindTextureByName("/btwmodtex/fcInfusedSkull.png");
+                var8 = m_InfusedModel;
+                break;
+            // END FCMOD
         }
 
         GL11.glPushMatrix();
@@ -106,6 +120,12 @@
         GL11.glScalef(-1.0F, -1.0F, 1.0F);
         GL11.glEnable(GL11.GL_ALPHA_TEST);
         var8.render((Entity)null, 0.0F, 0.0F, 0.0F, par5, 0.0F, var10);
+        // FCMOD: Added
+        if ( par6 == 5 )
+        {
+        	RenderInfusedEyes( var8, par5);
+        }
+        // END FCMOD
         GL11.glPopMatrix();
     }
 
@@ -113,4 +133,28 @@
     {
         this.renderTileEntitySkullAt((TileEntitySkull)par1TileEntity, par2, par4, par6, par8);
     }
+    
+    // FCMOD: Added
+    private void RenderInfusedEyes( ModelSkeletonHead model, float fYaw )
+    {
+        bindTextureByName("/btwmodtex/fcInfusedSkullEyes.png");
+        float var4 = 1.0F;
+        GL11.glEnable(GL11.GL_BLEND);
+        GL11.glDisable(GL11.GL_ALPHA_TEST);
+        GL11.glBlendFunc(GL11.GL_ONE, GL11.GL_ONE);
+        GL11.glDisable(GL11.GL_LIGHTING);
+        GL11.glDepthMask(true);
+        char var5 = 61680;
+        int var6 = var5 % 65536;
+        int var7 = var5 / 65536;
+        OpenGlHelper.setLightmapTextureCoords(OpenGlHelper.lightmapTexUnit, (float)var6 / 1.0F, (float)var7 / 1.0F);
+        GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
+        GL11.glEnable(GL11.GL_LIGHTING);
+        GL11.glEnable(GL11.GL_ALPHA_TEST);
+        GL11.glDisable(GL11.GL_BLEND);
+        GL11.glColor4f(1.0F, 1.0F, 1.0F, var4);
+        
+        model.render((Entity)null, 0.0F, 0.0F, 0.0F, fYaw, 0.0F, 0.0625F);
 }
+    // END FCMOD    
+}
