--- a/minecraft/net/minecraft/src/TileEntityRendererPiston.java
+++ b/minecraft/net/minecraft/src/TileEntityRendererPiston.java
@@ -48,8 +48,28 @@
             }
             else
             {
+            	// FCMOD: Changed (Client only)
+            	/*
                 this.blockRenderer.renderBlockAllFaces(var9, par1TileEntityPiston.xCoord, par1TileEntityPiston.yCoord, par1TileEntityPiston.zCoord);
+                */
+            	
+            	// I get the impression that under rare circumstances metadata can get out of sync
+            	// resulting in blocks attempting to render with erroneous data, which
+            	// would explain what's happening here: 
+            	// http://www.sargunster.com/btwforum/viewtopic.php?f=7&t=9630
+            	if ( par1TileEntityPiston.getBlockMetadata() == 
+            		par1TileEntityPiston.worldObj.getBlockMetadata(
+        			par1TileEntityPiston.xCoord,
+        			par1TileEntityPiston.yCoord,
+        			par1TileEntityPiston.zCoord ) )
+    			{
+	            	var9.m_currentBlockRenderer = blockRenderer;
+	            	
+	        		var9.RenderBlockMovedByPiston( blockRenderer, par1TileEntityPiston.xCoord, 
+	        			par1TileEntityPiston.yCoord, par1TileEntityPiston.zCoord );
             }
+        		// END FCMOD
+            }
 
             var10.setTranslation(0.0D, 0.0D, 0.0D);
             var10.draw();
