--- a/minecraft/net/minecraft/src/TileEntitySignRenderer.java
+++ b/minecraft/net/minecraft/src/TileEntitySignRenderer.java
@@ -1,5 +1,6 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.cc.ColorizeWorld;
 import org.lwjgl.opengl.GL11;
 
 public class TileEntitySignRenderer extends TileEntitySpecialRenderer
@@ -58,7 +59,7 @@
         GL11.glScalef(var12, -var12, var12);
         GL11.glNormal3f(0.0F, 0.0F, -1.0F * var12);
         GL11.glDepthMask(false);
-        byte var13 = 0;
+        int var13 = ColorizeWorld.colorizeSignText();
 
         for (int var14 = 0; var14 < par1TileEntitySign.signText.length; ++var14)
         {
