--- a/minecraft/net/minecraft/src/WorldGenReed.java
+++ b/minecraft/net/minecraft/src/WorldGenReed.java
@@ -1,30 +1,64 @@
 package net.minecraft.src;
 
+import java.util.ArrayList;
 import java.util.Random;
 
 public class WorldGenReed extends WorldGenerator
 {
-    public boolean generate(World par1World, Random par2Random, int par3, int par4, int par5)
+	private static ArrayList<BiomeGenBase> validBiomeList = new ArrayList();
+	
+	public static boolean isBiomeValid(BiomeGenBase biome) {
+		return validBiomeList.contains(biome);
+	}
+	
+	public static void addBiomeToGenerator(BiomeGenBase biome) {
+		validBiomeList.add(biome);
+	}
+	
+    public boolean generate(World par1World, Random par2Random, int x, int y, int z)
     {
+    	// FCMOD: Added
+    	BiomeGenBase currentBiome = par1World.getBiomeGenForCoords( x, z );
+    	
+        boolean isValidBiome = isBiomeValid(currentBiome);
+        // END FCMOD
+        
         for (int var6 = 0; var6 < 20; ++var6)
         {
-            int var7 = par3 + par2Random.nextInt(4) - par2Random.nextInt(4);
-            int var8 = par4;
-            int var9 = par5 + par2Random.nextInt(4) - par2Random.nextInt(4);
+            int plantX = x + par2Random.nextInt(4) - par2Random.nextInt(4);
+            int plantY = y;
+            int plantZ = z + par2Random.nextInt(4) - par2Random.nextInt(4);
 
-            if (par1World.isAirBlock(var7, par4, var9) && (par1World.getBlockMaterial(var7 - 1, par4 - 1, var9) == Material.water || par1World.getBlockMaterial(var7 + 1, par4 - 1, var9) == Material.water || par1World.getBlockMaterial(var7, par4 - 1, var9 - 1) == Material.water || par1World.getBlockMaterial(var7, par4 - 1, var9 + 1) == Material.water))
+            if (par1World.isAirBlock(plantX, y, plantZ) && (par1World.getBlockMaterial(plantX - 1, y - 1, plantZ) == Material.water || par1World.getBlockMaterial(plantX + 1, y - 1, plantZ) == Material.water || par1World.getBlockMaterial(plantX, y - 1, plantZ - 1) == Material.water || par1World.getBlockMaterial(plantX, y - 1, plantZ + 1) == Material.water))
             {
                 int var10 = 2 + par2Random.nextInt(par2Random.nextInt(3) + 1);
 
-                for (int var11 = 0; var11 < var10; ++var11)
+                // FCMOD: Added
+                if ( !isValidBiome )
                 {
-                    if (Block.reed.canBlockStay(par1World, var7, var8 + var11, var9))
+                	// must occur after all random number generation to avoid messing up world gen
+                	
+                	continue;
+                }
+
+                for (int i = 0; i < var10; ++i)
                     {
-                        par1World.setBlock(var7, var8 + var11, var9, Block.reed.blockID, 0, 2);
+                	Block reedBlock;
+                	
+                	if (i == 0) {
+                		reedBlock = FCBetterThanWolves.fcBlockReedRoots;
                     }
+                	else {
+                		reedBlock = FCBetterThanWolves.fcBlockReeds;
                 }
+                	
+                    if (reedBlock.canBlockStay(par1World, plantX, plantY + i, plantZ)) {
+                        par1World.setBlock(plantX, plantY + i, plantZ, reedBlock.blockID, 0, 2);
             }
         }
+                // END FCMOD
+            }
+        }
 
         return true;
     }
