--- a/minecraft/net/minecraft/src/WorldChunkManager.java
+++ b/minecraft/net/minecraft/src/WorldChunkManager.java
@@ -26,8 +26,10 @@
         this.biomesToSpawnIn.add(BiomeGenBase.taiga);
         this.biomesToSpawnIn.add(BiomeGenBase.taigaHills);
         this.biomesToSpawnIn.add(BiomeGenBase.forestHills);
-        this.biomesToSpawnIn.add(BiomeGenBase.jungle);
-        this.biomesToSpawnIn.add(BiomeGenBase.jungleHills);
+        // FCMOD: Removing jungle biomes from the valid spawn locations due to jungle spiders
+        //this.biomesToSpawnIn.add(BiomeGenBase.jungle);
+        //this.biomesToSpawnIn.add(BiomeGenBase.jungleHills);
+        // END FCMOD
     }
 
     public WorldChunkManager(long par1, WorldType par3WorldType)
