--- a/minecraft/net/minecraft/src/WorldManager.java
+++ b/minecraft/net/minecraft/src/WorldManager.java
@@ -68,7 +68,10 @@
      */
     public void markBlockForUpdate(int par1, int par2, int par3)
     {
-        this.theWorldServer.getPlayerManager().flagChunkForUpdate(par1, par2, par3);
+    	// FCMOD: Changed
+        //this.theWorldServer.getPlayerManager().flagChunkForUpdate(par1, par2, par3);
+        theWorldServer.GetChunkTracker().FlagBlockForClientUpdate( par1, par2, par3 );
+        // END FCMOD
     }
 
     /**
