--- a/minecraft/net/minecraft/src/WorldProviderHell.java
+++ b/minecraft/net/minecraft/src/WorldProviderHell.java
@@ -1,5 +1,8 @@
 package net.minecraft.src;
 
+import com.prupe.mcpatcher.cc.ColorizeWorld;
+import com.prupe.mcpatcher.cc.ColorizeWorld;
+
 public class WorldProviderHell extends WorldProvider
 {
     /**
@@ -18,7 +21,7 @@
      */
     public Vec3 getFogColor(float par1, float par2)
     {
-        return this.worldObj.getWorldVec3Pool().getVecFromPool(0.20000000298023224D, 0.029999999329447746D, 0.029999999329447746D);
+    	return this.worldObj.getWorldVec3Pool().getVecFromPool((double)ColorizeWorld.netherFogColor[0], (double)ColorizeWorld.netherFogColor[1], (double)ColorizeWorld.netherFogColor[2]);
     }
 
     /**
