--- a/minecraft/net/minecraft/src/WorldGenMinable.java
+++ b/minecraft/net/minecraft/src/WorldGenMinable.java
@@ -66,14 +66,37 @@
 
                                 if (var39 * var39 + var42 * var42 + var45 * var45 < 1.0D && par1World.getBlockId(var38, var41, var44) == this.field_94523_c)
                                 {
+                                	// FCMOD: Code change
+                                	/*
                                     par1World.setBlock(var38, var41, var44, this.minableBlockId, 0, 2);
+                                    */
+                                	int iMetadata = 0;
+                                	Block block = Block.blocksList[minableBlockId];
+                                	
+                                	if ( block.HasStrata() )
+                                	{
+                                		if ( var41 <= 48  + par1World.rand.nextInt( 2 ) )
+                                		{
+                                			int iStrataLevel = 1;
+                                			
+                                			if ( var41 <= 24 + par1World.rand.nextInt( 2 ) )
+                                			{
+                                				iStrataLevel = 2;                                				
                                 }
+                                			
+                            				iMetadata = block.GetMetadataConversionForStrataLevel( iStrataLevel, 0 );
                             }
                         }
+                                	
+                                    par1World.setBlock(var38, var41, var44, this.minableBlockId, iMetadata, 2);
+                                    // END FCMOD
                     }
                 }
             }
         }
+                }
+            }
+        }
 
         return true;
     }
